{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "nbsphinx": "hidden"
   },
   "source": [
    "### NOTE FOR LUCA\n",
    "\n",
    "**Remember to set/remove metadata as:**\n",
    "{\n",
    "  \"nbsphinx\": \"hidden\"\n",
    "}\n",
    "\n",
    "to enable/disable solutions view"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sample exam \n",
    "\n",
    "This is the text of the exam done on 23/01/2019\n",
    "\n",
    "## Before you start\n",
    "**Please write one single python script for the lab part and one text file with the answers to the theoretical questions.**\n",
    "\n",
    "**IMPORTANT: Add your name and ID (matricola) on top of the .py and text files!**\n",
    "\n",
    "\n",
    "## Theory\n",
    "\n",
    "**Please write the solution in a text file**.\n",
    "\n",
    "### Exercise 1:\n",
    "\n",
    "Let L a list of size n, and i and j two indeces. Return the computational complexity of function fun() with respect to n.\n",
    "\n",
    "def fun(L,i,j):\n",
    "    if i==j:\n",
    "        return 0\n",
    "    else:\n",
    "        m = (i+j)//2\n",
    "        count = 0\n",
    "        for x in L[i:m]:\n",
    "          for y in L[m:j+1]:\n",
    "             if x==y:\n",
    "                count = count+1\n",
    "        left = fun(L,i,m)\n",
    "        right = fun(L,m+1,j)\n",
    "        return left+right+count \n",
    "\n",
    "\n",
    "## Practical part\n",
    "\n",
    "\n",
    "## Exercise 1:\n",
    "\n",
    "The file [snp_data.tsv](file_samples/snp_data.tsv) is a tab separated file containing the following information regarding SNPs:\n",
    "\n",
    "```\n",
    "Chr Pos Ref Alt ReadDepth\n",
    "Chr01 120 A T 120\n",
    "...\n",
    "```\n",
    "\n",
    "As the header says, the first column is the chromosome where the SNP has been identified, Pos is the position of the SNP, Ref is the reference base, Alt is the alternative base and ReadDepth is the read depth at that locus. \n",
    "\n",
    "Read the file in input and store it in a suitable data structure ```data``` (hint: pandas?) then write the following methods:\n",
    "\n",
    "1. ```computestats(data)``` gets the data structure loaded and prints the total number of chromosomes, the total number of SNPs, the total number of SNPs for each possible type (i.e. TA, TC, TG, TT, AT,AC, AG, AA,...) and the average read depth (with 3 decimal places of precision); \n",
    "\n",
    "\n",
    "2. ```chrHistogram(data, chromosome, start, end)``` given the input data, a chromosome, a start postion and an end position,  plots an histogram with the number of SNPs for each type belonging to that chromosome having postion in the closed interval [start, end];\n",
    "\n",
    "\n",
    "3. ```readDepthStats(data, chromosome, start, end)``` given the input data and chromosome, a start and end position, the function computes  the average read depth of all the SNPs belonging to that chromosome (and in the [start,end] interval) and the average read depth for each type of SNP (TA, TC, ...).\n",
    "\n",
    "\n",
    "After loading the data, calling \n",
    "\n",
    "```\n",
    "computestats(data)\n",
    "print(\"\\n\")\n",
    "chrHistogram(data, \"Chr01\",500,4000)\n",
    "chrHistogram(data, \"Chr02\",1,1250)\n",
    "chrHistogram(data, \"Chr03\",1,50000)\n",
    "readDepthStats(data, \"Chr01\", 500,6000)\n",
    "print(\"\\n\")\n",
    "readDepthStats(data, \"Chr02\",1, 2000)\n",
    "```\n",
    "\n",
    "should return \n",
    "\n",
    "```\n",
    "The file contains 10407 SNPs\n",
    "Belonging to 3 chromosomes:\n",
    "Chr01\n",
    "Chr02\n",
    "Chr03\n",
    "Type CT counts 681 SNPs\n",
    "Type GT counts 673 SNPs\n",
    "Type GA counts 670 SNPs\n",
    "Type GG counts 668 SNPs\n",
    "Type CG counts 666 SNPs\n",
    "Type TG counts 666 SNPs\n",
    "Type TC counts 663 SNPs\n",
    "Type AT counts 658 SNPs\n",
    "Type AG counts 658 SNPs\n",
    "Type CC counts 638 SNPs\n",
    "Type AA counts 637 SNPs\n",
    "Type TT counts 634 SNPs\n",
    "Type CA counts 630 SNPs\n",
    "Type AC counts 628 SNPs\n",
    "Type GC counts 621 SNPs\n",
    "Type TA counts 616 SNPs\n",
    "Average read depth: 62.999 reads per site\n",
    "```\n",
    "![](img/examJan19/plot1.png)\n",
    "\n",
    "![](img/examJan19/plot2.png)\n",
    "\n",
    "![](img/examJan19/plot3.png)\n",
    "```\n",
    "Global Avg. read depth for SNPs in Chr01 [500,6000]: 65.973\n",
    "\tType: AA avg. read depth: 83.250\n",
    "\tType: AC avg. read depth: 39.600\n",
    "\tType: AG avg. read depth: 83.750\n",
    "\tType: AT avg. read depth: 49.750\n",
    "\tType: CA avg. read depth: 73.200\n",
    "\tType: CC avg. read depth: 67.889\n",
    "\tType: CG avg. read depth: 61.727\n",
    "\tType: CT avg. read depth: 67.200\n",
    "\tType: GA avg. read depth: 61.364\n",
    "\tType: GC avg. read depth: 62.800\n",
    "\tType: GG avg. read depth: 75.900\n",
    "\tType: GT avg. read depth: 84.909\n",
    "\tType: TA avg. read depth: 58.125\n",
    "\tType: TC avg. read depth: 68.538\n",
    "\tType: TG avg. read depth: 61.000\n",
    "\tType: TT avg. read depth: 52.333\n",
    "\n",
    "\n",
    "Global Avg. read depth for SNPs in Chr02 [1,2000]: 68.771\n",
    "\tType: AA avg. read depth: 50.800\n",
    "\tType: AC avg. read depth: 63.250\n",
    "\tType: AG avg. read depth: 78.500\n",
    "\tType: AT avg. read depth: 106.500\n",
    "\tType: CA avg. read depth: 102.667\n",
    "\tType: CC avg. read depth: 109.500\n",
    "\tType: CG avg. read depth: 34.750\n",
    "\tType: CT avg. read depth: 54.667\n",
    "\tType: GA avg. read depth: 56.000\n",
    "\tType: GC avg. read depth: 61.000\n",
    "\tType: GG avg. read depth: 82.000\n",
    "\tType: GT avg. read depth: 62.000\n",
    "\tType: TA avg. read depth: 85.000\n",
    "\tType: TC avg. read depth: 79.000\n",
    "\tType: TG avg. read depth: 68.750\n",
    "```\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "## Exercise 2:\n",
    "\n",
    "Recall that a Binary Search Tree (BST), is binary tree with the following properties:\n",
    "\n",
    "a. the left subtree of a tree contains only nodes with values lower than that of the root;\n",
    "\n",
    "b. the right subtree of a tree contains only nodes with values greater than that of the root;\n",
    "\n",
    "c. no duplicate nodes are allowed.\n",
    "\n",
    "The ```partB.py``` code provided implements a BinaryTree as seen in the practicals. Modify it adding the following methods:\n",
    "\n",
    "1. ```findExtremes(BST)``` given a binary search tree BST, returns a tuple with two elements that are the smallest and the biggest element in the BST;\n",
    "\n",
    "2. ```sumNodes(BST, N)``` given a binary search tree BST and a node N, returns the sum of all the elements that are in the subtree rooted at N or None if N is not in BST.\n",
    "\n",
    "3. ```findClosestValue(BST,value)``` given an integer ```value``` and a binary search tree BST, the method should return ```value``` if present in the BST,  or a list of all the values that are closest to ```value``` (higher or lower);\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Test the code with the following BST:\n",
    "\n",
    "![](img/testbsttree.png)\n",
    "\n",
    "\n",
    "that can be implemented with:\n",
    "\n",
    "```\n",
    "intL = [9,25, 32, 31, 144,156, 44, 40, -1, 72, 11, 4, -7, 3, 1, 6]\n",
    "    \n",
    "B = createBST(intL)\n",
    "r = findExtremes(B)\n",
    "print(\"Min value: {}\\nMax value: {}\".format(r[0],r[1]))\n",
    "\n",
    "print(\"Sum subtree rooted at 4: {}\".format(sumNodes(B,4)))\n",
    "print(\"Sum subtree rooted at -1: {}\".format(sumNodes(B,-1)))\n",
    "print(\"Sum subtree rooted at 25: {}\".format(sumNodes(B,25)))\n",
    "print(\"Sum subtree rooted at 144: {}\".format(sumNodes(B,144)))\n",
    "print(\"Sum subtree rooted at 71: {}\".format(sumNodes(B,71)))\n",
    "\n",
    "print(\"Element(s) closest to 25: {}\".format(findClosestValue(B,25)))\n",
    "print(\"Element(s) closest to 10: {}\".format(findClosestValue(B,10)))\n",
    "print(\"Element(s) closest to 5: {}\".format(findClosestValue(B,5)))\n",
    "print(\"Element(s) closest to 28: {}\".format(findClosestValue(B,28)))\n",
    "print(\"Element(s) closest to 36: {}\".format(findClosestValue(B,36)))\n",
    "```\n",
    "\n",
    "should return:\n",
    "\n",
    "```\n",
    "Min value: -7\n",
    "Max value: 156\n",
    "Sum subtree rooted at 4: 14\n",
    "Sum subtree rooted at -1: 6\n",
    "Sum subtree rooted at 25: 555\n",
    "Sum subtree rooted at 144: 456\n",
    "Sum subtree rooted at 71: None\n",
    "Element(s) closest to 25: 25\n",
    "Element(s) closest to 10: [9, 11]\n",
    "Element(s) closest to 5: [4, 6]\n",
    "Element(s) closest to 28: [25, 31]\n",
    "Element(s) closest to 36: [32, 40]\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "nbsphinx": "hidden"
   },
   "outputs": [],
   "source": [
    "#Chr01 120 A T 120\n",
    "import random\n",
    "\n",
    "bases = ['A', 'T', 'C', 'G']\n",
    "chrs = [\"Chr01\", \"Chr02\", \"Chr03\"]\n",
    "of = open(\"/tmp/snp_data.tsv\", \"w\")\n",
    "of.write(\"//Chr\\tPos\\tRef\\tAlt\\tReadDepth\\n\")\n",
    "for el in chrs:\n",
    "    SNPs = random.randint(3000,4000)\n",
    "    B = []\n",
    "    P = []\n",
    "    D = []\n",
    "    for i in range(SNPs):\n",
    "        b = random.randint(0,3)\n",
    "        b1 = random.randint(0,3)\n",
    "        pos = random.randint(1,120000)\n",
    "        d = random.randint(5,120)\n",
    "        \n",
    "        if pos not in P:    \n",
    "            B.append((bases[b],bases[b1]))\n",
    "            P.append(pos)\n",
    "            D.append(d)\n",
    "        \n",
    "        P.sort()\n",
    "    for ind in range(len(P)):\n",
    "        of.write(el+\"\\t\"+str(P[ind])+\"\\t\"+B[ind][0]+\"\\t\"+B[ind][1]+\"\\t\"+str(D[ind])+\"\\n\")\n",
    "\n",
    "of.close()\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "nbsphinx": "hidden"
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "NAME:\n",
    "SURNAME:\n",
    "Matricola: \n",
    "\n",
    "partB.py\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "class BinaryTree:\n",
    "    def __init__(self, value):\n",
    "        self.__data = value\n",
    "        self.__right = None\n",
    "        self.__left = None\n",
    "        self.__parent = None\n",
    "    \n",
    "        \n",
    "    def getValue(self):\n",
    "        return self.__data\n",
    "        \n",
    "    def getParent(self):\n",
    "        return self.__parent\n",
    "    \n",
    "    #needed because we are using private attributes\n",
    "    def setParent(self, tree):\n",
    "        self.__parent = tree\n",
    "    \n",
    "    def getRight(self):\n",
    "        return self.__right\n",
    "    \n",
    "    def getLeft(self):\n",
    "        return self.__left\n",
    "    \n",
    "    def insertRight(self, tree):\n",
    "        if self.__right == None:\n",
    "            self.__right = tree\n",
    "            tree.setParent(self)\n",
    "    \n",
    "                    \n",
    "    def insertLeft(self, tree):\n",
    "        if self.__left == None:\n",
    "            self.__left = tree\n",
    "            tree.setParent(self)\n",
    "            \n",
    "    def deleteRight(self):\n",
    "        self.__right = None\n",
    "        \n",
    "    def deleteLeft(self):\n",
    "        self.__left = None\n",
    "        \n",
    "\n",
    "def createBST(intList):\n",
    "    BST = None\n",
    "    if len(intList) > 0:\n",
    "        BST = BinaryTree(intList[0])\n",
    "        for el in intList[1:]:\n",
    "            cur_el = BST\n",
    "            alreadyPresent = False\n",
    "            prev_el = None\n",
    "            while cur_el != None:\n",
    "                prev_el = cur_el\n",
    "                cv = cur_el.getValue()\n",
    "                if  cv > el:\n",
    "                    cur_el = cur_el.getLeft()\n",
    "                elif cv < el:\n",
    "                    cur_el = cur_el.getRight()\n",
    "                else:\n",
    "                    #cv == el (el is already present)\n",
    "                    #not allowed by rule c. skip it\n",
    "                    alreadyPresent = True\n",
    "                    #print(\"El {} already present\".format(el))\n",
    "                    break\n",
    "                \n",
    "                \n",
    "            if not alreadyPresent:\n",
    "                node = BinaryTree(el)\n",
    "                node.setParent(prev_el)\n",
    "                if prev_el.getValue() > el:\n",
    "                    prev_el.insertLeft(node)\n",
    "                else:\n",
    "                    prev_el.insertRight(node)\n",
    "                \n",
    "        \n",
    "    return BST\n",
    "\n",
    "def findExtremes(BST):\n",
    "    \"\"\"\n",
    "    returns a tuple with two elements that are the smallest and the \n",
    "    biggest element in the BST\n",
    "    \"\"\"\n",
    "\n",
    "\n",
    "def sumNodes(BST, N):\n",
    "    \"\"\"\n",
    "    given a binary search tree BST and a node N, \n",
    "    returns the sum of all the elements that are in \n",
    "    the subtree rooted at N or None if N is not in BST\n",
    "    \"\"\"        \n",
    "    \n",
    "def findClosestValue(BST,value):\n",
    "    \"\"\"\n",
    "    given an integer value the method should return \n",
    "    value if present in the BST,  or a list of all the \n",
    "    values that are closest to value (higher or lower)\n",
    "    \"\"\"\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    import pygraphviz as pgv\n",
    "    intL = [9,25, 32, 31, 144,156, 44, 40, -1, 72, 11, 4, -7, 3, 1, 6]\n",
    "    \n",
    "    B = createBST(intL)\n",
    "    r = findExtremes(B)\n",
    "    print(\"Min value: {}\\nMax value: {}\".format(r[0],r[1]))\n",
    "    plotGraph(B)\n",
    "    print(\"Sum subtree rooted at 4: {}\".format(sumNodes(B,4)))\n",
    "    print(\"Sum subtree rooted at -1: {}\".format(sumNodes(B,-1)))\n",
    "    print(\"Sum subtree rooted at 25: {}\".format(sumNodes(B,25)))\n",
    "    print(\"Sum subtree rooted at 144: {}\".format(sumNodes(B,144)))\n",
    "    print(\"Sum subtree rooted at 71: {}\".format(sumNodes(B,71)))\n",
    "    \n",
    "    print(\"Element(s) closest to 25: {}\".format(findClosestValue(B,25)))\n",
    "    print(\"Element(s) closest to 10: {}\".format(findClosestValue(B,10)))\n",
    "    print(\"Element(s) closest to 5: {}\".format(findClosestValue(B,5)))\n",
    "    print(\"Element(s) closest to 28: {}\".format(findClosestValue(B,28)))\n",
    "    print(\"Element(s) closest to 36: {}\".format(findClosestValue(B,36)))           "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### A possible solution: \n",
    "\n",
    "## Theory\n",
    "\n",
    "Exercise 1. Solution:\n",
    "$O(n^2)$\n",
    "\n",
    "## Practical part"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "nbsphinx": "hidden"
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "NAME:\n",
    "SURNAME:\n",
    "Matricola: \n",
    "\n",
    "partA.py\n",
    "\"\"\"\n",
    "import pandas as pd\n",
    "from matplotlib import pyplot as plt\n",
    "\n",
    "def computestats(data):\n",
    "    \"\"\"\n",
    "    gets the data structure loaded and prints the total \n",
    "    number of chromosomes, the total number of SNPs and \n",
    "    the total number of SNPs for each possible type \n",
    "    (i.e. TA, TC, TG, TT, AT,AC, AG, AA,...) and \n",
    "    the average read depth (with 3 decimal places of precision)\n",
    "    \"\"\"\n",
    "    print(\"The file contains {} SNPs\".format(data.shape[0]))\n",
    "    \n",
    "    countsByC = data['Chr'].value_counts()\n",
    "    print(\"Belonging to {} chromosomes:\".format(len(countsByC)))\n",
    "    chrs = []\n",
    "    for i in countsByC.index:\n",
    "        chrs.append(i)\n",
    "    chrs.sort()\n",
    "    print(\"\\n\".join(chrs))\n",
    "    \n",
    "    data['type'] = data['Ref']+data['Alt']\n",
    "    countsByType = data['type'].value_counts()\n",
    "    for el in countsByType.index:\n",
    "        print(\"Type {} counts {} SNPs\".format(el, countsByType[el]))\n",
    "        \n",
    "    print(\"Average read depth: {:.3f} reads per site\".format(data[\"ReadDepth\"].mean()))\n",
    "    \n",
    "    \n",
    "def chrHistogram(data,chromosome, start, end):\n",
    "    \"\"\"\n",
    "    given a chromosome, a start postion and an end position,  \n",
    "    plots an histogram with the number of SNPs for each type\n",
    "    belonging to that chromosome having postion in \n",
    "    the interval [start, end]\n",
    "    \"\"\"\n",
    "    chrFilt = data[data['Chr'] == chromosome]\n",
    "    tmp = chrFilt[chrFilt['Pos'] >= start]\n",
    "    filtSNPs = tmp[tmp['Pos'] <= end]\n",
    "    countsByType = filtSNPs['type'].value_counts()\n",
    "    countsByType.plot(kind = 'bar')\n",
    "    plt.title('# SNPs of {} in [{},{}]'.format(chromosome, start,end))\n",
    "    plt.show()\n",
    "    pass\n",
    "    \n",
    "\n",
    "\n",
    "def readDepthStats(data, chromosome, start, end):\n",
    "    \"\"\"\n",
    "    given the input data and chromosome, a start and end position, \n",
    "    the function computes  the average read depth of all the SNPs \n",
    "    belonging to that chromosome (and in the [start,end] interval) \n",
    "    and the average read depth for each \n",
    "    type of SNP (TA, TC, ...).\n",
    "    \"\"\"\n",
    "    chrFilt = data[data['Chr'] == chromosome]\n",
    "    tmp = chrFilt[chrFilt['Pos'] >= start]\n",
    "    filtSNPs = tmp[tmp['Pos'] <= end]\n",
    "    print(\"Global Avg. read depth for SNPs in {} [{},{}]: {:.3f}\".format(chromosome,\n",
    "                                                             start,\n",
    "                                                             end,\n",
    "                                                             filtSNPs['ReadDepth'].mean()))\n",
    "    \n",
    "    byType = filtSNPs.groupby('type')['ReadDepth'].mean()\n",
    "    for el in byType.index:\n",
    "        print(\"\\tType: {} avg. read depth: {:.3f}\".format(el,byType[el]))\n",
    "    pass\n",
    "\n",
    "dataFile = \"file_samples/snp_data.tsv\"\n",
    "\n",
    "data = pd.read_csv(dataFile,\n",
    "                     sep=\"\\t\",  header = 0)\n",
    "    \n",
    "\n",
    "computestats(data)\n",
    "print(\"\\n\")\n",
    "chrHistogram(data, \"Chr01\",500,4000)\n",
    "chrHistogram(data, \"Chr02\",1,1250)\n",
    "chrHistogram(data, \"Chr03\",1,50000)\n",
    "readDepthStats(data, \"Chr01\", 500,6000)\n",
    "print(\"\\n\")\n",
    "readDepthStats(data, \"Chr02\",1, 2000)\n",
    "\n",
    "    \n",
    "    \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\n",
      "ERROR:root:Invalid alias: The name more can't be aliased because it is another magic command.\n",
      "ERROR:root:Invalid alias: The name less can't be aliased because it is another magic command.\n",
      "ERROR:root:Invalid alias: The name man can't be aliased because it is another magic command.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Min value: -7\n",
      "Max value: 156\n",
      "Sum subtree rooted at 4: 14\n",
      "Sum subtree rooted at -1: 6\n",
      "Sum subtree rooted at 25: 555\n",
      "Sum subtree rooted at 144: 456\n",
      "Sum subtree rooted at 71: None\n",
      "Element(s) closest to 25: 25\n",
      "Element(s) closest to 10: [9, 11]\n",
      "Element(s) closest to 5: [4, 6]\n",
      "Element(s) closest to 28: [25, 31]\n",
      "Element(s) closest to 36: [32, 40]\n"
     ]
    }
   ],
   "source": [
    "%reset -f \n",
    "\"\"\"\n",
    "NAME:\n",
    "SURNAME:\n",
    "Matricola: \n",
    "\n",
    "partB.py\n",
    "\"\"\"\n",
    "\n",
    "from collections import deque\n",
    "\n",
    "class BinaryTree:\n",
    "    def __init__(self, value):\n",
    "        self.__data = value\n",
    "        self.__right = None\n",
    "        self.__left = None\n",
    "        self.__parent = None\n",
    "    \n",
    "        \n",
    "    def getValue(self):\n",
    "        return self.__data\n",
    "        \n",
    "    def getParent(self):\n",
    "        return self.__parent\n",
    "    \n",
    "    #needed because we are using private attributes\n",
    "    def setParent(self, tree):\n",
    "        self.__parent = tree\n",
    "    \n",
    "    def getRight(self):\n",
    "        return self.__right\n",
    "    \n",
    "    def getLeft(self):\n",
    "        return self.__left\n",
    "    \n",
    "    def insertRight(self, tree):\n",
    "        if self.__right == None:\n",
    "            self.__right = tree\n",
    "            tree.setParent(self)\n",
    "    \n",
    "                    \n",
    "    def insertLeft(self, tree):\n",
    "        if self.__left == None:\n",
    "            self.__left = tree\n",
    "            tree.setParent(self)\n",
    "            \n",
    "    def deleteRight(self):\n",
    "        self.__right = None\n",
    "        \n",
    "    def deleteLeft(self):\n",
    "        self.__left = None\n",
    "        \n",
    "\n",
    "def createBST(intList):\n",
    "    BST = None\n",
    "    if len(intList) > 0:\n",
    "        BST = BinaryTree(intList[0])\n",
    "        for el in intList[1:]:\n",
    "            cur_el = BST\n",
    "            alreadyPresent = False\n",
    "            prev_el = None\n",
    "            while cur_el != None:\n",
    "                prev_el = cur_el\n",
    "                cv = cur_el.getValue()\n",
    "                if  cv > el:\n",
    "                    cur_el = cur_el.getLeft()\n",
    "                elif cv < el:\n",
    "                    cur_el = cur_el.getRight()\n",
    "                else:\n",
    "                    #cv == el (el is already present)\n",
    "                    #not allowed by rule c. skip it\n",
    "                    alreadyPresent = True\n",
    "                    #print(\"El {} already present\".format(el))\n",
    "                    break\n",
    "                \n",
    "                \n",
    "            if not alreadyPresent:\n",
    "                node = BinaryTree(el)\n",
    "                node.setParent(prev_el)\n",
    "                if prev_el.getValue() > el:\n",
    "                    prev_el.insertLeft(node)\n",
    "                else:\n",
    "                    prev_el.insertRight(node)\n",
    "                \n",
    "        \n",
    "    return BST\n",
    "\n",
    "def findExtremes(BST):\n",
    "    \"\"\"\n",
    "    returns a tuple with two elements that are the smaller and the \n",
    "    bigger element in the BST\n",
    "    \"\"\"\n",
    "    root = BST\n",
    "    minVal = None\n",
    "    maxVal = None\n",
    "    if root == None:\n",
    "        return (minVal, maxVal)\n",
    "    \n",
    "    #Find the smallest element\n",
    "    curNode = root\n",
    "    while curNode.getLeft() != None:\n",
    "        curNode = curNode.getLeft()\n",
    "        \n",
    "    minVal = curNode.getValue()\n",
    "    \n",
    "    #Find the biggest element\n",
    "    curNode = root\n",
    "    while curNode.getRight() != None:\n",
    "        curNode = curNode.getRight()\n",
    "    maxVal = curNode.getValue()\n",
    "    \n",
    "    return (minVal, maxVal)\n",
    "\n",
    "\n",
    "def sumNodes(BST, N):\n",
    "    \"\"\"\n",
    "    given a binary search tree BST and a node N, \n",
    "    returns the sum of all the elements that are in \n",
    "    the subtree rooted at N or None if N is not in BST\n",
    "    \"\"\"\n",
    "    #find node N\n",
    "    curNode = BST\n",
    "    while curNode != None and curNode.getValue() != N:\n",
    "        if N > curNode.getValue():\n",
    "            curNode = curNode.getRight()\n",
    "        else:\n",
    "            curNode = curNode.getLeft()\n",
    "            \n",
    "    if curNode == None:\n",
    "        return None\n",
    "    else:\n",
    "        res = curNode.getValue()\n",
    "        subtree = deque()\n",
    "        subtree.append(curNode.getRight())\n",
    "        subtree.append(curNode.getLeft())\n",
    "        while len(subtree) > 0:\n",
    "            curNode = subtree.popleft()\n",
    "            if curNode != None:\n",
    "                res += curNode.getValue()\n",
    "                subtree.append(curNode.getRight())\n",
    "                subtree.append(curNode.getLeft())\n",
    "        return res\n",
    "        \n",
    "    \n",
    "def findClosestValue(BST,value):    \n",
    "    \"\"\"\n",
    "    given an integer value the method should return \n",
    "    value if present in the BST,  or a list of all the \n",
    "    values that are closest to value (higher or lower)\n",
    "    \"\"\"\n",
    "    #find node value\n",
    "    dist = abs(BST.getValue() - value)\n",
    "    #this will be filled in the first loop of while\n",
    "    closestEls = []\n",
    "    curNode = BST\n",
    "    \n",
    "    while curNode != None:\n",
    "        curVal = curNode.getValue()\n",
    "        if curVal == value:\n",
    "            return value\n",
    "        tmpD = abs(curVal - value)\n",
    "        if tmpD < dist:\n",
    "            closestEls = [curVal]\n",
    "            dist = tmpD\n",
    "        else:\n",
    "            if tmpD == dist:\n",
    "                closestEls.append(curVal)\n",
    "        #move to next\n",
    "        if value > curVal:\n",
    "            curNode = curNode.getRight()\n",
    "        else:\n",
    "            curNode = curNode.getLeft()\n",
    "    \n",
    "    \n",
    "    return closestEls\n",
    "    \n",
    "    \n",
    "def plotGraph(tree):    \n",
    "    #uses pygraphviz \n",
    "    G=pgv.AGraph(directed=True)\n",
    "    \n",
    "    levels = [tree]\n",
    "    while len(levels) > 0:\n",
    "        cur_n = levels.pop()\n",
    "        if cur_n != None:\n",
    "            G.add_node(cur_n.getValue(), color = 'black')\n",
    "            r = cur_n.getRight()\n",
    "            l = cur_n.getLeft()\n",
    "            if l != None:\n",
    "                G.add_node(l.getValue(), color = 'black')\n",
    "                G.add_edge(cur_n.getValue(), l.getValue())\n",
    "                levels.append(l)\n",
    "            if r != None:\n",
    "                G.add_node(r.getValue(), color = 'black')\n",
    "                G.add_edge(cur_n.getValue(), r.getValue())\n",
    "                levels.append(r)\n",
    "    G.layout(prog='dot') # use dot\n",
    "    #In the following, the path of the graph. Change\n",
    "    #it to your liking\n",
    "    G.draw('img/testbsttree.png')\n",
    "    \n",
    "    \n",
    "if __name__ == \"__main__\":\n",
    "    import pygraphviz as pgv\n",
    "    intL = [9,25, 32, 31, 144,156, 44, 40, -1, 72, 11, 4, -7, 3, 1, 6]\n",
    "    \n",
    "    B = createBST(intL)\n",
    "    r = findExtremes(B)\n",
    "    print(\"Min value: {}\\nMax value: {}\".format(r[0],r[1]))\n",
    "    plotGraph(B)\n",
    "    print(\"Sum subtree rooted at 4: {}\".format(sumNodes(B,4)))\n",
    "    print(\"Sum subtree rooted at -1: {}\".format(sumNodes(B,-1)))\n",
    "    print(\"Sum subtree rooted at 25: {}\".format(sumNodes(B,25)))\n",
    "    print(\"Sum subtree rooted at 144: {}\".format(sumNodes(B,144)))\n",
    "    print(\"Sum subtree rooted at 71: {}\".format(sumNodes(B,71)))\n",
    "    \n",
    "    print(\"Element(s) closest to 25: {}\".format(findClosestValue(B,25)))\n",
    "    print(\"Element(s) closest to 10: {}\".format(findClosestValue(B,10)))\n",
    "    print(\"Element(s) closest to 5: {}\".format(findClosestValue(B,5)))\n",
    "    print(\"Element(s) closest to 28: {}\".format(findClosestValue(B,28)))\n",
    "    print(\"Element(s) closest to 36: {}\".format(findClosestValue(B,36)))\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "nbsphinx": "hidden"
   },
   "source": [
    "![](img/testbsttree.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "nbsphinx": "hidden"
   },
   "outputs": [],
   "source": [
    "\"CODE NOT SHOWN\"\n",
    "\n",
    "#Drawing a graph in pygraphviz\n",
    "import pygraphviz as pgv\n",
    "\n",
    "G=pgv.AGraph(directed=True)\n",
    "\n",
    "#Attributes can be added when adding nodes or edge\n",
    "G.add_node(\"9\", color='black')\n",
    "G.add_node(\"25\", color='black')\n",
    "G.add_node(\"32\", color='black')\n",
    "G.add_node(\"144\", color='black')\n",
    "G.add_node(\"44\", color='black')\n",
    "G.add_node(\"-1\", color='black')\n",
    "G.add_node(\"72\", color='black')\n",
    "G.add_node(\"11\", color='black')\n",
    "G.add_node(\"4\", color='black')\n",
    "G.add_node(\"-7\", color='black')\n",
    "G.add_node(\"3\", color='black')\n",
    "G.add_node(\"1\", color='black')\n",
    "G.add_node(\"7\", color='black')\n",
    "\n",
    "\n",
    "\n",
    "G.add_edge(\"9\" ,\"-1\", color='blue')\n",
    "G.add_edge(\"9\", \"25\", color='blue')\n",
    "G.add_edge(\"-1\" ,\"-7\", color='blue')\n",
    "G.add_edge(\"-1\" ,\"4\", color='blue')\n",
    "G.add_edge(\"25\" ,\"11\", color='blue')\n",
    "G.add_edge(\"25\" ,\"32\", color='blue')\n",
    "G.add_edge(\"4\" ,\"3\", color='blue')\n",
    "G.add_edge(\"4\" ,\"7\", color='blue')\n",
    "G.add_edge(\"3\" ,\"1\", color='blue')\n",
    "G.add_edge(\"32\" ,\"144\", color='blue')\n",
    "G.add_edge(\"32\" ,\"72\", color='blue')\n",
    "# write to a dot file\n",
    "#G.write('test.dot')\n",
    "\n",
    "#create a png file\n",
    "G.layout(prog='dot') # use dot\n",
    "G.draw('img/testbsttree.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "celltoolbar": "Edit Metadata",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
