{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "nbsphinx": "hidden"
   },
   "source": [
    "### NOTE FOR LUCA\n",
    "\n",
    "**Remember to set/remove metadata as:**\n",
    "{\n",
    "  \"nbsphinx\": \"hidden\"\n",
    "}\n",
    "\n",
    "to enable/disable solutions view"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Midterm 2018\n",
    "\n",
    "## Before you start\n",
    "**Please write one single python script.**\n",
    "\n",
    "**IMPORTANT: Add your name and ID (matricola) on top of the .py file!**\n",
    "\n",
    "## The Problem\n",
    "\n",
    "The dataset we are going to use is a tab separated version of the Sequence Alignment Format (SAM), which contains, among the other things (that have been removed), the following fields:\n",
    "\n",
    "```\n",
    "readID\tChr\tPos\tSequence\tMismatchingPositions\n",
    "DJB77P1:536:H880AADXX:1:1116:16537:22664\tChr02\t41\tGGAACTCGAACA...TAATCAGGAACT\t26A124\n",
    "DJB77P1:536:H880AADXX:1:1111:5187:26828\tChr02\t61\tTCACCCAAACTCAA...ACATGATGC\t24G3T40G81\n",
    "DJB77P1:536:H880AADXX:1:1111:20900:20916\tChr02\t9\tTGAACTCGAG...AACTTCAA\t151\n",
    "```\n",
    "\n",
    "as the header (first line) describes, the first column is the identifier of the read, the second and third are the chromosome and position (starting position) to which the read is aligned, the fourth is the sequence and the fifth is a string that reports any mismatches (it is normally more complex than this but we can skip insertions/deletions). \n",
    "\n",
    "For example, the mismatchingPositions string:\n",
    "```26A124``` means that the read aligned with 26 matches, one mismatch (A in the reference) and 124 more matches (for a total of 151 bases that is the length of the read);\n",
    "```24G3T40G81``` means 24 matches, one G (in the reference), 3 matches, T, 40 matches, G, 81 matches (total 151 bases)\n",
    "```151``` means 151 matches and no mismatch.\n",
    "\n",
    "The alignment:\n",
    "```\n",
    "reference: ATACACG\n",
    "read:      ATCCACG\n",
    "```\n",
    "would be represented as ```2A4```\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Implement the following python functions:\n",
    "\n",
    "1. ```countMismatches(mis_str)```: gets a mismatch string and returns the number of mismatches. Please note that reads might contain \"N\" as bases. Hint: count characters that represent nucleotides.\n",
    "\n",
    "Calling:\n",
    "```\n",
    "mismatches = [\"9C15T125\", \n",
    "              \"9G4C136\", \n",
    "              \"151\", \n",
    "              \"9G7T2C24A24C11A0G5G6G1A0G51\", \n",
    "              \"9T13A16T1G5A6A6A6A81\"]\n",
    "\n",
    "for m in mismatches:\n",
    "    print(\"{} has {} mismatches\".format(m, countMismatches(m)))\n",
    "```\n",
    "Should give:\n",
    "```\n",
    "9C15T125 has 2 mismatches\n",
    "9G4C136 has 2 mismatches\n",
    "151 has 0 mismatches\n",
    "9G7T2C24A24C11A0G5G6G1A0G51 has 11 mismatches\n",
    "9T13A16T1G5A6A6A6A81 has 8 mismatches\n",
    "```\n",
    "\n",
    "\n",
    "2. ```getReadsInInterval(filename, chr, startPos, endPos, max_diff)```: reads all entries in the file ```filename``` and returns all the reads that aligned to the chromosome ```chr``` and have at least one base in the closed interval ```[startPos,endPos]``` with no more than ```max_diff``` mismatches. The function should return a dictionary with readIDs as keys and all the other elements (in a list) as values. The function should also print some information on the reads selected, like their number, readID and starting position (see below).\n",
    "\n",
    "given the entry:\n",
    "```\n",
    "DJB77P1:536:H880AADXX:1:1116:16537:22664\tChr02\t41\tGGAACTCGAACA...TAATCAGGAACT\t26A124\n",
    "```\n",
    "the returned dictionary should be:\n",
    "```\n",
    "{\"DJB77P1:536:H880AADXX:1:1116:16537:22664\" : ['Chr02',41,'GGAACTCGAACA...TAATCAGGAACT', '26A124']}\n",
    "```\n",
    "\n",
    "Calling:\n",
    "```\n",
    "fn = \"./test_data.tsv\"\n",
    "chromosome = \"Chr01\"\n",
    "reads = getReadsInInterval(fn, chromosome, 100,600, 2)\n",
    "```\n",
    "should give:\n",
    "```\n",
    "4 reads cover the interval Chr01: 100-600\n",
    "They are:\n",
    "\tDJB77P1:536:H880AADXX:1:1204:4499:50848 in Chr01 starts at 174\n",
    "\tDJB77P1:536:H880AADXX:1:1111:13063:11094 in Chr01 starts at 270\n",
    "\tDJB77P1:536:H880AADXX:1:1101:3693:67653 in Chr01 starts at 183\n",
    "\tDJB77P1:536:H880AADXX:1:1106:18413:36334 in Chr01 starts at 250\n",
    "```\n",
    "\n",
    "3.  ```getCoverage(reads, chr,startP, endP)```: based on the ```reads``` selected with ```getReadsInInterval``` this function computes the coverage for each position of chromosome ```chr``` starting from ```startP``` to ```endP``` both included (i.e. number of reads covering each position of the interval). For each position, the \n",
    "\n",
    "Calling:\n",
    "```\n",
    "getCoverage(reads, chromosome, 172,400)\n",
    "```\n",
    "should print:\n",
    "```\n",
    "Chr01 172 has coverage 0\n",
    "Chr01 173 has coverage 0\n",
    "Chr01 174 has coverage 1\n",
    "Chr01 175 has coverage 1\n",
    "Chr01 176 has coverage 1\n",
    "Chr01 177 has coverage 1\n",
    "...\n",
    "Chr01 382 has coverage 2\n",
    "Chr01 383 has coverage 2\n",
    "Chr01 384 has coverage 2\n",
    "...\n",
    "Chr01 396 has coverage 2\n",
    "Chr01 397 has coverage 2\n",
    "Chr01 398 has coverage 2\n",
    "Chr01 399 has coverage 2\n",
    "Chr01 400 has coverage 2\n",
    "```\n",
    "\n",
    "**Note: \"...\" are just missing lines, but the output should go from startP (172) to endP (400) included.**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download the data\n",
    "\n",
    "Create a \"qcbsciprolab-2018-11-13-NAME-SURNAME-ID\" folder on the desktop. Download the following data to the folder that you just created.\n",
    "\n",
    "The .tsv alignment file is: [test_data_unique10k.tsv](2018mt/test_data_unique10k.tsv)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### A possible solution: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9C15T125 has 2 mismatches\n",
      "9G4C136 has 2 mismatches\n",
      "151 has 0 mismatches\n",
      "9G7T2C24A24C11A0G5G6G1A0G51 has 11 mismatches\n",
      "9T13A16T1G5A6A6A6A81 has 8 mismatches\n",
      "\n",
      "\n",
      "4 reads cover the interval Chr01: 100-600\n",
      "They are:\n",
      "\tDJB77P1:536:H880AADXX:1:1111:13063:11094 in Chr01 starts at 270\n",
      "\tDJB77P1:536:H880AADXX:1:1204:4499:50848 in Chr01 starts at 174\n",
      "\tDJB77P1:536:H880AADXX:1:1101:3693:67653 in Chr01 starts at 183\n",
      "\tDJB77P1:536:H880AADXX:1:1106:18413:36334 in Chr01 starts at 250\n",
      "\n",
      "\n",
      "Chr01 172 has coverage 0\n",
      "Chr01 173 has coverage 0\n",
      "Chr01 174 has coverage 1\n",
      "Chr01 175 has coverage 1\n",
      "Chr01 176 has coverage 1\n",
      "Chr01 177 has coverage 1\n",
      "Chr01 178 has coverage 1\n",
      "Chr01 179 has coverage 1\n",
      "Chr01 180 has coverage 1\n",
      "Chr01 181 has coverage 1\n",
      "Chr01 182 has coverage 1\n",
      "Chr01 183 has coverage 2\n",
      "Chr01 184 has coverage 2\n",
      "Chr01 185 has coverage 2\n",
      "Chr01 186 has coverage 2\n",
      "Chr01 187 has coverage 2\n",
      "Chr01 188 has coverage 2\n",
      "Chr01 189 has coverage 2\n",
      "Chr01 190 has coverage 2\n",
      "Chr01 191 has coverage 2\n",
      "Chr01 192 has coverage 2\n",
      "Chr01 193 has coverage 2\n",
      "Chr01 194 has coverage 2\n",
      "Chr01 195 has coverage 2\n",
      "Chr01 196 has coverage 2\n",
      "Chr01 197 has coverage 2\n",
      "Chr01 198 has coverage 2\n",
      "Chr01 199 has coverage 2\n",
      "Chr01 200 has coverage 2\n",
      "Chr01 201 has coverage 2\n",
      "Chr01 202 has coverage 2\n",
      "Chr01 203 has coverage 2\n",
      "Chr01 204 has coverage 2\n",
      "Chr01 205 has coverage 2\n",
      "Chr01 206 has coverage 2\n",
      "Chr01 207 has coverage 2\n",
      "Chr01 208 has coverage 2\n",
      "Chr01 209 has coverage 2\n",
      "Chr01 210 has coverage 2\n",
      "Chr01 211 has coverage 2\n",
      "Chr01 212 has coverage 2\n",
      "Chr01 213 has coverage 2\n",
      "Chr01 214 has coverage 2\n",
      "Chr01 215 has coverage 2\n",
      "Chr01 216 has coverage 2\n",
      "Chr01 217 has coverage 2\n",
      "Chr01 218 has coverage 2\n",
      "Chr01 219 has coverage 2\n",
      "Chr01 220 has coverage 2\n",
      "Chr01 221 has coverage 2\n",
      "Chr01 222 has coverage 2\n",
      "Chr01 223 has coverage 2\n",
      "Chr01 224 has coverage 2\n",
      "Chr01 225 has coverage 2\n",
      "Chr01 226 has coverage 2\n",
      "Chr01 227 has coverage 2\n",
      "Chr01 228 has coverage 2\n",
      "Chr01 229 has coverage 2\n",
      "Chr01 230 has coverage 2\n",
      "Chr01 231 has coverage 2\n",
      "Chr01 232 has coverage 2\n",
      "Chr01 233 has coverage 2\n",
      "Chr01 234 has coverage 2\n",
      "Chr01 235 has coverage 2\n",
      "Chr01 236 has coverage 2\n",
      "Chr01 237 has coverage 2\n",
      "Chr01 238 has coverage 2\n",
      "Chr01 239 has coverage 2\n",
      "Chr01 240 has coverage 2\n",
      "Chr01 241 has coverage 2\n",
      "Chr01 242 has coverage 2\n",
      "Chr01 243 has coverage 2\n",
      "Chr01 244 has coverage 2\n",
      "Chr01 245 has coverage 2\n",
      "Chr01 246 has coverage 2\n",
      "Chr01 247 has coverage 2\n",
      "Chr01 248 has coverage 2\n",
      "Chr01 249 has coverage 2\n",
      "Chr01 250 has coverage 3\n",
      "Chr01 251 has coverage 3\n",
      "Chr01 252 has coverage 3\n",
      "Chr01 253 has coverage 3\n",
      "Chr01 254 has coverage 3\n",
      "Chr01 255 has coverage 3\n",
      "Chr01 256 has coverage 3\n",
      "Chr01 257 has coverage 3\n",
      "Chr01 258 has coverage 3\n",
      "Chr01 259 has coverage 3\n",
      "Chr01 260 has coverage 3\n",
      "Chr01 261 has coverage 3\n",
      "Chr01 262 has coverage 3\n",
      "Chr01 263 has coverage 3\n",
      "Chr01 264 has coverage 3\n",
      "Chr01 265 has coverage 3\n",
      "Chr01 266 has coverage 3\n",
      "Chr01 267 has coverage 3\n",
      "Chr01 268 has coverage 3\n",
      "Chr01 269 has coverage 3\n",
      "Chr01 270 has coverage 4\n",
      "Chr01 271 has coverage 4\n",
      "Chr01 272 has coverage 4\n",
      "Chr01 273 has coverage 4\n",
      "Chr01 274 has coverage 4\n",
      "Chr01 275 has coverage 4\n",
      "Chr01 276 has coverage 4\n",
      "Chr01 277 has coverage 4\n",
      "Chr01 278 has coverage 4\n",
      "Chr01 279 has coverage 4\n",
      "Chr01 280 has coverage 4\n",
      "Chr01 281 has coverage 4\n",
      "Chr01 282 has coverage 4\n",
      "Chr01 283 has coverage 4\n",
      "Chr01 284 has coverage 4\n",
      "Chr01 285 has coverage 4\n",
      "Chr01 286 has coverage 4\n",
      "Chr01 287 has coverage 4\n",
      "Chr01 288 has coverage 4\n",
      "Chr01 289 has coverage 4\n",
      "Chr01 290 has coverage 4\n",
      "Chr01 291 has coverage 4\n",
      "Chr01 292 has coverage 4\n",
      "Chr01 293 has coverage 4\n",
      "Chr01 294 has coverage 4\n",
      "Chr01 295 has coverage 4\n",
      "Chr01 296 has coverage 4\n",
      "Chr01 297 has coverage 4\n",
      "Chr01 298 has coverage 4\n",
      "Chr01 299 has coverage 4\n",
      "Chr01 300 has coverage 4\n",
      "Chr01 301 has coverage 4\n",
      "Chr01 302 has coverage 4\n",
      "Chr01 303 has coverage 4\n",
      "Chr01 304 has coverage 4\n",
      "Chr01 305 has coverage 4\n",
      "Chr01 306 has coverage 4\n",
      "Chr01 307 has coverage 4\n",
      "Chr01 308 has coverage 4\n",
      "Chr01 309 has coverage 4\n",
      "Chr01 310 has coverage 4\n",
      "Chr01 311 has coverage 4\n",
      "Chr01 312 has coverage 4\n",
      "Chr01 313 has coverage 4\n",
      "Chr01 314 has coverage 4\n",
      "Chr01 315 has coverage 4\n",
      "Chr01 316 has coverage 4\n",
      "Chr01 317 has coverage 4\n",
      "Chr01 318 has coverage 4\n",
      "Chr01 319 has coverage 4\n",
      "Chr01 320 has coverage 4\n",
      "Chr01 321 has coverage 4\n",
      "Chr01 322 has coverage 4\n",
      "Chr01 323 has coverage 4\n",
      "Chr01 324 has coverage 4\n",
      "Chr01 325 has coverage 3\n",
      "Chr01 326 has coverage 3\n",
      "Chr01 327 has coverage 3\n",
      "Chr01 328 has coverage 3\n",
      "Chr01 329 has coverage 3\n",
      "Chr01 330 has coverage 3\n",
      "Chr01 331 has coverage 3\n",
      "Chr01 332 has coverage 3\n",
      "Chr01 333 has coverage 3\n",
      "Chr01 334 has coverage 2\n",
      "Chr01 335 has coverage 2\n",
      "Chr01 336 has coverage 2\n",
      "Chr01 337 has coverage 2\n",
      "Chr01 338 has coverage 2\n",
      "Chr01 339 has coverage 2\n",
      "Chr01 340 has coverage 2\n",
      "Chr01 341 has coverage 2\n",
      "Chr01 342 has coverage 2\n",
      "Chr01 343 has coverage 2\n",
      "Chr01 344 has coverage 2\n",
      "Chr01 345 has coverage 2\n",
      "Chr01 346 has coverage 2\n",
      "Chr01 347 has coverage 2\n",
      "Chr01 348 has coverage 2\n",
      "Chr01 349 has coverage 2\n",
      "Chr01 350 has coverage 2\n",
      "Chr01 351 has coverage 2\n",
      "Chr01 352 has coverage 2\n",
      "Chr01 353 has coverage 2\n",
      "Chr01 354 has coverage 2\n",
      "Chr01 355 has coverage 2\n",
      "Chr01 356 has coverage 2\n",
      "Chr01 357 has coverage 2\n",
      "Chr01 358 has coverage 2\n",
      "Chr01 359 has coverage 2\n",
      "Chr01 360 has coverage 2\n",
      "Chr01 361 has coverage 2\n",
      "Chr01 362 has coverage 2\n",
      "Chr01 363 has coverage 2\n",
      "Chr01 364 has coverage 2\n",
      "Chr01 365 has coverage 2\n",
      "Chr01 366 has coverage 2\n",
      "Chr01 367 has coverage 2\n",
      "Chr01 368 has coverage 2\n",
      "Chr01 369 has coverage 2\n",
      "Chr01 370 has coverage 2\n",
      "Chr01 371 has coverage 2\n",
      "Chr01 372 has coverage 2\n",
      "Chr01 373 has coverage 2\n",
      "Chr01 374 has coverage 2\n",
      "Chr01 375 has coverage 2\n",
      "Chr01 376 has coverage 2\n",
      "Chr01 377 has coverage 2\n",
      "Chr01 378 has coverage 2\n",
      "Chr01 379 has coverage 2\n",
      "Chr01 380 has coverage 2\n",
      "Chr01 381 has coverage 2\n",
      "Chr01 382 has coverage 2\n",
      "Chr01 383 has coverage 2\n",
      "Chr01 384 has coverage 2\n",
      "Chr01 385 has coverage 2\n",
      "Chr01 386 has coverage 2\n",
      "Chr01 387 has coverage 2\n",
      "Chr01 388 has coverage 2\n",
      "Chr01 389 has coverage 2\n",
      "Chr01 390 has coverage 2\n",
      "Chr01 391 has coverage 2\n",
      "Chr01 392 has coverage 2\n",
      "Chr01 393 has coverage 2\n",
      "Chr01 394 has coverage 2\n",
      "Chr01 395 has coverage 2\n",
      "Chr01 396 has coverage 2\n",
      "Chr01 397 has coverage 2\n",
      "Chr01 398 has coverage 2\n",
      "Chr01 399 has coverage 2\n",
      "Chr01 400 has coverage 2\n"
     ]
    }
   ],
   "source": [
    "%reset -f\n",
    "\n",
    "\n",
    "#gets the MismatchingPositions string and counts how many mismatches are there.\n",
    "def countMismatches(mis_str):\n",
    "    t_str = mis_str.upper()\n",
    "    res = [t_str.count(x) for x in \"ATCGN\"]\n",
    "    return sum(res)\n",
    "\n",
    "def getReadsInInterval(filename, chr, startPos, endPos, max_diff = -1):\n",
    "    reads = {}\n",
    "    inF = open(filename, \"r\")\n",
    "\n",
    "    for line in inF:\n",
    "        line = line.strip()\n",
    "        \n",
    "        if(not line.startswith(\"readID\")):\n",
    "            els = line.split('\\t')\n",
    "   \n",
    "            L = len(els[3])\n",
    "            readStart = int(els[2])\n",
    "            if(els[1] == chr):\n",
    "                for pos in range(startPos, endPos+1):\n",
    "                    if(els[1] == chr and pos >= readStart and pos < readStart + L): #note: not = !!! \n",
    "                        if(max_diff == -1 or max_diff >= countMismatches(els[4])):\n",
    "                            if(els[0] not in reads):\n",
    "                                reads[els[0]] = els[1:]\n",
    "    print(\"{} reads cover the interval {}: {}-{}\\nThey are:\".format(len(reads), chromosome, 100,600 ))\n",
    "    for r in reads:\n",
    "        print(\"\\t{} in {} starts at {}\".format(r,reads[r][0],reads[r][1]))\n",
    "    \n",
    "\n",
    "    return reads\n",
    "\n",
    "\n",
    "\n",
    "            \n",
    "def getCoverage(reads, chr,startP, endP):\n",
    "\n",
    "    coverage = [0 for x in range(startP, endP+1)]\n",
    "    intervals = []\n",
    "    for r in reads:\n",
    "        read = reads[r]\n",
    "        if(read[0] == chr):\n",
    "            pos = int(read[1])\n",
    "            L = len(read[2])\n",
    "            intervals.append([pos, pos+L])\n",
    "    for i in range(startP, endP + 1):\n",
    "        for s,e in intervals:\n",
    "            if(i >= s and i < e):\n",
    "                coverage[i-startP] += 1\n",
    "    for ind in range(len(coverage)):\n",
    "        print(\"{} {} has coverage {}\".format(chr, ind + startP, coverage[ind]))\n",
    "        \n",
    "#    return [ (i + startP, coverage[i]) for i in range(len(coverage))]        \n",
    "    \n",
    "#2x average coverage\n",
    "def getRepeats(cov):\n",
    "    values = [x[1] for x in cov]\n",
    "    A = np.mean(values)\n",
    "    repeats = [x for x in cov if x[1] >= 2*A]\n",
    "    return [A, repeats]\n",
    "    \n",
    "\n",
    "mismatches = [\"9C15T125\", \n",
    "              \"9G4C136\", \n",
    "              \"151\", \n",
    "              \"9G7T2C24A24C11A0G5G6G1A0G51\", \n",
    "              \"9T13A16T1G5A6A6A6A81\"]\n",
    "\n",
    "for m in mismatches:\n",
    "    print(\"{} has {} mismatches\".format(m, countMismatches(m)))\n",
    "print(\"\\n\")\n",
    "fn = \"./2018mt/test_data_unique10k.tsv\"\n",
    "chromosome = \"Chr01\"\n",
    "reads = getReadsInInterval(fn, chromosome, 100,600, 2)\n",
    "#print(reads)\n",
    "print(\"\\n\")\n",
    "getCoverage(reads, chromosome, 172,400)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "celltoolbar": "Edit Metadata",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
