{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "nbsphinx": "hidden"
   },
   "source": [
    "### NOTE FOR LUCA\n",
    "\n",
    "**Remember to set/remove metadata as:**\n",
    "{\n",
    "  \"nbsphinx\": \"hidden\"\n",
    "}\n",
    "\n",
    "to enable/disable solutions view"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# Practical 14\n",
    "\n",
    "In this practical we will start working with sorting algorithms. In particular we will work with **selection sort** and **insertion sort**.\n",
    "\n",
    "## Slides\n",
    "\n",
    "The slides of the introduction can be found here: [Intro](docs/Practical14.pdf)\n",
    "\n",
    "## Sorting algorithms\n",
    "\n",
    "The basic principle of sorting algorithms is quite simple: we have an input sequence ($S$) of un-sorted elements $U=u_{1},u_{2}, ..., u_{n}$ and output a new sequence $S = s_{1}, s_{2}, ... , s_{n}$ which is a permutation of $U$ such that $s_{0}\\leq s_{1}, ..., \\leq s_{n}$.\n",
    "\n",
    "There are several sorting algorithms the first one we will work with is **selection sort**.\n",
    "\n",
    "### Selection sort\n",
    "\n",
    "Selection sort is the simplest of the sorting algorithms. The idea of **selection sort** is that given $U=u_{1},u_{2},...,u_{n}$ is to loop through all the elements of $U$, find the minimum $u_m$ and place it at the beginning of the sequence $U$, swapping $u_{1}$ with $u_m$. The next iteration is to continue looking for the minimum starting from $u_2$ and so on.\n",
    "\n",
    "If $U$ has $n$ elements, for each position $i=0,..,n-1$ in the list we need to perform the following two steps:\n",
    "\n",
    "1. (argmin) Find index of the minimum element in the sublist $U[i+1:]$, let's call it $m$ (i.e. $u_{m} = min(U[i:])$);\n",
    "\n",
    "2. (swap) Swap $u_m$ with $u_i$;\n",
    "\n",
    "A reminder on how selection sort works is reported in the following picture (taken from the lecture). Yellow cells are the minimum found at each iteration, while orange cells are those already sorted.\n",
    "\n",
    "![](img/pract14/selection_sort.png)\n",
    "\n",
    "### "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Insertion sort\n",
    "\n",
    "\n",
    "## A testing class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%reset -f \n",
    "\n",
    "class MySorting:\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The line ```if __name__ == \"__main__\":``` is used to specify if the code is executed as a script (i.e. it is not invoked as an imported module somewhere else in another piece of code).\n",
    "\n",
    "Executing the code as it is will not give any particular error as the tests we set-up passed correctly. But if we uncomment the ```res.append(1)``` the tests will fail and we produce the following testing output, which reports the three failed tests, the expected values and the obtained values: \n",
    "\n",
    "![](img/pract13/doctest.png)\n",
    "\n",
    "Another way of testing the code is through unit tests, we will see this later on. \n",
    "\n",
    "### Raising exceptions and using assertions\n",
    "\n",
    "One thing we can do is to raise exceptions whenever some pre-conditions are not met in order to insure that these do not lead to erroneous behaviours. This can be done with ```raise Exception(\"exception text\")```. More info on raising exceptions are [here](https://docs.python.org/3/tutorial/errors.html#raising-exceptions). \n",
    "\n",
    "**Example**: \n",
    "Consider the following ```MyIntPair``` class that works with integers. If we want to make sure it only contains integers we can add a ```raise Exception``` in case it is not an integer. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercises\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Implement a class SelectionSort that has one parameter called ```data``` (the actual data to sort), ```swaps``` (initialized to 0) that counts how many swaps have been done to perform the sorting, ```comparisons``` (initialized to 0) that counts how many comparisons have been done and ```verbose``` a boolean (default= True) that is used to decide if the method should report what is happening at each step and some stats or not. The class has one method called ```sort``` that implements the selection sort algorithm (two more methods might be needed to compute ```swap``` and ```argmin``` -- see description above). \n",
    "\n",
    "Once you implemented the class you can test it with some data like:\n",
    "```\n",
    "[7, 5, 10, -11 ,3, -4, 99, 1]\n",
    "```\n",
    "or you can create a random list of N integers with:\n",
    "```\n",
    "import random\n",
    "for i in range(0,N):\n",
    "        d.append(random.randint(0,1000))\n",
    "```\n",
    "Test the class wit N = 10000\n",
    "Add a private ```__time``` variable that computes the time spent doing the sorting. This can be done by:\n",
    "```\n",
    "import time\n",
    "...\n",
    "start_t = time.time()\n",
    "...\n",
    "end_t = time.time()\n",
    "self._time = end_t - start_t\n",
    "```\n",
    "How long does it take with a list of 10000 elements? With 20000?\n",
    "\n",
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex1');\">Show/Hide Solution</div>\n",
    "<div id=\"ex1\" style=\"display:none;\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "nbsphinx": "hidden"
   },
   "outputs": [],
   "source": [
    "%reset -f\n",
    "\n",
    "import random\n",
    "import time\n",
    "\n",
    "class SelectionSort:\n",
    "    def __init__(self,data, verbose = True):\n",
    "        self.__data = data\n",
    "        self.__comparisons = 0\n",
    "        self.__swaps = 0\n",
    "        self.__verbose = verbose\n",
    "        self.__time = 0\n",
    "        \n",
    "    def getData(self):\n",
    "        return self.__data\n",
    "    \n",
    "    def getTime(self):\n",
    "        return self.__time\n",
    "    \n",
    "    def getComparisons(self):\n",
    "        return self.__comparisons\n",
    "    \n",
    "    def getSwaps(self):\n",
    "        return self.__swaps\n",
    "    \n",
    "    def swap(self, i, j):\n",
    "        \"\"\"\n",
    "        swaps elements i and j in data.\n",
    "        \"\"\"\n",
    "        if(i != j): #no point in swapping if i==j\n",
    "            tmp = self.__data[i]\n",
    "            self.__data[i] = self.__data[j]\n",
    "            self.__data[j] = tmp\n",
    "        \n",
    "    def argmin(self, i):\n",
    "        \"\"\"\n",
    "        returns the index of the smallest element of\n",
    "        self.__data[i:]\n",
    "        \"\"\"\n",
    "        mpos = i\n",
    "        N = len(self.__data)\n",
    "        minV = self.__data[mpos]\n",
    "        for j in range(i + 1,N): # from i+1 to N. U[i+1:]\n",
    "            if(self.__data[j] < minV):\n",
    "                mpos = j\n",
    "                minV = self.__data[j]\n",
    "            #just for checking\n",
    "            self.__comparisons += 1\n",
    "        \n",
    "        return mpos\n",
    "    \n",
    "    def sort(self):\n",
    "        self.__comparisons = 0\n",
    "        self.__swaps = 1\n",
    "        if self.__verbose:\n",
    "            print(\"Initial list:\")\n",
    "            print(self.__data)\n",
    "            print(\"\\n\")\n",
    "            \n",
    "        #to check performance    \n",
    "        start_t = time.time()\n",
    "        for i in range(len(self.__data) - 1):\n",
    "                j = self.argmin(i)\n",
    "                self.swap(i,j) \n",
    "                self.__swaps += 1\n",
    "                if self.__verbose:\n",
    "                    print(\"It. {}. data[{}]<->data[{}] {}<->{}\".format(i,\n",
    "                                                                       i,\n",
    "                                                                       j,\n",
    "                                                                       self.__data[i],\n",
    "                                                                       self.__data[j]))\n",
    "                    print(self.__data)    \n",
    "        end_t = time.time()\n",
    "        \n",
    "        self.__time = end_t - start_t\n",
    "        \n",
    "        if self.__verbose:\n",
    "            print(self.__data)\n",
    "            print(\"\\nNumber of comparisons: {}\".format(self.__comparisons))\n",
    "            print(\"Number of swaps: {}\".format(self.__swaps))\n",
    "            print(\"In {:.4f}s\".format(self.__time))\n",
    "\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    d = [7, 5, 10, -11 ,3, -4, 99, 1]\n",
    "    selSorter = SelectionSort(d, verbose = True)\n",
    "    selSorter.sort()\n",
    "    d = []\n",
    "    for i in range(0,10000):\n",
    "        d.append(random.randint(0,1000))\n",
    "    selSorter = SelectionSort(d, verbose = False)\n",
    "    selSorter.sort()\n",
    "    print(\"\\nNumber of elements: {}\".format(len(d)))\n",
    "    print(\"Number of comparisons: {}\".format(selSorter.getComparisons()))\n",
    "    print(\"Number of swaps: {}\".format(selSorter.getSwaps()))\n",
    "    print(\"In {:.4f}s\".format(selSorter.getTime()))\n",
    "    test = True\n",
    "    for el in range(1,len(d)-1):\n",
    "        test = test and (d[el]<= d[el+1])\n",
    "    print(\"Sorting test passed? {}\".format(test))\n",
    "    \n",
    "    d = []\n",
    "    for i in range(0,20000):\n",
    "        d.append(random.randint(0,1000))\n",
    "    selSorter = SelectionSort(d, verbose = False)\n",
    "    selSorter.sort()\n",
    "    print(\"\\nNumber of elements: {}\".format(len(d)))\n",
    "    print(\"Number of comparisons: {}\".format(selSorter.getComparisons()))\n",
    "    print(\"Number of swaps: {}\".format(selSorter.getSwaps()))\n",
    "    print(\"In {:.4f}s\".format(selSorter.getTime()))\n",
    "    test = True\n",
    "    for el in range(1,len(d)-1):\n",
    "        test = test and (d[el]<= d[el+1])\n",
    "    print(\"Sorting test passed? {}\".format(test))\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2. Define a unittest class (see previous practical) to test the SelectionSort class. Some examples of things to check are:\n",
    "\n",
    "a. swap(i,j);swap(i,j) == original data;\n",
    "b. length of data does not change after swap;\n",
    "c. swap does not change elements other than i and j;\n",
    "d. if j=argmin(i) , data[j] is lower or equal than all elements in data[i:];\n",
    "e. after sort $\\forall$ i=0,..,n-2 : data[i]<data[i+1];\n",
    "f. Any other check?\n",
    "\n",
    "Perform unit testing on the SelectionSort class (copy the code from your previous exercise).\n",
    "\n",
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex2');\">Show/Hide Solution</div>\n",
    "<div id=\"ex2\" style=\"display:none;\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "nbsphinx": "hidden"
   },
   "outputs": [],
   "source": [
    "import random\n",
    "import time\n",
    "import unittest\n",
    "\n",
    "class SelectionSort:\n",
    "    def __init__(self,data, verbose = True):\n",
    "        self.__data = data\n",
    "        self.__comparisons = 0\n",
    "        self.__swaps = 0\n",
    "        self.__verbose = verbose\n",
    "        self.__time = 0\n",
    "        \n",
    "    def getData(self):\n",
    "        return self.__data\n",
    "    \n",
    "    def getTime(self):\n",
    "        return self.__time\n",
    "    \n",
    "    def getComparisons(self):\n",
    "        return self.__comparisons\n",
    "    \n",
    "    def getSwaps(self):\n",
    "        return self.__swaps\n",
    "    \n",
    "    def swap(self, i, j):\n",
    "        \"\"\"\n",
    "        swaps elements i and j in data.\n",
    "        \"\"\"\n",
    "        if(i != j): #no point in swapping if i==j\n",
    "            tmp = self.__data[i]\n",
    "            self.__data[i] = self.__data[j]\n",
    "            self.__data[j] = tmp\n",
    "        \n",
    "    def argmin(self, i):\n",
    "        \"\"\"\n",
    "        returns the index of the smallest element of\n",
    "        self.__data[i:]\n",
    "        \"\"\"\n",
    "        mpos = i\n",
    "        N = len(self.__data)\n",
    "        minV = self.__data[mpos]\n",
    "        for j in range(i + 1,N): # from i+1 to N. U[i+1:]\n",
    "            if(self.__data[j] < minV):\n",
    "                mpos = j\n",
    "                minV = self.__data[j]\n",
    "            #just for checking\n",
    "            self.__comparisons += 1\n",
    "        \n",
    "        return mpos\n",
    "    \n",
    "    def sort(self):\n",
    "        self.__comparisons = 0\n",
    "        self.__swaps = 1\n",
    "        if self.__verbose:\n",
    "            print(\"Initial list:\")\n",
    "            print(self.__data)\n",
    "            print(\"\\n\")\n",
    "            \n",
    "        #to check performance    \n",
    "        start_t = time.time()\n",
    "        for i in range(len(self.__data) - 1):\n",
    "                j = self.argmin(i)\n",
    "                self.swap(i,j) \n",
    "                self.__swaps += 1\n",
    "                if self.__verbose:\n",
    "                    print(\"It. {}. data[{}]<->data[{}] {}<->{}\".format(i,\n",
    "                                                                       i,\n",
    "                                                                       j,\n",
    "                                                                       self.__data[i],\n",
    "                                                                       self.__data[j]))\n",
    "                    print(self.__data)    \n",
    "        end_t = time.time()\n",
    "        \n",
    "        self.__time = end_t - start_t\n",
    "        \n",
    "        if self.__verbose:\n",
    "            print(self.__data)\n",
    "            print(\"\\nNumber of comparisons: {}\".format(self.__comparisons))\n",
    "            print(\"Number of swaps: {}\".format(self.__swaps))\n",
    "            print(\"In {:.4f}s\".format(self.__time))\n",
    "\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    unittest.main()\n",
    "\n",
    "\n",
    "class Test(unittest.TestCase):\n",
    "    def __init__(self, *args, **kwargs):\n",
    "        super(Test, self).__init__(*args, **kwargs)\n",
    "        #create a test list:\n",
    "        x = []\n",
    "        for i in range(300):    \n",
    "            x.append(random.randint(-100,100))\n",
    "        self.sorter = SelectionSort(x, verbose = False)\n",
    "                          \n",
    "    def test_swap1(self):\n",
    "        \"\"\"swap of swap is identical to beginning\"\"\"\n",
    "        #let's copy data\n",
    "        dcopy = self.sorter.getData()[:]\n",
    "        for i in range(40):\n",
    "            i1 = random.randint(0,len(dcopy) - 1)\n",
    "            i2 = random.randint(0,len(dcopy) - 1)\n",
    "            self.sorter.swap(i1,i2)\n",
    "            self.sorter.swap(i1,i2)\n",
    "            self.assertTrue(self.sorter.getData() == dcopy)\n",
    "    def test_swap2(self):\n",
    "        \"\"\"length of swapped data does not change\"\"\"\n",
    "        \n",
    "        l = len(self.sorter.getData())\n",
    "        for i in range(40):\n",
    "            i1 = random.randint(0,l - 1)\n",
    "            i2 = random.randint(0,l - 1)\n",
    "            self.sorter.swap(i1,i2)\n",
    "            self.assertTrue(len(self.sorter.getData()) == l)\n",
    "        \n",
    "    def test_swap3(self):\n",
    "        \"\"\"swapping only changes i and j indexes (if i!=j)\"\"\"\n",
    "        #let's copy data\n",
    "        dcopy = self.sorter.getData()[:]\n",
    "        for i in range(40):\n",
    "            #let's copy data\n",
    "            dcopy = self.sorter.getData()[:]\n",
    "            i1 = random.randint(0,len(dcopy) - 1)\n",
    "            i2 = random.randint(0,len(dcopy) - 1)\n",
    "            if i1 != i2:\n",
    "                self.sorter.swap(i1,i2)\n",
    "                for ind in range(0,len(dcopy)):\n",
    "                    if ind != i1 and ind != i2:\n",
    "                        self.assertTrue(dcopy[ind] == self.sorter.getData()[ind])\n",
    "    \n",
    "    def test_argmin(self):\n",
    "        \"\"\"\n",
    "        tests if j=argmin(i) then j <= data[i:]\n",
    "        \"\"\"\n",
    "        l = len(self.sorter.getData())\n",
    "        for i in range(40):\n",
    "            ind = random.randint(0,l - 1)\n",
    "            minP = self.sorter.argmin(ind)\n",
    "            for j in self.sorter.getData()[ind:]:\n",
    "                self.assertTrue(self.sorter.getData()[minP] <= j)\n",
    "                \n",
    "    def test_sort(self):\n",
    "        \"\"\"tests if the sort works\"\"\"\n",
    "        self.sorter.sort()\n",
    "        d = self.sorter.getData()\n",
    "        for el in range(0,len(d) - 2):\n",
    "            self.assertTrue(d[el] <= d[el+1])\n",
    "    \n",
    "    def test_empty(self):\n",
    "            \"\"\"sorting of empty list is empty\"\"\"\n",
    "            self.assertEqual(SelectionSort([]).getData(),[])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-info\">\n",
    "\n",
    "**Note:** \n",
    "Note the line that I used to initialize the Test class. \n",
    "```\n",
    "def __init__(self, *args, **kwargs):\n",
    "        super(Test, self).__init__(*args, **kwargs)\n",
    "```\n",
    "this allows us to define the random test data within the Test class (these lines are basically because we need to pass the super-class constructor with all the parameters it needs). \n",
    "\n",
    "</div>\n",
    "\n",
    "\n",
    "You can find a solution to run unittests here: [selection_sort_test.py](file_samples/selection_sort_test.py)\n",
    "Reminder: you can run the unittest with:\n",
    "\n",
    "```\n",
    "python3 -m unittest selection_sort_test.py\n",
    "```\n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2. CRISPR-Cas9 is quite a neat system to perform genome editing. Guide RNAs (gRNAs) can transport Cas9 to anywhere in the genome for gene editing, but no editing can occur at any site other than one at which Cas9 recognizes the protospacer adjacent motif (PAM). The PAM site is a 2-6 base pair DNA sequence immediately following the DNA sequence targeted by the Cas9 nuclease in the CRISPR bacterial adaptive immune system. Some used PAMs are the following:\n",
    "```\n",
    "NGG (where N is any base)\n",
    "NGA\n",
    "YG (where Y is a Pyrimidine, i.e. C or T)\n",
    "TTTN\n",
    "YTN\n",
    "```\n",
    "write a function that loads the fasta sequences [contig82.fasta](file_samples/contig82.fasta) and for each sequence reports the number of sites of each of the above PAMs and its frequency (i.e. number over the length of the sequence). Hint: load the sequence with biopython and SeqIO.\n",
    "\n",
    "\n",
    "\n",
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex2');\">Show/Hide Solution</div>\n",
    "<div id=\"ex2\" style=\"display:none;\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "nbsphinx": "hidden"
   },
   "outputs": [],
   "source": [
    "%reset -f\n",
    "\n",
    "from Bio import SeqIO\n",
    "import re\n",
    "\n",
    "def countPAMs(filename):\n",
    "    for seq_record in SeqIO.parse(filename, \"fasta\"):\n",
    "        s = seq_record.seq\n",
    "        l = len(s)\n",
    "        ident = seq_record.id\n",
    "        m = re.findall(\"[ATCG]GG\", str(s))\n",
    "        NGG = len(m)\n",
    "        m = re.findall(\"[ATCG]GA\", str(s))\n",
    "        NGA = len(m)\n",
    "        m = re.findall(\"[C|T]G\", str(s))\n",
    "        YG = len(m)\n",
    "        m = re.findall(\"TTT[ATCG]\", str(s))\n",
    "        TTTN = len(m)\n",
    "        m = re.findall(\"[C|T]T[ATCG]\", str(s))\n",
    "        YTN = len(m)\n",
    "        print(\"{} (len:{}):\\n\\tNGG:{} (1 on {} bases)\".format(ident, l,NGG, l/NGG))\n",
    "        print(\"\\tNGA:{} (1 on {} bases)\".format(NGA, l/NGA))\n",
    "        print(\"\\tYG:{} (1 on {} bases)\".format(YG, l/YG))\n",
    "        print(\"\\tTTTN:{} (1 on {} bases)\".format( TTTN, l/TTTN))\n",
    "        print(\"\\tYTN:{} (1 on {} bases)\".format( YTN, l/YTN))\n",
    "        \n",
    "        \n",
    "fn = \"file_samples/contigs82.fasta\"\n",
    "countPAMs(fn)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3. Write a python function ```sortCSV(mystr)``` that gets a comma separated string and returns a comma separated string with all elements sorted in alphabetically decreasing order. Define some unittests to check if the function has been implemented correctly (some things to check: length of initial string is the same as that of the final string, number of elements is the same, each element in the output string must come after the next one in lexicographical order,...). \n",
    "\n",
    "\n",
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex3');\">Show/Hide Solution</div>\n",
    "<div id=\"ex3\" style=\"display:none;\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "nbsphinx": "hidden"
   },
   "outputs": [],
   "source": [
    "%reset -f\n",
    "\n",
    "import random\n",
    "import unittest\n",
    "\n",
    "def sortCSV(mystr):\n",
    "    tmp = mystr.split(\",\")\n",
    "    tmp.sort(reverse=True)\n",
    "    return \",\".join(tmp)\n",
    "\n",
    "\n",
    "\n",
    "class Testing(unittest.TestCase):\n",
    "    def __init__(self, *args, **kwargs):\n",
    "        super(Testing, self).__init__(*args, **kwargs)\n",
    "        #create a random string\n",
    "        self.alphabet = \"abcdefghkjilmnopqrstuvwyz\"\n",
    "        self.data = \"\"\n",
    "        #create 15 random strings\n",
    "        for i in range(15):\n",
    "            word = \"\"\n",
    "            #each of them has a random length up to 20\n",
    "            j = random.randint(1,20)\n",
    "            for ind in range(j):\n",
    "                #pick up to 20 random letters\n",
    "                t = random.randint(1,len(self.alphabet)-1)\n",
    "                word += self.alphabet[t]\n",
    "            if(len(self.data) == 0):\n",
    "                self.data = word\n",
    "            else:\n",
    "                self.data += \",\" + word\n",
    "\n",
    "\n",
    "    def test_reslen(self):\n",
    "        self.assertTrue(len(self.data) == len(sortCSV(self.data)))\n",
    "\n",
    "    def test_elcount(self):\n",
    "        res = sortCSV(self.data).split(\",\")\n",
    "        self.assertTrue(len(self.data.split(\",\")) == len(res))\n",
    "\n",
    "    def test_elsorting(self):\n",
    "        res = sortCSV(self.data).split(\",\")\n",
    "        for ind in range(len(res)-1):\n",
    "            self.assertTrue(res[ind]> res[ind+1])\n",
    "\n",
    "    def test_empty(self):\n",
    "        self.assertEqual(sortCSV(\"\"),\"\")\n",
    "    \n",
    "    def test_onlyOne(self):\n",
    "        j = random.randint(1,20)\n",
    "        word = \"\"\n",
    "        for ind in range(j):\n",
    "            #pick up to 20 random letters\n",
    "            t = random.randint(0,len(self.alphabet)-1)\n",
    "            word += self.alphabet[t]\n",
    "        self.assertEqual(sortCSV(word), word)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    mystr = \"book,tree,final,example,testing,zed,all,hair,lady,figure,tap,spring,test,fin,tail\"\n",
    "    print(\"Original:\\n{}\".format(mystr))\n",
    "    print(\"Sorted:\\n{}\".format(sortCSV(mystr)))\n",
    "    unittest.main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "You can find a solution to run unittests here: [pract13_ex3.py](file_samples/pract13_ex3.py)\n",
    "\n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4. Solve a modified version of the following exercise of Practical 11 implementing a function that parses the \"ExpXml\" text through several regular expressions.\n",
    "\n",
    "The exercise is reported below:\n",
    "\n",
    "Write a python script that retrieves all the information present in SRA regarding PacBio sequencing performed on E.coli strain K12 (query term is “E.coli K12 wgs PacBio”). \n",
    "\n",
    "Print the number of results and for each id report the title, the experiment accession, the instrument, the library strategy, the library source, the total number of spots and total number of bases sequenced.\n",
    "Sample output:\n",
    "\n",
    "```\n",
    "Entries found: 9\n",
    "\n",
    "[1] Results for id 357838:\n",
    "Title: E. coli K12 PacBio RS C2 CCS sequencing\n",
    "Experiment accession: SRX255779\n",
    "Instrument: PACBIO_SMRT\n",
    "Library strategy: WGS\n",
    "Library source: GENOMIC\n",
    "Total spots:1798302\n",
    "Total bases:4228754616\n",
    "\n",
    " ...\n",
    " ...\n",
    "```\n",
    "\n",
    "A sample \"ExpXml\" string:\n",
    "\n",
    "```\n",
    "<Summary><Title>E. coli K12 PacBio RS C2 CCS sequencing</Title><Platform instrument_model=\"PacBio RS\">PACBIO_SMRT</Platform><Statistics total_runs=\"22\" total_spots=\"1798302\" total_bases=\"4228754616\" total_size=\"16799546700\" load_done=\"true\" cluster_name=\"public\"/></Summary><Submitter acc=\"SRA071585\" center_name=\"NBACC\" contact_name=\"Sergey Koren\" lab_name=\"\"/><Experiment acc=\"SRX255779\" ver=\"1\" status=\"public\" name=\"E. coli K12 PacBio RS C2 CCS sequencing\"/><Study acc=\"SRP020003\" name=\"Escherichia coli K12 Re-sequencing\"/><Organism taxid=\"511145\" ScientificName=\"Escherichia coli str. K-12 substr. MG1655\"/><Sample acc=\"SRS000462\" name=\"\"/><Instrument PACBIO_SMRT=\"PacBio RS\"/><Library_descriptor><LIBRARY_NAME>PacBio RS CCS</LIBRARY_NAME><LIBRARY_STRATEGY>WGS</LIBRARY_STRATEGY><LIBRARY_SOURCE>GENOMIC</LIBRARY_SOURCE><LIBRARY_SELECTION>unspecified</LIBRARY_SELECTION><LIBRARY_LAYOUT> <SINGLE/> </LIBRARY_LAYOUT></Library_descriptor><Bioproject>PRJNA194437</Bioproject><Biosample>SAMN00000224</Biosample>\n",
    "```\n",
    "\n",
    "\n",
    "\n",
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex4');\">Show/Hide Solution</div>\n",
    "<div id=\"ex4\" style=\"display:none;\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "nbsphinx": "hidden"
   },
   "outputs": [],
   "source": [
    "%reset -f \n",
    "\n",
    "import re\n",
    "from Bio import Entrez\n",
    "\n",
    "\n",
    "\n",
    "def parseExp(expStr):\n",
    "    m = re.search(\"<Title>([A-Za-z0-9_\\. \\(\\)]*)</Title>\", expStr)\n",
    "    if(m):\n",
    "        title = \"Title: \" + m.group()[7:-8]\n",
    "        print(title)\n",
    "        \n",
    "    m = re.search(\"<Experiment acc=\\\"([A-Z0-9]*)\\\"\", expStr)\n",
    "    if(m):\n",
    "        acc = \"Experiment accession: \" + m.groups()[0]\n",
    "        print(acc)\n",
    "    m = re.search(\"<Platform ([A-Za-z0-9_=\\\" \\(\\)]*)>([A-Za-z0-9_\\(\\)]*)</Platform>\", expStr)\n",
    "    if(m):\n",
    "        platform = \"Instrument: \" + m.groups()[1]\n",
    "        print(platform)\n",
    "    m = re.search(\"<LIBRARY_STRATEGY>([A-Za-z0-9_=\\\" \\(\\)]*)</LIBRARY_STRATEGY>\", expStr)\n",
    "    if(m):\n",
    "        src = \"Library strategy: \"  + m.groups()[0]\n",
    "        print(src)\n",
    "    m = re.search(\"<LIBRARY_SOURCE>([A-Za-z0-9_=\\\" \\(\\)]*)</LIBRARY_SOURCE>\", expStr)\n",
    "    if(m):\n",
    "        src = \"Library source: \"  + m.groups()[0]    \n",
    "        print(src)\n",
    "    m = re.search(\"total_spots=\\\"([0-9]*)\\\" total_bases=\\\"([0-9]*)\\\"\",expStr)\n",
    "    if(m):\n",
    "        spots = \"Total spots:\" + m.groups()[0] +\"\\nTotal bases:\" + m.groups()[1]\n",
    "        print(spots)\n",
    "\n",
    "    \n",
    "Entrez.email = \"my_email\"\n",
    "handle = Entrez.esearch(db=\"sra\", term=\"E.coli K12 wgs PacBio\", retmax = 10)\n",
    "res = Entrez.read(handle)\n",
    "#uncomment to see all fields:\n",
    "#for el in res.keys():\n",
    "#    print(el , \" : \", res[el])\n",
    "\n",
    "print(\"Entries found: {}\".format(res[\"Count\"]))\n",
    "\n",
    "cnt = 1\n",
    "for ids in res[\"IdList\"]:\n",
    "    print(\"\\n[{}] Results for id {}:\".format(cnt, ids))\n",
    "    handle = Entrez.esummary(db=\"sra\",  id = ids)\n",
    "    res = Entrez.read(handle)\n",
    "    cnt += 1\n",
    "          \n",
    "    for r in res:\n",
    "        info = r['ExpXml']\n",
    "        #print(info)\n",
    "        parseExp(info)\n",
    "        \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "5. The file (DNA_seq.fasta)[file_samples/DNA_seq.fasta] contains a synthetic DNA sequence. Let's assume to have two restriction enzymes LagI and JagII that respectively cut at the site CNC/ATT, and GAGRK/TNG (where N is any site, R is A or G and K is A or C or T. Note that \"/\" is just a representation of where the enzyme cuts, therefore we do not need to specify this in the regular expression, but we need to take it into account when we cut the DNA.\n",
    "\n",
    "Ex. if the sequence is:\n",
    "ATACATTCCCCCGGAATCGCCCCCCCTCCATTCC\n",
    "digesting the sequence with LagI would give:\n",
    "[\"ATAC\",\"ATTCCCCCGGAATCGCCCCCCCTCC\", \"ATTCC\"]\n",
    "digesting this further with JagII would give:\n",
    "[\"ATAC\",\"ATTCCCCCGGAA\", \"TCGCCCCCCCTCC\", \"ATTCC\"]\n",
    "\n",
    "Write a python script that simulates a digestion with LagI, size selection to keep only the fragments higher than 50 base pairs, and a digestion with JagII, printing the lengths of the obtained fragments. What happens to the fragments if we digest first with JagII and then with LagI?\n",
    "\n",
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex5');\">Show/Hide Solution</div>\n",
    "<div id=\"ex5\" style=\"display:none;\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "nbsphinx": "hidden"
   },
   "outputs": [],
   "source": [
    "%reset -f \n",
    "\n",
    "from Bio import SeqIO\n",
    "import re\n",
    "\n",
    "#in our case overhang can be 3 or 4 (3 for LagI and 4 for JagII)\n",
    "def digestSequence(seq, regex, overhang):\n",
    "    digests = []\n",
    "    sP = 0\n",
    "    matches = re.finditer(regex,seq)\n",
    "    if matches:\n",
    "        for site in matches:\n",
    "            print(\"\\tRestriction sites:\")\n",
    "            print(\"\\t{} {} {}\".format(site.start(),\n",
    "                                   site.end(),\n",
    "                                   site.group()))\n",
    "            digests.append(seq[sP:site.start()+overhang])\n",
    "            sP = site.start() + overhang\n",
    "        #last element:\n",
    "        digests.append(seq[sP:])\n",
    "    return digests\n",
    "\n",
    "fn = \"file_samples/DNA_seq.fasta\"\n",
    "\n",
    "myseq = SeqIO.read(fn, \"fasta\")\n",
    "\n",
    "regexLagI = \"C[ATCG]CATT\"\n",
    "regexJagII = \"GAG[AG][ACT]T[ATCG]G\"\n",
    "\n",
    "\n",
    "s = str(myseq.seq)\n",
    "\n",
    "print(\"Initial sequence:\")\n",
    "print(s)\n",
    "print(\"\\n\")\n",
    "\n",
    "        \n",
    "print(\"LagI restriction:\")\n",
    "digests = digestSequence(s, regexLagI, 3)\n",
    "#filter digests\n",
    "dig = [x for x in digests if len(x) > 50]\n",
    "print(\"Lengths:\")\n",
    "print([len(x) for x in dig])\n",
    "finalDigests = []\n",
    "\n",
    "print(\"\\nJagII restriction:\")\n",
    "for d in dig:\n",
    "    tmp = digestSequence(d, regexJagII, 5)\n",
    "    for t in tmp:\n",
    "        finalDigests.append(t)\n",
    "\n",
    "#print(\"Final digests:\" + str(finalDigests))\n",
    "print(\"Final lengths: \" + str([len(x) for x in finalDigests])) \n",
    "\n",
    "\n",
    "#Other JagII first and LagI second:\n",
    "print(\"\\n###########################\")\n",
    "print(\"##### JagII first #########\")\n",
    "print(\"###########################\")\n",
    "print(\"LagI restriction:\")\n",
    "digests = digestSequence(s, regexJagII, 5)\n",
    "#filter digests\n",
    "dig = [x for x in digests if len(x) > 50]\n",
    "print(\"Lengths:\")\n",
    "print([len(x) for x in dig])\n",
    "finalDigests = []\n",
    "\n",
    "print(\"\\nJagII restriction:\")\n",
    "for d in dig:\n",
    "    tmp = digestSequence(d, regexLagI, 3)\n",
    "    for t in tmp:\n",
    "        finalDigests.append(t)\n",
    "\n",
    "#print(\"Final digests:\" + str(finalDigests))\n",
    "print(\"Final lengths: \" + str([len(x) for x in finalDigests])) \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "</div>"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Edit Metadata",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
