{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "nbsphinx": "hidden"
   },
   "source": [
    "### NOTE FOR LUCA\n",
    "\n",
    "**Remember to set/remove metadata as:**\n",
    "{\n",
    "  \"nbsphinx\": \"hidden\"\n",
    "}\n",
    "\n",
    "to enable/disable solutions view"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# Practical 13\n",
    "\n",
    "In this practical we will continue with object oriented programming and will focus on code testing in Python. Finally, we will introduce regular expressions.\n",
    "\n",
    "## Slides\n",
    "\n",
    "The slides of the introduction can be found here: [Intro](docs/Practical13.pdf)\n",
    "\n",
    "## Testing your code \n",
    "\n",
    "Testing the code is quite an important step to make sure that the code is predictable and although some bugs can always slip through, **testing is the process of making the code as predictable as possible**.\n",
    "\n",
    "Two types of testing exist: the **white-box** testing methodology looks into every detail of the implemented code to inspect its correctness, **black-box** testing does not look at the details of how the code is implemented, but it just focus on the correctness of the output produced by the code.\n",
    "\n",
    "Testing is quite a complex and articulated topic, here we will just scratch the tip of the iceberg (there are many books on the topic if you are interested).\n",
    "\n",
    "A **test** is a piece of code written with the sole purpose of checking the correctness of another piece of code. \n",
    "\n",
    "Testing requires three successive moments: first of all we need to **set up** (or prepare) the test setting up connections/interfaces to test data, the second step is to **execute** the piece of code we are aiming to test using the interfaces devised at the previous step, and the third is the **verification** of the results to make sure they look as they were expected to.\n",
    "\n",
    "### Doctest \n",
    "\n",
    "A very simple way to specifying tests for the code is by using an embedded module called **doctest**. It will basically search for pieces of code in your python file that look like **interactive python sessions** (that are lines starting with ```>>> ```) and will execute them to check if they run giving the result specified in the next line.\n",
    "\n",
    "<div class=\"alert alert-warning\">\n",
    "\n",
    "**Important:** \n",
    "\n",
    "Note the space after \">>> \". That is where the test starts. An example:\n",
    "```\n",
    "\"\"\"\n",
    "This is a function that returns three values in a list...\n",
    ">>> fun(x)\n",
    "[x, y, z]\n",
    "\"\"\"\n",
    "```\n",
    "</div>\n",
    "\n",
    "**Example:**\n",
    "Let's define some doctest tests for the simple function computing the first N prime numbers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%reset -f \n",
    "\n",
    "def getFirstNprimes(N):\n",
    "    '''\n",
    "    This function should output the first N prime numbers.\n",
    "    >>> getFirstNprimes(1)\n",
    "    [2]\n",
    "    >>> getFirstNprimes(2)\n",
    "    [2, 3]\n",
    "    >>> getFirstNprimes(10)\n",
    "    [2, 3, 5, 7, 11, 13, 17, 19, 23, 29]\n",
    "    '''\n",
    "    if N == 0:\n",
    "        return []\n",
    "    res = [2]\n",
    "    current = 3\n",
    "    while len(res) < N:\n",
    "        if len([x for x in res if current % x == 0]) == 0:\n",
    "            res.append(current)\n",
    "        current += 1\n",
    "    #uncomment next line to introduce a bug\n",
    "    #res.append(1)\n",
    "    return res        \n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    import doctest\n",
    "    doctest.testmod()\n",
    "    print(getFirstNprimes(20))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The line ```if __name__ == \"__main__\":``` is used to specify if the code is executed as a script (i.e. it is not invoked as an imported module somewhere else in another piece of code).\n",
    "\n",
    "Executing the code as it is will not give any particular error as the tests we set-up passed correctly. But if we uncomment the ```res.append(1)``` the tests will fail and we produce the following testing output, which reports the three failed tests, the expected values and the obtained values: \n",
    "\n",
    "![](img/pract13/doctest.png)\n",
    "\n",
    "Another way of testing the code is through unit tests, we will see this later on. \n",
    "\n",
    "### Raising exceptions and using assertions\n",
    "\n",
    "Exceptions are a very good way to inform the program that some unexpected thing has happened (e.g. division by zero, tentative to access a position in a list that does not exist, summing a string and an integer...).\n",
    "\n",
    "One thing we can do is to raise exceptions whenever some pre-conditions are not met in order to insure that these do not lead to erroneous behaviours. This can be done with ```raise Exception(\"exception text\")```. More info on raising exceptions are [here](https://docs.python.org/3/tutorial/errors.html#raising-exceptions). \n",
    "\n",
    "**Example**: \n",
    "Consider the following ```MyIntPair``` class that works with integers. If we want to make sure it only contains integers we can add a ```raise Exception``` in case it is not an integer. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class MyIntPair:\n",
    "    def __init__(self, x,y):\n",
    "        if not type(x) == int:\n",
    "            raise Exception(\"x {} is not integer\".format(x))\n",
    "        if not type(y) == int:\n",
    "            raise Exception(\"y {} is not integer\".format(y))\n",
    "        self.x = x\n",
    "        self.y = y\n",
    "        \n",
    "    def __add__(self, other):\n",
    "        return (self.x + other.x, self.y + other.y)\n",
    "    \n",
    "A = MyIntPair(5,10)\n",
    "B = MyIntPair(3,6)\n",
    "print(A + B)\n",
    "C = MyIntPair(1, \"two\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that if we try to pass it something other than an integer, the execution will stop. To amend this behaviour we need to intercept the exception and deal with it appropriately. This topic is quite articulated, you can find more info on it [here](https://docs.python.org/3/tutorial/errors.html#handling-exceptions).\n",
    "\n",
    "A simple exception handling can be done by using the ```try - except``` construct that tries to run some statements, being ready to intercept and handle any exception that might occur:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class MyIntPair:\n",
    "    def __init__(self, x,y):\n",
    "        if not type(x) == int:\n",
    "            raise Exception(\"x: {} is not integer\".format(x))\n",
    "        if not type(y) == int:\n",
    "            raise Exception(\"y: {} is not integer\".format(y))\n",
    "        self.x = x\n",
    "        self.y = y\n",
    "        \n",
    "    def __add__(self, other):\n",
    "        return (self.x + other.x, self.y + other.y)\n",
    "    \n",
    "try: \n",
    "    A = MyIntPair(5,10)\n",
    "    B = MyIntPair(3,6)\n",
    "    #Uncomment to see a different error\n",
    "    #print(A/0)\n",
    "    print(A + B)\n",
    "    C = MyIntPair(1, \"two\")\n",
    "    print(A + C)\n",
    "\n",
    "except Exception as e:\n",
    "    print(\"Whoops something went wrong. Ignore the rest.\")\n",
    "    print(str(e))\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Similarly, we can produce assertions that test if a condition is True, otherwise the execution is stopped with an ```AssertionError```. This error can be caught and handled appropriately. Note that in the code below, only the AssertionError is caught and a division by 0 (e.g. print(10/0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class MyIntPair:\n",
    "    def __init__(self, x,y):\n",
    "        assert type(x) == int, \"x: {} is not integer\".format(x)\n",
    "        assert type(y) == int, \"y: {} is not integer\".format(y)\n",
    "        self.x = x\n",
    "        self.y = y\n",
    "        \n",
    "        \n",
    "    def __add__(self, other):\n",
    "        return (self.x + other.x, self.y + other.y)\n",
    "    \n",
    "try: \n",
    "    A = MyIntPair(5,10)\n",
    "    B = MyIntPair(3,6)\n",
    "    #Uncomment to see a different error (not captured!)\n",
    "    #print(10/0)\n",
    "    print(A + B)\n",
    "    C = MyIntPair(1, \"two\")\n",
    "    print(A + C)\n",
    "\n",
    "except AssertionError as e:\n",
    "    print(\"Whoops something went wrong. Ignore the rest.\")\n",
    "    print(str(e))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Unit tests\n",
    "\n",
    "Python comes with a fully-fledgeted testing module which is called ```unittest```. Have a look [here](https://docs.python.org/3.8/library/unittest.html) for detailed information on this module.\n",
    "\n",
    "The module ```unittest``` must be imported first with ```import unittest``` and then the Test class must be implemented to perform the tests.\n",
    "\n",
    "In a nutshell, to create some unit tests we need to define a ```Testing``` class (we are free to call it as we like) which is a subclass of the class ```unittest.TestCase```. Within this class, we have then to specify the tests we want to run. Every test is a method and its name **must start** with ```test_``` (e.g. ```test_length```).\n",
    "\n",
    "Tests can use assertions ```assertEqual(value1, value2)```, ```assertTrue(condition)``` or ```assertFalse(condition)``` that respectively allow to check the equality of two values (i.e. the known result and the output of the method to be tested) and the truth value of a condition (typically computed on the output of the method under test).\n",
    "\n",
    "We can add the invocation to the ```unittest``` module within the script by adding (```unittest.main()```) for example in the ```main``` part through ```if __name__ == \"__main__\":```. In this case, we can perform the tests just by calling something like:\n",
    "```\n",
    "python3 my_testing_function.py\n",
    "```\n",
    "\n",
    "without specifying ```unittest.main()``` in the script, we need to call the unittest with:\n",
    "```\n",
    "python3 -m unittest my_testing_function.py\n",
    "```\n",
    "\n",
    "The unittest will provide us feedback on how the tests performed. In particular, if all the tests are passed we should get something like:\n",
    "\n",
    "```\n",
    "python3 -m unittest file_samples/my_testing_function.py \n",
    "\n",
    ".....\n",
    "----------------------------------------------------------------------\n",
    "Ran 5 tests in 0.387s\n",
    "\n",
    "OK\n",
    "```\n",
    "otherwise, in case of some errors like:\n",
    "\n",
    "```\n",
    "...FF\n",
    "======================================================================\n",
    "FAIL: test_one (file_samples.my_testing_function.Testing)\n",
    "----------------------------------------------------------------------\n",
    "Traceback (most recent call last):\n",
    "  File \"/home/biancol/Google Drive/work/courses/QCBsciprolab/file_samples/my_testing_function.py\", line 33, in test_one\n",
    "    self.assertEqual(getFirstNprimes(1),[2])\n",
    "AssertionError: Lists differ: [10] != [2]\n",
    "\n",
    "First differing element 0:\n",
    "10\n",
    "2\n",
    "\n",
    "- [10]\n",
    "+ [2]\n",
    "\n",
    "======================================================================\n",
    "FAIL: test_ten (file_samples.my_testing_function.Testing)\n",
    "----------------------------------------------------------------------\n",
    "Traceback (most recent call last):\n",
    "  File \"/home/biancol/Google Drive/work/courses/QCBsciprolab/file_samples/my_testing_function.py\", line 37, in test_ten\n",
    "    [2, 3, 5, 7, 11, 13, 17, 19, 23, 29])\n",
    "AssertionError: Lists differ: [2, 3, 5, 7, 11, 13, 17, 10, 23, 29] != [2, 3, 5, 7, 11, 13, 17, 19, 23, 29]\n",
    "\n",
    "First differing element 7:\n",
    "10\n",
    "19\n",
    "\n",
    "- [2, 3, 5, 7, 11, 13, 17, 10, 23, 29]\n",
    "?                           ^\n",
    "\n",
    "+ [2, 3, 5, 7, 11, 13, 17, 19, 23, 29]\n",
    "?                           ^\n",
    "\n",
    "\n",
    "----------------------------------------------------------------------\n",
    "Ran 5 tests in 0.770s\n",
    "\n",
    "FAILED (failures=2)\n",
    "\n",
    "```\n",
    "\n",
    "**Example:**\n",
    "Let's define some doctest tests for the simple function computing the first N prime numbers. The file is available here: [my_testing_function.py](file_samples/my_testing_function.py)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%reset -f \n",
    "\"\"\"\n",
    "In file my_testing_function.py\n",
    "\"\"\"\n",
    "\n",
    "import unittest\n",
    "import random\n",
    "\n",
    "def getFirstNprimes(N):\n",
    "    '''\n",
    "    This function should output the first N prime numbers.\n",
    "     '''\n",
    "    if N <= 0:\n",
    "        return []\n",
    "    res = [2]\n",
    "    current = 3\n",
    "    while len(res) < N:\n",
    "        if len([x for x in res if current % x == 0]) == 0:\n",
    "            res.append(current)\n",
    "        current += 1\n",
    "    #uncomment next line to introduce a bug\n",
    "    #res.append(1)\n",
    "    #or a more subtle error:\n",
    "    #ind = random.randint(len(res))\n",
    "    #res[ind] = 10\n",
    "    return res        \n",
    "\n",
    "class Testing(unittest.TestCase):\n",
    "    def test_empty(self):\n",
    "        self.assertEqual(getFirstNprimes(0),[])\n",
    "    \n",
    "    def test_one(self):\n",
    "        self.assertEqual(getFirstNprimes(1),[2])\n",
    "    \n",
    "    def test_ten(self):\n",
    "        self.assertEqual(getFirstNprimes(10),\n",
    "                         [2, 3, 5, 7, 11, 13, 17, 19, 23, 29])\n",
    "    \n",
    "    def test_len(self):\n",
    "        for i in range(0,10):\n",
    "            n = random.randint(1,1000)\n",
    "            self.assertFalse(len(getFirstNprimes(n)) != n)\n",
    "    \n",
    "    \n",
    "    def test_negative(self):\n",
    "        self.assertTrue(len(getFirstNprimes(-1)) == 0)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    #uncomment to run the tests in the main (without -m)\n",
    "    #unittest.main()\n",
    "    print(getFirstNprimes(20))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Running the tests on the previous code without any \"intentional bug\" will produce an OK message as above. Note that the random error is quite tricky as it affects different positions all the time (try running the test several times). This type of errors is quite difficult to detect and requires several rounds of testing as if some random steps are taking place, the error might not occur all the times."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Regular Expressions \n",
    "\n",
    "A **regular expression** (**regex**) is a string of characters defining a search pattern with which we can carry out operations such as pattern and string matching, find/replace etc.\n",
    "\n",
    "There are two types of characters: **normal characters** (which have to match amongst themselves) and **special characters** (which are used to specify repetitions (```*, ?, +, {x,y}```), a set of elements (```[]```), negation (```[^]```), beginning (```^```) of a string, end of a string (```$```), etc.\n",
    "\n",
    "As seen in the lecture, the syntax of the most common regular expressions is the following:\n",
    "\n",
    "![](img/pract13/regex.png)\n",
    "\n",
    "With regex it is possible to filter strings depending on some patterns that could include specific words (present 0 or several times using the ```?```, ```+``` or ```*``` or from ```n``` to ```m``` occurrences ```{n,m}```) at the beginning of the string ```^``` or end of the string with  ```$```. Subsets of characters can be specified with ```[]```, negation of a subset of characters with ```[^]```.\n",
    "\n",
    "Let's see some examples:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What does the following regex match?\n",
    "\n",
    "```\n",
    "regex = \"[A-Z]__[0-9]{1,4}__[a-z:-]*__[A-Z]\"\n",
    "```\n",
    "One capital letter, two underscores, from 1 to 4 numbers, two underscores, from 0 to infinite small letters, : or -, two underscores and one capital letter. Examples:\n",
    "\n",
    "```\n",
    "B__2786__prxywflh-bityeqwdmuzbygwpjtadzbvjyzmq:prucyz-rkrgjmytczdjejsgvpn-__X\n",
    "Q__5879__rarwqytmpqa-l__M\n",
    "A__45__lwonmfel-qdbcd__X\n",
    "T__5776__lmeqq__T\n",
    "A__126____T\n",
    "\n",
    "```\n",
    "\n",
    "What does the following regex match (assuming only IUPAC ambiguous nucleotide alphabet)?\n",
    "\n",
    "```\n",
    "regex = \"[ATCG]*([^ATCG]+[ATCG]*)*\"\n",
    "```\n",
    "macthes all DNA strings with zero or more A,T,C,G bases, followed by a non A,T,C,G base (e.g. N), followed by zero or more A,T,C,G bases:\n",
    "```\n",
    "ATACAYYATATACA\n",
    "CGCTTANTAT\n",
    "AAATTCGW\n",
    "TTACWWWWWWWWA\n",
    "```\n",
    "\n",
    "What does the following regex encode for?\n",
    "\n",
    "```\n",
    "regex=\"^\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}$```\n",
    "```\n",
    "it matches:\n",
    "\n",
    "```\n",
    "61.92.71.6\n",
    "497.7.0.88\n",
    "168.9.620.755\n",
    "21.5.5.1\n",
    "6.994.06.45\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In python there is a specific regular expression module to deal with these. It is called ```re``` and must be imported with ```import re``` before using it.\n",
    "\n",
    "### Match, search, finditer and findall\n",
    "The first two methods we are going to see are *match*, *search* and *finditer*, which return a ```MatchObject```:\n",
    "\n",
    "* ```re.match(regex, str)``` where ```regex``` is a string representing the regular expression, ```str``` is the input string. This method tries to match the regex on the string starting **from the beginning** (i.e. left-to-right) of the string. It will return an ```MatchObject``` or None if not match could be found;\n",
    "\n",
    "* ```re.search(regex,str)``` searches ```regex``` in the whole string and returns a ```MatchObject``` with the first occurrence of the pattern or None if the regex could not match anything;\n",
    "\n",
    "* ```re.finditer(regex,str)```: returns an iterator to ```MatchObject``` instances over all **non-overlapping** matches for the regular expression pattern in string. The string is scanned left-to-right, and matches are returned in the order found;\n",
    "\n",
    "\n",
    "Given a ```MatchObject```, if not None, provides the following information:\n",
    "\n",
    "* ```MatchObject.group()``` : the matched string;\n",
    "\n",
    "* ```MatchObject.start()``` : the starting point of the matched string in the tested string (str);\n",
    "\n",
    "* ```MatchObject.end()``` : the ending point of the matched string in the tested string (str); \n",
    "\n",
    "* ```MatchObject.groups()``` : when defining a regular expression, we can define subgroups by using \"()\". This method returns a tuple containing all the subgroups.\n",
    "\n",
    "Another method available is:\n",
    "\n",
    "```re.findall(regex,str)``` returns a **list** with all the occurrences of the regex in the string. Note that **if a group is specified** (with \"()\"), this method reports only the group.\n",
    "\n",
    "Let's see some examples of match:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%reset -f\n",
    "\n",
    "import re \n",
    "\n",
    "\n",
    "myStr = \"hi there, i am using Python and learning hOw to UsE Regular Expressions AKA REGEX\"\n",
    "\n",
    "m = re.search(\"123\", myStr)\n",
    "print(myStr)\n",
    "if(m):\n",
    "    print(\"123 is in myStr\")\n",
    "else:\n",
    "    print(\"123 is NOT in myStr\")\n",
    "    print(\"m is {}\".format(m))\n",
    "    \n",
    "a = myStr.split()\n",
    "print(\"\\nCapitalized words:\")\n",
    "for word in a:\n",
    "    match = re.match(\"[A-Z]+[a-zA-Z]*\",word)\n",
    "    if(match):\n",
    "        print(word)\n",
    "        \n",
    "print(\"\\nOn the whole string with SEARCH:\")\n",
    "result = re.search(\" [A-Z]+[a-zA-Z]*|^[A-Z]+[a-zA-Z]*| [A-Z]+[a-zA-Z]*$\", myStr)\n",
    "\n",
    "print(\"{}: starts:{} ends: {} that is: \\\"{}\\\"\".format(result.group(), \n",
    "                                        result.start(), \n",
    "                                        result.end(), \n",
    "                                        myStr[result.start():result.end()]))\n",
    "\n",
    "print(\"\\nIteratively with finditer\")\n",
    "#get all the words starting with a capital letter\n",
    "for m in re.finditer(\" [A-Z]+[a-zA-Z]*|^[A-Z]+[a-zA-Z]*| [A-Z]+[a-zA-Z]*$\", myStr):\n",
    "    print(type(m))\n",
    "    print(\"{}: starts:{} ends: {} that is: \\\"{}\\\"\".format(m.group(), \n",
    "                                        m.start(), \n",
    "                                        m.end(), \n",
    "                                        myStr[m.start():m.end()]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Example:**\n",
    "There are three types of valid Uniprot accessions:\n",
    "\n",
    "1. [O,P,Q][0-9][A-Z,0-9][A-Z,0-9][A-Z,0-9][0-9]\n",
    "2. [A-N,R-Z][0-9][A-Z][A-Z,0-9][A-Z,0-9][0-9]\n",
    "3. [A-N,R-Z][0-9][A-Z][A-Z,0-9][A-Z,0-9][0-9][A-Z][A-Z,0-9][A-Z,0-9][0-9]\n",
    "\n",
    "Devise a single regex capable of capturing all of them and use these to filter out the strings that do not encode for valid Uniprot accessions from the following list:\n",
    "\n",
    " [\"P68250\", \"SRX4799393\", \"XT44312B\",\"LFKT12DDF\", \"A0A022YWF9\", \n",
    " \"PRJNA230403\", \"P15163\", \"Q86US8\",\"W0075458831\", \n",
    " \"A0A2K5MMC5\", \"SAMN10177882\",\"Q8LKW0\", \"Q8IEH2_PLAF7\"]\n",
    "\n",
    "Test the strings that passed the filter by accessing: \n",
    "https://www.uniprot.org/uniprot/xxxxxxx where \"xxxxxxx\" is a valid Uniprot accession."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%reset -f \n",
    "import re\n",
    "\n",
    "accessionList = [\"P68250\", \"SRX4799393\", \"XT44312B\",\"LFKT12DDF\",\n",
    "                 \"A0A022YWF9\", \"PRJNA230403\", \"P15163\", \n",
    "                 \"Q86US8\",\"W0075458831\", \"A0A2K5MMC5\", \"SAMN10177882\",\n",
    "                \"Q8LKW0\", \"Q8IEH2_PLAF7\"]\n",
    "for el in accessionList:\n",
    "    match = re.match(\"[OPQ][0-9][A-Z0-9]{3}[0-9]|[A-NR-Z][0-9]([A-Z][A-Z0-9]{2}[0-9]){1,2}\",el)\n",
    "    if(match):\n",
    "        print(\"https://www.uniprot.org/uniprot/\" + match.group())\n",
    "    else:\n",
    "        print(\"skipping:\" + el)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Example:**\n",
    "The BisI restriction enzyme cuts the DNA at a specific site (restriction site): \n",
    "```\n",
    "GCNGC where N is any base \n",
    "```\n",
    "given the following DNA string:\n",
    "```\n",
    "DNA = \"ACAAAGACCAAGAAAATGGAATACTCGGAGGGGTAGACCACTAGAAGTGAGCTGCATAAGCGCCGCAAGCGGATTATTTGCCACTCAGTCATTGGCAGGAGTTGCAGGCTGCCAGAGTTCGACAGTTGCAGTATGCCGCGGATCCACGTTTGGAGCAGCGCCAGTTTGGATAGACACCAGGACCCAGAAGCATTACAAAATGCGGCGTAACCTCGCTGTAGACCATGAAACTACGACCGGGTGAATCAGTCCACTCTGCTGCTGGAGCTACCATTGGCAGCACCGGTTGTGGGTAAGGAAAGCTCACCGTTAATCTACTGCAGATGGACAGAGTCCGGGTCCTACACACGCGGCGAGACACCCATTTGTGGCCAGTAGCCGCTTCAGGTCGAGTAATCTGCCAGCAAATCGCGGCGAGCGAGTGACTTGTCGACGAAGCAAAGCAATGCAAACCGTTGGCGGCAGCAATCTCCTGCATATGATGGTGGGACGCTACAATAATCATGTACCAGGCAGCCTATGCATAATATGGCTGCCCACCTACGGGCTATTATAGTGCGACAGCTACCAAAAAAGGTCGCTAGCATGCGCGATACAGCTAACTTTTACGCAGCATACCTTTCCAGCTAGCCGCGATTGT\" \n",
    "```\n",
    "write a regex and some python code that answers the following questions:\n",
    "\n",
    "1. print the start and end position of the first restriction site and the corresponding sequence;\n",
    "2. print all the positions of the restriction sites and the corresponding sequence;\n",
    "3. count how many restrictions sites are there and print a list with all the distances between consecutive restriction sites\n",
    "4. define two groups to get the whole restriction site and also the base in between GCs;\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "DNA = \"\"\"ACAAAGACCAAGAAAATGGAATACTCGGAGGGGTAGACCACTAGAAGTGAGCTGCATAAGCGCCGCAAGCGGAT\n",
    "TATTTGCCACTCAGTCATTGGCAGGAGTTGCAGGCTGCCAGAGTTCGACAGTTGCAGTATGCCGCGGATCCACGTTTGGAGCA\n",
    "GCGCCAGTTTGGATAGACACCAGGACCCAGAAGCATTACAAAATGCGGCGTAACCTCGCTGTAGACCATGAAACTACGACCGG\n",
    "GTGAATCAGTCCACTCTGCTGCTGGAGCTACCATTGGCAGCACCGGTTGTGGGTAAGGAAAGCTCACCGTTAATCTACTGCAG\n",
    "ATGGACAGAGTCCGGGTCCTACACACGCGGCGAGACACCCATTTGTGGCCAGTAGCCGCTTCAGGTCGAGTAATCTGCCAGCA\n",
    "AATCGCGGCGAGCGAGTGACTTGTCGACGAAGCAAAGCAATGCAAACCGTTGGCGGCAGCAATCTCCTGCATATGATGGTGGG\n",
    "ACGCTACAATAATCATGTACCAGGCAGCCTATGCATAATATGGCTGCCCACCTACGGGCTATTATAGTGCGACAGCTACCAAA\n",
    "AAAGGTCGCTAGCATGCGCGATACAGCTAACTTTTACGCAGCATACCTTTCCAGCTAGCCGCGATTGT\"\"\" \n",
    "\n",
    "DNA = DNA.replace(\"\\n\",\"\")\n",
    "regex = \"GC[ATCG]GC\"\n",
    "#let's try with match\n",
    "match = re.match(regex,DNA)\n",
    "if match:\n",
    "    print(\"DNA starts with a restriction site\\n\")\n",
    "else:\n",
    "    print(\"DNA does NOT start with a restriction site\\n\")\n",
    "#with search\n",
    "first = re.search(regex,DNA)\n",
    "if first:\n",
    "    print(\"1st Restriction site starts at {} ends at {}: {}\\n\".format(first.start(),\n",
    "                                                                first.end(),\n",
    "                                                                first.group()\n",
    "                                                               ))\n",
    "allrest = re.finditer(regex,DNA)\n",
    "if allrest: \n",
    "    cnt = 0\n",
    "    curPos = 0\n",
    "    distances = []\n",
    "    for site in allrest:\n",
    "        print(\"{} res. site starts at {} ends at {}: {}\".format(cnt + 1,site.start(),\n",
    "                                                                site.end(),\n",
    "                                                                site.group()\n",
    "                                                               ))\n",
    "        cnt +=1\n",
    "        distances.append(site.start() - curPos -1 )\n",
    "        curPos = site.end()\n",
    "\n",
    "print(\"\\nWe have {} restriction sites\".format(cnt))\n",
    "print(\"Distances:\\n{}\".format(distances))\n",
    "\n",
    "regex = \"(GC([ATCG])GC)\"\n",
    "allrest = re.finditer(regex,DNA)\n",
    "if allrest: \n",
    "    for site in allrest:\n",
    "        print(\"res. site S:{} E:{}: {} whole group: {} base: {}\".format(site.start(),\n",
    "                                                                site.end(),\n",
    "                                                                site.group(),\n",
    "                                                                site.groups()[0],\n",
    "                                                                site.groups()[1]\n",
    "                                                               ))\n",
    "\n",
    "print(\"\\nFindall test (tuple with all groups!):\")\n",
    "allres = re.findall(regex,DNA)\n",
    "print(allres)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A lot of information on regex can be found [here](https://docs.python.org/3.7/library/re.html) and [here](https://docs.python.org/3/howto/regex.html).\n",
    "\n",
    "If you want to practice with regular expressions, you can use this website: [https://regex101.com/r/gU7oT6/2](https://regex101.com/r/gU7oT6/2)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercises\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. The following function is supposed to get two lists of integers (let's call then X and Y) and return the list of elements that are contained in both (let's call it B). Is it correct? Devise a unit test to check if it is correct or not. In the latter case propose a correct version of the function.\n",
    "\n",
    "```\n",
    "def myListIntersection(X,Y):\n",
    "    tmp = X + Y\n",
    "    vals = [x for x in tmp if tmp.count(x) == 2]\n",
    "    return list(set(vals))\n",
    "    \n",
    "A = [1, 2, 3, 4, 7, 12]\n",
    "B = [4, 1, 7, 120]\n",
    "C = [120, 6]\n",
    "D = []\n",
    "\n",
    "print(\"A, B: {}\".format(myListIntersection(A,B)))\n",
    "print(\"A, C: {}\".format(myListIntersection(A,C)))\n",
    "print(\"B, C: {}\".format(myListIntersection(B,C)))\n",
    "print(\"A, D: {}\".format(myListIntersection(A,D)))\n",
    "```\n",
    "\n",
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex1');\">Show/Hide Solution</div>\n",
    "<div id=\"ex1\" style=\"display:none;\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "nbsphinx": "hidden"
   },
   "outputs": [],
   "source": [
    "%reset -f\n",
    "\n",
    "import unittest\n",
    "import random\n",
    "\n",
    "#this has problems (test X=[1,1,2,2,3], Y=[4,3])\n",
    "def myListIntersection2(X,Y):\n",
    "    tmp = X + Y\n",
    "    vals = [x for x in tmp if tmp.count(x) == 2]\n",
    "    return list(set(vals))\n",
    "\n",
    "##Correct!\n",
    "def myListIntersection(X,Y):\n",
    "    inx = [ x for x in X if x in Y]\n",
    "    iny = [y for y in Y if y in X]\n",
    "    return list(set(inx + iny))\n",
    "    \n",
    "A = [1, 2, 3, 4, 7, 12]\n",
    "B = [4, 1, 7, 120]\n",
    "C = [120, 6]\n",
    "D = []\n",
    "\n",
    "\n",
    "class Test(unittest.TestCase):\n",
    "    def __init__(self, *args, **kwargs):\n",
    "        super(Test, self).__init__(*args, **kwargs)\n",
    "        #create two test lists:\n",
    "        self.x = []\n",
    "        self.y = []\n",
    "        for i in range(15):    \n",
    "            self.x.append(random.randint(0,10))\n",
    "            self.y.append(random.randint(0,10))\n",
    "        #print(\"{}\\n{}\".format(self.x,self.y))\n",
    "                          \n",
    "    def test_reslen(self):\n",
    "        r = myListIntersection(self.x, self.y)\n",
    "        self.assertFalse(len(r) > len(self.x))\n",
    "        self.assertFalse(len(r) > len(self.y))\n",
    "        s1 = [a for a in self.x if a in self.y]\n",
    "        s2 = [a for a in self.y if a in self.x]\n",
    "        S = set(s1+s2)\n",
    "        self.assertTrue(len(r) == len(S))\n",
    "    \n",
    "    def test_empty(self):\n",
    "            self.assertEqual(myListIntersection(self.x, []),[])\n",
    "            self.assertEqual(myListIntersection([], self.y),[]) \n",
    "    \n",
    "    def test_transitivity(self):\n",
    "            v = myListIntersection(self.x, self.y).sort()\n",
    "            v1 = myListIntersection(self.y, self.x).sort()              \n",
    "            self.assertEqual(v,v1)\n",
    "    \n",
    "    def test_doubleEls(self):\n",
    "        dX = self.x + self.x\n",
    "        dY = self.y + self.y\n",
    "        v1 = myListIntersection(dX,dY)\n",
    "        v1.sort()\n",
    "        v2 = myListIntersection(self.x,self.y)\n",
    "        v2.sort()\n",
    "        self.assertEqual(v1, v2)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    print(\"A, B: {}\".format(myListIntersection(A,B)))\n",
    "    print(\"A, C: {}\".format(myListIntersection(A,C)))\n",
    "    print(\"B, C: {}\".format(myListIntersection(B,C)))\n",
    "    print(\"A, D: {}\".format(myListIntersection(A,D)))   \n",
    "    unittest.main()\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-info\">\n",
    "\n",
    "**Note:** \n",
    "Note the line that I used to initialize the Test class. \n",
    "```\n",
    "def __init__(self, *args, **kwargs):\n",
    "        super(Test, self).__init__(*args, **kwargs)\n",
    "```\n",
    "this allows us to define the random test data within the Test class (these lines are basically because we need to call the super-class constructor with all the parameters it needs). \n",
    "\n",
    "</div>\n",
    "\n",
    "\n",
    "You can find a solution to run unittests here: [pract13_ex1.py](file_samples/pract13_ex1.py)\n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2. CRISPR-Cas9 is quite a neat system to perform genome editing. Guide RNAs (gRNAs) can transport Cas9 to anywhere in the genome for gene editing, but no editing can occur at any site other than one at which Cas9 recognizes the protospacer adjacent motif (PAM). The PAM site is a 2-6 base pair DNA sequence immediately following the DNA sequence targeted by the Cas9 nuclease in the CRISPR bacterial adaptive immune system. Some used PAMs are the following:\n",
    "```\n",
    "NGG (where N is any base)\n",
    "NGA\n",
    "YG (where Y is a Pyrimidine, i.e. C or T)\n",
    "TTTN\n",
    "YTN\n",
    "```\n",
    "write a function that loads the fasta sequences [contig82.fasta](file_samples/contig82.fasta) and for each sequence reports the number of sites of each of the above PAMs and its frequency (i.e. number over the length of the sequence). Hint: load the sequence with biopython and SeqIO.\n",
    "\n",
    "\n",
    "\n",
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex2');\">Show/Hide Solution</div>\n",
    "<div id=\"ex2\" style=\"display:none;\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "nbsphinx": "hidden"
   },
   "outputs": [],
   "source": [
    "%reset -f\n",
    "\n",
    "from Bio import SeqIO\n",
    "import re\n",
    "\n",
    "def countPAMs(filename):\n",
    "    for seq_record in SeqIO.parse(filename, \"fasta\"):\n",
    "        s = seq_record.seq\n",
    "        l = len(s)\n",
    "        ident = seq_record.id\n",
    "        m = re.findall(\"[ATCG]GG\", str(s))\n",
    "        NGG = len(m)\n",
    "        m = re.findall(\"[ATCG]GA\", str(s))\n",
    "        NGA = len(m)\n",
    "        m = re.findall(\"[C|T]G\", str(s))\n",
    "        YG = len(m)\n",
    "        m = re.findall(\"TTT[ATCG]\", str(s))\n",
    "        TTTN = len(m)\n",
    "        m = re.findall(\"[C|T]T[ATCG]\", str(s))\n",
    "        YTN = len(m)\n",
    "        print(\"{} (len:{}):\\n\\tNGG:{} (1 on {} bases)\".format(ident, l,NGG, l/NGG))\n",
    "        print(\"\\tNGA:{} (1 on {} bases)\".format(NGA, l/NGA))\n",
    "        print(\"\\tYG:{} (1 on {} bases)\".format(YG, l/YG))\n",
    "        print(\"\\tTTTN:{} (1 on {} bases)\".format( TTTN, l/TTTN))\n",
    "        print(\"\\tYTN:{} (1 on {} bases)\".format( YTN, l/YTN))\n",
    "        \n",
    "        \n",
    "fn = \"file_samples/contigs82.fasta\"\n",
    "countPAMs(fn)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3. Write a python function ```sortCSV(mystr)``` that gets a comma separated string and returns a comma separated string with all elements sorted in alphabetically decreasing order. Define some unittests to check if the function has been implemented correctly (some things to check: length of initial string is the same as that of the final string, number of elements is the same, each element in the output string must come after the next one in lexicographical order,...). \n",
    "\n",
    "\n",
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex3');\">Show/Hide Solution</div>\n",
    "<div id=\"ex3\" style=\"display:none;\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "nbsphinx": "hidden"
   },
   "outputs": [],
   "source": [
    "%reset -f\n",
    "\n",
    "import random\n",
    "import unittest\n",
    "\n",
    "def sortCSV(mystr):\n",
    "    tmp = mystr.split(\",\")\n",
    "    tmp.sort(reverse=True)\n",
    "    return \",\".join(tmp)\n",
    "\n",
    "\n",
    "\n",
    "class Testing(unittest.TestCase):\n",
    "    def __init__(self, *args, **kwargs):\n",
    "        super(Testing, self).__init__(*args, **kwargs)\n",
    "        #create a random string\n",
    "        self.alphabet = \"abcdefghkjilmnopqrstuvwyz\"\n",
    "        self.data = \"\"\n",
    "        #create 15 random strings\n",
    "        for i in range(15):\n",
    "            word = \"\"\n",
    "            #each of them has a random length up to 20\n",
    "            j = random.randint(1,20)\n",
    "            for ind in range(j):\n",
    "                #pick up to 20 random letters\n",
    "                t = random.randint(1,len(self.alphabet)-1)\n",
    "                word += self.alphabet[t]\n",
    "            if(len(self.data) == 0):\n",
    "                self.data = word\n",
    "            else:\n",
    "                self.data += \",\" + word\n",
    "\n",
    "\n",
    "    def test_reslen(self):\n",
    "        self.assertTrue(len(self.data) == len(sortCSV(self.data)))\n",
    "\n",
    "    def test_elcount(self):\n",
    "        res = sortCSV(self.data).split(\",\")\n",
    "        self.assertTrue(len(self.data.split(\",\")) == len(res))\n",
    "\n",
    "    def test_elsorting(self):\n",
    "        res = sortCSV(self.data).split(\",\")\n",
    "        for ind in range(len(res)-1):\n",
    "            self.assertTrue(res[ind]> res[ind+1])\n",
    "\n",
    "    def test_empty(self):\n",
    "        self.assertEqual(sortCSV(\"\"),\"\")\n",
    "    \n",
    "    def test_onlyOne(self):\n",
    "        j = random.randint(1,20)\n",
    "        word = \"\"\n",
    "        for ind in range(j):\n",
    "            #pick up to 20 random letters\n",
    "            t = random.randint(0,len(self.alphabet)-1)\n",
    "            word += self.alphabet[t]\n",
    "        self.assertEqual(sortCSV(word), word)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    mystr = \"book,tree,final,example,testing,zed,all,hair,lady,figure,tap,spring,test,fin,tail\"\n",
    "    print(\"Original:\\n{}\".format(mystr))\n",
    "    print(\"Sorted:\\n{}\".format(sortCSV(mystr)))\n",
    "    unittest.main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "You can find a solution to run unittests here: [pract13_ex3.py](file_samples/pract13_ex3.py)\n",
    "\n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4. Solve a modified version of the following exercise of Practical 11 implementing a function that parses the \"ExpXml\" text through several regular expressions.\n",
    "\n",
    "The exercise is reported below:\n",
    "\n",
    "Write a python script that retrieves all the information present in SRA regarding PacBio sequencing performed on E.coli strain K12 (query term is “E.coli K12 wgs PacBio”). \n",
    "\n",
    "Print the number of results and for each id report the title, the experiment accession, the instrument, the library strategy, the library source, the total number of spots and total number of bases sequenced.\n",
    "Sample output:\n",
    "\n",
    "```\n",
    "Entries found: 9\n",
    "\n",
    "[1] Results for id 357838:\n",
    "Title: E. coli K12 PacBio RS C2 CCS sequencing\n",
    "Experiment accession: SRX255779\n",
    "Instrument: PACBIO_SMRT\n",
    "Library strategy: WGS\n",
    "Library source: GENOMIC\n",
    "Total spots:1798302\n",
    "Total bases:4228754616\n",
    "\n",
    " ...\n",
    " ...\n",
    "```\n",
    "\n",
    "A sample \"ExpXml\" string:\n",
    "\n",
    "```\n",
    "<Summary><Title>E. coli K12 PacBio RS C2 CCS sequencing</Title><Platform instrument_model=\"PacBio RS\">PACBIO_SMRT</Platform><Statistics total_runs=\"22\" total_spots=\"1798302\" total_bases=\"4228754616\" total_size=\"16799546700\" load_done=\"true\" cluster_name=\"public\"/></Summary><Submitter acc=\"SRA071585\" center_name=\"NBACC\" contact_name=\"Sergey Koren\" lab_name=\"\"/><Experiment acc=\"SRX255779\" ver=\"1\" status=\"public\" name=\"E. coli K12 PacBio RS C2 CCS sequencing\"/><Study acc=\"SRP020003\" name=\"Escherichia coli K12 Re-sequencing\"/><Organism taxid=\"511145\" ScientificName=\"Escherichia coli str. K-12 substr. MG1655\"/><Sample acc=\"SRS000462\" name=\"\"/><Instrument PACBIO_SMRT=\"PacBio RS\"/><Library_descriptor><LIBRARY_NAME>PacBio RS CCS</LIBRARY_NAME><LIBRARY_STRATEGY>WGS</LIBRARY_STRATEGY><LIBRARY_SOURCE>GENOMIC</LIBRARY_SOURCE><LIBRARY_SELECTION>unspecified</LIBRARY_SELECTION><LIBRARY_LAYOUT> <SINGLE/> </LIBRARY_LAYOUT></Library_descriptor><Bioproject>PRJNA194437</Bioproject><Biosample>SAMN00000224</Biosample>\n",
    "```\n",
    "\n",
    "\n",
    "\n",
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex4');\">Show/Hide Solution</div>\n",
    "<div id=\"ex4\" style=\"display:none;\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "nbsphinx": "hidden"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Entries found: 9\n",
      "\n",
      "[1] Results for id 357838:\n",
      "Title: E. coli K12 PacBio RS C2 CCS sequencing\n",
      "Experiment accession: SRX255779\n",
      "Instrument: PACBIO_SMRT\n",
      "Library strategy: WGS\n",
      "Library source: GENOMIC\n",
      "Total spots:1798302\n",
      "Total bases:4228754616\n",
      "\n",
      "[2] Results for id 357018:\n",
      "Title: E. coli K12 MG1655 PacBio RS C2 Sequencing\n",
      "Experiment accession: SRX255228\n",
      "Instrument: PACBIO_SMRT\n",
      "Library strategy: WGS\n",
      "Library source: GENOMIC\n",
      "Total spots:1389597\n",
      "Total bases:3032527263\n",
      "\n",
      "[3] Results for id 357016:\n",
      "Title: E. Coli K12 454 Sequencing To Use For PacBio RS Correction\n",
      "Experiment accession: SRX255226\n",
      "Instrument: LS454\n",
      "Library strategy: WGS\n",
      "Library source: GENOMIC\n",
      "Total spots:1385764\n",
      "Total bases:720878709\n",
      "\n",
      "[4] Results for id 285700:\n",
      "Title: WGA E. coli genome DNA sequenced by C2 chemistry (Control)\n",
      "Experiment accession: SRX209661\n",
      "Instrument: PACBIO_SMRT\n",
      "Library strategy: WGS\n",
      "Library source: GENOMIC\n",
      "Total spots:245223\n",
      "Total bases:373483301\n",
      "\n",
      "[5] Results for id 285699:\n",
      "Title: WGA E. coli genome DNA sequenced by C2 chemistry (Null)\n",
      "Experiment accession: SRX209660\n",
      "Instrument: PACBIO_SMRT\n",
      "Library strategy: WGS\n",
      "Library source: GENOMIC\n",
      "Total spots:245223\n",
      "Total bases:432837815\n",
      "\n",
      "[6] Results for id 285698:\n",
      "Title: Native E. coli genome DNA sequenced by C2 chemistry\n",
      "Experiment accession: SRX209659\n",
      "Instrument: PACBIO_SMRT\n",
      "Library strategy: WGS\n",
      "Library source: GENOMIC\n",
      "Total spots:245223\n",
      "Total bases:340971357\n",
      "\n",
      "[7] Results for id 285697:\n",
      "Title: WGA E. coli genome DNA sequenced by FCR chemistry\n",
      "Experiment accession: SRX209658\n",
      "Instrument: PACBIO_SMRT\n",
      "Library strategy: WGS\n",
      "Library source: GENOMIC\n",
      "Total spots:2917014\n",
      "Total bases:2838893381\n",
      "\n",
      "[8] Results for id 285673:\n",
      "Title: WGA E. coli Plasmid DNA with m4C\n",
      "Experiment accession: SRX209634\n",
      "Instrument: PACBIO_SMRT\n",
      "Library strategy: WGS\n",
      "Library source: GENOMIC\n",
      "Total spots:121922\n",
      "Total bases:183635674\n",
      "\n",
      "[9] Results for id 285672:\n",
      "Title: Native E. coli plasmid DNA with m4C\n",
      "Experiment accession: SRX209633\n",
      "Instrument: PACBIO_SMRT\n",
      "Library strategy: WGS\n",
      "Library source: GENOMIC\n",
      "Total spots:60961\n",
      "Total bases:168647652\n"
     ]
    }
   ],
   "source": [
    "%reset -f \n",
    "\n",
    "import re\n",
    "from Bio import Entrez\n",
    "\n",
    "\n",
    "\n",
    "def parseExp(expStr):\n",
    "    m = re.search(\"<Title>([A-Za-z0-9_\\. \\(\\)]*)</Title>\", expStr)\n",
    "    if(m):\n",
    "        title = \"Title: \" + m.group()[7:-8]\n",
    "        print(title)\n",
    "        \n",
    "    m = re.search(\"<Experiment acc=\\\"([A-Z0-9]*)\\\"\", expStr)\n",
    "    if(m):\n",
    "        acc = \"Experiment accession: \" + m.groups()[0]\n",
    "        print(acc)\n",
    "    m = re.search(\"<Platform ([A-Za-z0-9_=\\\" \\(\\)]*)>([A-Za-z0-9_\\(\\)]*)</Platform>\", expStr)\n",
    "    if(m):\n",
    "        platform = \"Instrument: \" + m.groups()[1]\n",
    "        print(platform)\n",
    "    m = re.search(\"<LIBRARY_STRATEGY>([A-Za-z0-9_=\\\" \\(\\)]*)</LIBRARY_STRATEGY>\", expStr)\n",
    "    if(m):\n",
    "        src = \"Library strategy: \"  + m.groups()[0]\n",
    "        print(src)\n",
    "    m = re.search(\"<LIBRARY_SOURCE>([A-Za-z0-9_=\\\" \\(\\)]*)</LIBRARY_SOURCE>\", expStr)\n",
    "    if(m):\n",
    "        src = \"Library source: \"  + m.groups()[0]    \n",
    "        print(src)\n",
    "    m = re.search(\"total_spots=\\\"([0-9]*)\\\" total_bases=\\\"([0-9]*)\\\"\",expStr)\n",
    "    if(m):\n",
    "        spots = \"Total spots:\" + m.groups()[0] +\"\\nTotal bases:\" + m.groups()[1]\n",
    "        print(spots)\n",
    "\n",
    "    \n",
    "Entrez.email = \"my_email\"\n",
    "handle = Entrez.esearch(db=\"sra\", term=\"E.coli K12 wgs PacBio\", retmax = 10)\n",
    "res = Entrez.read(handle)\n",
    "#uncomment to see all fields:\n",
    "#for el in res.keys():\n",
    "#    print(el , \" : \", res[el])\n",
    "\n",
    "print(\"Entries found: {}\".format(res[\"Count\"]))\n",
    "\n",
    "cnt = 1\n",
    "for ids in res[\"IdList\"]:\n",
    "    print(\"\\n[{}] Results for id {}:\".format(cnt, ids))\n",
    "    handle = Entrez.esummary(db=\"sra\",  id = ids)\n",
    "    res = Entrez.read(handle)\n",
    "    cnt += 1\n",
    "          \n",
    "    for r in res:\n",
    "        info = r['ExpXml']\n",
    "        #print(info)\n",
    "        parseExp(info)\n",
    "        \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "5. The file (DNA_seq.fasta)[file_samples/DNA_seq.fasta] contains a synthetic DNA sequence. Let's assume to have two restriction enzymes LagI and JagII that respectively cut at the site CNC/ATT, and GAGRK/TNG (where N is any site, R is A or G and K is A or C or T. Note that \"/\" is just a representation of where the enzyme cuts, therefore we do not need to specify this in the regular expression, but we need to take it into account when we cut the DNA.\n",
    "\n",
    "Ex. if the sequence is:\n",
    "ATACATTCCCCCGGAATCGCCCCCCCTCCATTCC\n",
    "digesting the sequence with LagI would give:\n",
    "[\"ATAC\",\"ATTCCCCCGGAATCGCCCCCCCTCC\", \"ATTCC\"]\n",
    "digesting this further with JagII would give:\n",
    "[\"ATAC\",\"ATTCCCCCGGAA\", \"TCGCCCCCCCTCC\", \"ATTCC\"]\n",
    "\n",
    "Write a python script that simulates a digestion with LagI, size selection to keep only the fragments higher than 50 base pairs, and a digestion with JagII, printing the lengths of the obtained fragments. What happens to the fragments if we digest first with JagII and then with LagI?\n",
    "\n",
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex5');\">Show/Hide Solution</div>\n",
    "<div id=\"ex5\" style=\"display:none;\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "nbsphinx": "hidden"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initial sequence:\n",
      "CTTACATGGCAATAACCCCCCGTTTCTACTTCTAGAGGAGAAAAGTATTGACATGAGCGCTCCCGGCACAAGGGCCAAAGAAGTCTCCAATTTCTTATTTCCGAATGACATGCGTCTCCTTGCGGGTAAATCACCGACCGCAATTCATAGAAGCCTGGGGGAACAGATAGGTCTAATTAGCTTAAGAGAGTAAATCCTGGGATCATTCAGTAGTAACCATAAACTTACGCTGGGGCTTCTTCGGCGGATTTTTACAGTTACCAACCAGGAGATTTGAAGTAAATCAGTTGAGGATTTAGCCGCGCTATCCGGTAATCTCCAAATTAAAACATACCGTTCCATGAAGGCTAGAATTACTTACCGGCCTTTTCCATGCCTGCGCTATACCCCCCCACTCTCCCGCTTATCCGTCCGAGCGGAGGCAGTGCGATCCTCCGTTAAGATATTCTTACGTGTGACGTAGCTATGTATTTTGCAGAGCTGGCGAACGCGTTGAACACTTCACAGATGGTAGGGATTCGGGTAAAGGGCGTATAATTGGGGACTAACATAGGCGTAGACTACGATGGCGCCAACTCAATCGCAGCTCGAGCGCCCTGAATAACGTACTCATCTCAACTCATTCTCGGCAATCTACCGAGCGACTCGATTATCAACGGCTGTCTAGCAGTTCTAATCTTTTGCCAGCATCGTAATAGCCTCCAAGAGATTGATGATAGCTATCGGCACAGAACTGAGACGGCGCCGATGGATAGCGGACTTTCGGTCAACCACAATTCCCCACGGGACAGGTCCTGCGGTGCGCATCACTCTGAATGTACAAGCAACCCAAGTGGGCCGAGCCTGGACTCAGCTGGTTCCTGCGTGAGCTCGAGACTCGGGATGACAGCTCTTTAAACATAGAGCGGGGGCGTCGAACGGTCGAGAAAGTCATAGTACCTCGGGTACCAACTTACTCAGGTTATTGCTTGAAGCTGTACTATTTTAGGGGGGGAGCGCTGAAGGTCTCTTCTTCTCATGACTGAACTCGCGAGGGTCGTGAAGTCGGTTCCTTCAATGGTTAAAAAACAAAGGCTTACTGTGCGCAGAGGAACGCCCATCTAGCGGCTGGCGTCTTGAATGCTCGGTCCCCTTTGTCATTCCGGATTAATCCATTTCCCTCATTCACGAGCTTGCGAAGTCTACATTGGTATATGAATGCGACCTAGAAGAGGGCGCTTAAAATTGGCAGTGGTTGATGCTCTAAACTCCATTTGGTTTACTCGTGCATCACCGCGATAGGCTGACAAAGGTTTAACATTGAATAGCAAGGCACTTCCGGTCTCAATGAACGGCCGGGAAAGGTACGCGCGCGGTATGGGAGGATCAAGGGGCCAATAGAGAGGCTCCTCTCTCACTCGCTAGGAGGCAAATGTAAAACAATGGTTACTGCATCGATACATAAAACATGTCCATCGGTTGCCCAAAGTGTTAAGTGTCTATCACCCCTAGGGCCGTTTCCCGCATATAAACGCCAGGTTGTATCCGCATTTGATGCTACCGTGGATGAGTCTGCGTCGAGCGCGCCGCACGAATGTTGCAATGTATTGCATGAGTAGGGTTGACTAAGAGCCGTTAGATGCGTCGCTGTACTAATAGTTGTCGACAGACCGTCGAGATTAGAAAATGGTACCAGCATTTTCGGAGGTTCTCTAACTAGTATGGATTGCGGTGTCTTCACTGTGCTGCGGCTACCCATCGCCTGAAATCCAGCTGGTGTCAAGCCATCCCCTCTCCGGGACGCCGCATGTAGTGAAACATATACGTTGCACGGGTTCACCGCGGTCCGTTCTGAGTCGACCAAGGACACAATCGAGCTCCGATCCGTACCCTCGACAAACTTGTACCCGACCCCCGGAGCTTGCCAGCTCCTCGGGTATCATGGAGCCTGTGGTTCATCGCGTCCGATATCAAACTTCGTCATGATAAAGTCCCCCCCTCGGGAGTACCAGAGAAGATGACTACTGAGTTGTGCGATAAACGGATAAA\n",
      "\n",
      "\n",
      "LagI restriction:\n",
      "\tRestriction sites:\n",
      "\t618 624 CTCATT\n",
      "\tRestriction sites:\n",
      "\t1159 1165 CTCATT\n",
      "\tRestriction sites:\n",
      "\t1525 1531 CGCATT\n",
      "Lengths:\n",
      "[621, 541, 366, 500]\n",
      "\n",
      "JagII restriction:\n",
      "\tRestriction sites:\n",
      "\t268 276 GAGATTTG\n",
      "\tRestriction sites:\n",
      "\t251 259 GAGACTCG\n",
      "\tRestriction sites:\n",
      "\t126 134 GAGATTAG\n",
      "Final lengths: [273, 348, 256, 285, 366, 131, 369]\n",
      "\n",
      "###########################\n",
      "##### JagII first #########\n",
      "###########################\n",
      "LagI restriction:\n",
      "\tRestriction sites:\n",
      "\t268 276 GAGATTTG\n",
      "\tRestriction sites:\n",
      "\t872 880 GAGACTCG\n",
      "\tRestriction sites:\n",
      "\t1654 1662 GAGATTAG\n",
      "Lengths:\n",
      "[273, 604, 782, 369]\n",
      "\n",
      "JagII restriction:\n",
      "\tRestriction sites:\n",
      "\t345 351 CTCATT\n",
      "\tRestriction sites:\n",
      "\t282 288 CTCATT\n",
      "\tRestriction sites:\n",
      "\t648 654 CGCATT\n",
      "Final lengths: [273, 348, 256, 285, 366, 131, 369]\n"
     ]
    }
   ],
   "source": [
    "%reset -f \n",
    "\n",
    "from Bio import SeqIO\n",
    "import re\n",
    "\n",
    "#in our case overhang can be 3 or 4 (3 for LagI and 4 for JagII)\n",
    "def digestSequence(seq, regex, overhang):\n",
    "    digests = []\n",
    "    sP = 0\n",
    "    matches = re.finditer(regex,seq)\n",
    "    if matches:\n",
    "        for site in matches:\n",
    "            print(\"\\tRestriction sites:\")\n",
    "            print(\"\\t{} {} {}\".format(site.start(),\n",
    "                                   site.end(),\n",
    "                                   site.group()))\n",
    "            digests.append(seq[sP:site.start()+overhang])\n",
    "            sP = site.start() + overhang\n",
    "        #last element:\n",
    "        digests.append(seq[sP:])\n",
    "    return digests\n",
    "\n",
    "fn = \"file_samples/DNA_seq.fasta\"\n",
    "\n",
    "myseq = SeqIO.read(fn, \"fasta\")\n",
    "\n",
    "regexLagI = \"C[ATCG]CATT\"\n",
    "regexJagII = \"GAG[AG][ACT]T[ATCG]G\"\n",
    "\n",
    "\n",
    "s = str(myseq.seq)\n",
    "\n",
    "print(\"Initial sequence:\")\n",
    "print(s)\n",
    "print(\"\\n\")\n",
    "\n",
    "        \n",
    "print(\"LagI restriction:\")\n",
    "digests = digestSequence(s, regexLagI, 3)\n",
    "#filter digests\n",
    "dig = [x for x in digests if len(x) > 50]\n",
    "print(\"Lengths:\")\n",
    "print([len(x) for x in dig])\n",
    "finalDigests = []\n",
    "\n",
    "print(\"\\nJagII restriction:\")\n",
    "for d in dig:\n",
    "    tmp = digestSequence(d, regexJagII, 5)\n",
    "    for t in tmp:\n",
    "        finalDigests.append(t)\n",
    "\n",
    "#print(\"Final digests:\" + str(finalDigests))\n",
    "print(\"Final lengths: \" + str([len(x) for x in finalDigests])) \n",
    "\n",
    "\n",
    "#Other JagII first and LagI second:\n",
    "print(\"\\n###########################\")\n",
    "print(\"##### JagII first #########\")\n",
    "print(\"###########################\")\n",
    "print(\"LagI restriction:\")\n",
    "digests = digestSequence(s, regexJagII, 5)\n",
    "#filter digests\n",
    "dig = [x for x in digests if len(x) > 50]\n",
    "print(\"Lengths:\")\n",
    "print([len(x) for x in dig])\n",
    "finalDigests = []\n",
    "\n",
    "print(\"\\nJagII restriction:\")\n",
    "for d in dig:\n",
    "    tmp = digestSequence(d, regexLagI, 3)\n",
    "    for t in tmp:\n",
    "        finalDigests.append(t)\n",
    "\n",
    "#print(\"Final digests:\" + str(finalDigests))\n",
    "print(\"Final lengths: \" + str([len(x) for x in finalDigests])) \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "</div>"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Edit Metadata",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
