{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "nbsphinx": "hidden"
   },
   "source": [
    "### NOTE FOR LUCA\n",
    "\n",
    "**Remember to set/remove metadata as:**\n",
    "{\n",
    "  \"nbsphinx\": \"hidden\"\n",
    "}\n",
    "\n",
    "to enable/disable solutions view"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# Practical 13\n",
    "\n",
    "In this practical we will continue with object oriented programming and will focus on code testing in Python. Finally, we will introduce regular expressions.\n",
    "\n",
    "## Slides\n",
    "\n",
    "The slides of the introduction can be found here: [Intro](docs/Practical13.pdf)\n",
    "\n",
    "## Testing your code \n",
    "\n",
    "Testing the code is quite an important step to make sure that the code is predictable and although some bugs can always slip through, **testing is the process of making the code as predictable as possible**.\n",
    "\n",
    "Two types of testing exist: the **white-box** testing methodology looks into every detail of the implemented code to inspect its correctness, **black-box** testing does not look at the details of how the code is implemented, but it just focus on the correctness of the output produced by the code.\n",
    "\n",
    "Testing is quite a complex and articulated topic, here we will just scratch the tip of the iceberg (there are many books on the topic if you are interested).\n",
    "\n",
    "A **test** is a piece of code written with the sole purpose of checking the correctness of another piece of code. \n",
    "\n",
    "Testing requires three successive moments: first of all we need to **set up** (or prepare) the test setting up connections/interfaces to test data, the second step is to **execute** the piece of code we are aiming to test using the interfaces devised at the previous step, and the third is the **verification** of the results to make sure they look as they were expected to.\n",
    "\n",
    "### Doctest \n",
    "\n",
    "A very simple way to specifying tests for the code is by using an embedded module called **doctest**. It will basically search for pieces of code in your python file that look like **interactive python sessions** (that are lines starting with ```>>> ```) and will execute them to check if they run giving the result specified in the next line.\n",
    "\n",
    "<div class=\"alert alert-warning\">\n",
    "\n",
    "**Important:** \n",
    "\n",
    "Note the space after \">>> \". That is where the test starts. An example:\n",
    "```\n",
    "\"\"\"\n",
    "This is a function that returns three values in a list...\n",
    ">>> fun(x)\n",
    "[x, y, z]\n",
    "\"\"\"\n",
    "```\n",
    "</div>\n",
    "\n",
    "**Example:**\n",
    "Let's define some doctest tests for the simple function computing the first N prime numbers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%reset -f \n",
    "\n",
    "def getFirstNprimes(N):\n",
    "    '''\n",
    "    This function should output the first N prime numbers.\n",
    "    >>> getFirstNprimes(1)\n",
    "    [2]\n",
    "    >>> getFirstNprimes(2)\n",
    "    [2, 3]\n",
    "    >>> getFirstNprimes(10)\n",
    "    [2, 3, 5, 7, 11, 13, 17, 19, 23, 29]\n",
    "    '''\n",
    "    if N == 0:\n",
    "        return []\n",
    "    res = [2]\n",
    "    current = 3\n",
    "    while len(res) < N:\n",
    "        if len([x for x in res if current % x == 0]) == 0:\n",
    "            res.append(current)\n",
    "        current += 1\n",
    "    #uncomment next line to introduce a bug\n",
    "    #res.append(1)\n",
    "    return res        \n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    import doctest\n",
    "    doctest.testmod()\n",
    "    print(getFirstNprimes(20))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The line ```if __name__ == \"__main__\":``` is used to specify if the code is executed as a script (i.e. it is not invoked as an imported module somewhere else in another piece of code).\n",
    "\n",
    "Executing the code as it is will not give any particular error as the tests we set-up passed correctly. But if we uncomment the ```res.append(1)``` the tests will fail and we produce the following testing output, which reports the three failed tests, the expected values and the obtained values: \n",
    "\n",
    "![](img/pract13/doctest.png)\n",
    "\n",
    "Another way of testing the code is through unit tests, we will see this later on. \n",
    "\n",
    "### Raising exceptions and using assertions\n",
    "\n",
    "Exceptions are a very good way to inform the program that some unexpected thing has happened (e.g. division by zero, tentative to access a position in a list that does not exist, summing a string and an integer...).\n",
    "\n",
    "One thing we can do is to raise exceptions whenever some pre-conditions are not met in order to insure that these do not lead to erroneous behaviours. This can be done with ```raise Exception(\"exception text\")```. More info on raising exceptions are [here](https://docs.python.org/3/tutorial/errors.html#raising-exceptions). \n",
    "\n",
    "**Example**: \n",
    "Consider the following ```MyIntPair``` class that works with integers. If we want to make sure it only contains integers we can add a ```raise Exception``` in case it is not an integer. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class MyIntPair:\n",
    "    def __init__(self, x,y):\n",
    "        if not type(x) == int:\n",
    "            raise Exception(\"x {} is not integer\".format(x))\n",
    "        if not type(y) == int:\n",
    "            raise Exception(\"y {} is not integer\".format(y))\n",
    "        self.x = x\n",
    "        self.y = y\n",
    "        \n",
    "    def __add__(self, other):\n",
    "        return (self.x + other.x, self.y + other.y)\n",
    "    \n",
    "A = MyIntPair(5,10)\n",
    "B = MyIntPair(3,6)\n",
    "print(A + B)\n",
    "C = MyIntPair(1, \"two\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that if we try to pass it something other than an integer, the execution will stop. To amend this behaviour we need to intercept the exception and deal with it appropriately. This topic is quite articulated, you can find more info on it [here](https://docs.python.org/3/tutorial/errors.html#handling-exceptions).\n",
    "\n",
    "A simple exception handling can be done by using the ```try - except``` construct that tries to run some statements, being ready to intercept and handle any exception that might occur:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class MyIntPair:\n",
    "    def __init__(self, x,y):\n",
    "        if not type(x) == int:\n",
    "            raise Exception(\"x: {} is not integer\".format(x))\n",
    "        if not type(y) == int:\n",
    "            raise Exception(\"y: {} is not integer\".format(y))\n",
    "        self.x = x\n",
    "        self.y = y\n",
    "        \n",
    "    def __add__(self, other):\n",
    "        return (self.x + other.x, self.y + other.y)\n",
    "    \n",
    "try: \n",
    "    A = MyIntPair(5,10)\n",
    "    B = MyIntPair(3,6)\n",
    "    #Uncomment to see a different error\n",
    "    #print(A/0)\n",
    "    print(A + B)\n",
    "    C = MyIntPair(1, \"two\")\n",
    "    print(A + C)\n",
    "\n",
    "except Exception as e:\n",
    "    print(\"Whoops something went wrong. Ignore the rest.\")\n",
    "    print(str(e))\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Similarly, we can produce assertions that test if a condition is True, otherwise the execution is stopped with an ```AssertionError```. This error can be caught and handled appropriately. Note that in the code below, only the AssertionError is caught and a division by 0 (e.g. print(10/0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Whoops something went wrong. Ignore the rest.\n",
      "division by zero\n"
     ]
    }
   ],
   "source": [
    "class MyIntPair:\n",
    "    def __init__(self, x,y):\n",
    "        assert type(x) == int, \"x: {} is not integer\".format(x)\n",
    "        assert type(y) == int, \"y: {} is not integer\".format(y)\n",
    "        self.x = x\n",
    "        self.y = y\n",
    "        \n",
    "        \n",
    "    def __add__(self, other):\n",
    "        return (self.x + other.x, self.y + other.y)\n",
    "    \n",
    "try: \n",
    "    A = MyIntPair(5,10)\n",
    "    B = MyIntPair(3,6)\n",
    "    #Uncomment to see a different error (not captured!)\n",
    "    #print(10/0)\n",
    "    print(A + B)\n",
    "    C = MyIntPair(1, \"two\")\n",
    "    print(A + C)\n",
    "\n",
    "except AssertionError as e:\n",
    "    print(\"Whoops something went wrong. Ignore the rest.\")\n",
    "    print(str(e))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Unit tests\n",
    "\n",
    "Python comes with a fully-fledgeted testing module which is called ```unittest```. Have a look [here](https://docs.python.org/3.8/library/unittest.html) for detailed information on this module.\n",
    "\n",
    "The module ```unittest``` must be imported first with ```import unittest``` and then the Test class must be implemented to perform the tests.\n",
    "\n",
    "In a nutshell, to create some unit tests we need to define a ```Testing``` class (we are free to call it as we like) which is a subclass of the class ```unittest.TestCase```. Within this class, we have then to specify the tests we want to run. Every test is a method and its name **must start** with ```test_``` (e.g. ```test_length```).\n",
    "\n",
    "Tests can use assertions ```assertEqual(value1, value2)```, ```assertTrue(condition)``` or ```assertFalse(condition)``` that respectively allow to check the equality of two values (i.e. the known result and the output of the method to be tested) and the truth value of a condition (typically computed on the output of the method under test).\n",
    "\n",
    "We can add the invocation to the ```unittest``` module within the script by adding (```unittest.main()```) for example in the ```main``` part through ```if __name__ == \"__main__\":```. In this case, we can perform the tests just by calling something like:\n",
    "```\n",
    "python3 my_testing_function.py\n",
    "```\n",
    "\n",
    "without specifying ```unittest.main()``` in the script, we need to call the unittest with:\n",
    "```\n",
    "python3 -m unittest my_testing_function.py\n",
    "```\n",
    "\n",
    "The unittest will provide us feedback on how the tests performed. In particular, if all the tests are passed we should get something like:\n",
    "\n",
    "```\n",
    "python3 -m unittest file_samples/my_testing_function.py \n",
    "\n",
    ".....\n",
    "----------------------------------------------------------------------\n",
    "Ran 5 tests in 0.387s\n",
    "\n",
    "OK\n",
    "```\n",
    "otherwise, in case of some errors like:\n",
    "\n",
    "```\n",
    "...FF\n",
    "======================================================================\n",
    "FAIL: test_one (file_samples.my_testing_function.Testing)\n",
    "----------------------------------------------------------------------\n",
    "Traceback (most recent call last):\n",
    "  File \"/home/biancol/Google Drive/work/courses/QCBsciprolab/file_samples/my_testing_function.py\", line 33, in test_one\n",
    "    self.assertEqual(getFirstNprimes(1),[2])\n",
    "AssertionError: Lists differ: [10] != [2]\n",
    "\n",
    "First differing element 0:\n",
    "10\n",
    "2\n",
    "\n",
    "- [10]\n",
    "+ [2]\n",
    "\n",
    "======================================================================\n",
    "FAIL: test_ten (file_samples.my_testing_function.Testing)\n",
    "----------------------------------------------------------------------\n",
    "Traceback (most recent call last):\n",
    "  File \"/home/biancol/Google Drive/work/courses/QCBsciprolab/file_samples/my_testing_function.py\", line 37, in test_ten\n",
    "    [2, 3, 5, 7, 11, 13, 17, 19, 23, 29])\n",
    "AssertionError: Lists differ: [2, 3, 5, 7, 11, 13, 17, 10, 23, 29] != [2, 3, 5, 7, 11, 13, 17, 19, 23, 29]\n",
    "\n",
    "First differing element 7:\n",
    "10\n",
    "19\n",
    "\n",
    "- [2, 3, 5, 7, 11, 13, 17, 10, 23, 29]\n",
    "?                           ^\n",
    "\n",
    "+ [2, 3, 5, 7, 11, 13, 17, 19, 23, 29]\n",
    "?                           ^\n",
    "\n",
    "\n",
    "----------------------------------------------------------------------\n",
    "Ran 5 tests in 0.770s\n",
    "\n",
    "FAILED (failures=2)\n",
    "\n",
    "```\n",
    "\n",
    "**Example:**\n",
    "Let's define some doctest tests for the simple function computing the first N prime numbers. The file is available here: [my_testing_function.py](file_samples/my_testing_function.py)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%reset -f \n",
    "\"\"\"\n",
    "In file my_testing_function.py\n",
    "\"\"\"\n",
    "\n",
    "import unittest\n",
    "import random\n",
    "\n",
    "def getFirstNprimes(N):\n",
    "    '''\n",
    "    This function should output the first N prime numbers.\n",
    "     '''\n",
    "    if N <= 0:\n",
    "        return []\n",
    "    res = [2]\n",
    "    current = 3\n",
    "    while len(res) < N:\n",
    "        if len([x for x in res if current % x == 0]) == 0:\n",
    "            res.append(current)\n",
    "        current += 1\n",
    "    #uncomment next line to introduce a bug\n",
    "    #res.append(1)\n",
    "    #or a more subtle error:\n",
    "    #ind = random.randint(len(res))\n",
    "    #res[ind] = 10\n",
    "    return res        \n",
    "\n",
    "class Testing(unittest.TestCase):\n",
    "    def test_empty(self):\n",
    "        self.assertEqual(getFirstNprimes(0),[])\n",
    "    \n",
    "    def test_one(self):\n",
    "        self.assertEqual(getFirstNprimes(1),[2])\n",
    "    \n",
    "    def test_ten(self):\n",
    "        self.assertEqual(getFirstNprimes(10),\n",
    "                         [2, 3, 5, 7, 11, 13, 17, 19, 23, 29])\n",
    "    \n",
    "    def test_len(self):\n",
    "        for i in range(0,10):\n",
    "            n = random.randint(1,1000)\n",
    "            self.assertFalse(len(getFirstNprimes(n)) != n)\n",
    "    \n",
    "    \n",
    "    def test_negative(self):\n",
    "        self.assertTrue(len(getFirstNprimes(-1)) == 0)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    #uncomment to run the tests in the main (without -m)\n",
    "    #unittest.main()\n",
    "    print(getFirstNprimes(20))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Running the tests on the previous code without any \"intentional bug\" will produce an OK message as above. Note that the random error is quite tricky as it affects different positions all the time (try running the test several times). This type of errors is quite difficult to detect and requires several rounds of testing as if some random steps are taking place, the error might not occur all the times."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Regular Expressions \n",
    "\n",
    "A **regular expression** (**regex**) is a string of characters defining a search pattern with which we can carry out operations such as pattern and string matching, find/replace etc.\n",
    "\n",
    "There are two types of characters: **normal characters** (which have to match amongst themselves) and **special characters** (which are used to specify repetitions (```*, ?, +, {x,y}```), a set of elements (```[]```), negation (```[^]```), beginning (```^```) of a string, end of a string (```$```), etc.\n",
    "\n",
    "As seen in the lecture, the syntax of the most common regular expressions is the following:\n",
    "\n",
    "![](img/pract13/regex.png)\n",
    "\n",
    "With regex it is possible to filter strings depending on some patterns that could include specific words (present 0 or several times using the ```?```, ```+``` or ```*``` or from ```n``` to ```m``` occurrences ```{n,m}```) at the beginning of the string ```^``` or end of the string with  ```$```. Subsets of characters can be specified with ```[]```, negation of a subset of characters with ```[^]```.\n",
    "\n",
    "Let's see some examples:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What does the following regex match?\n",
    "\n",
    "```\n",
    "regex = \"[A-Z]__[0-9]{1,4}__[a-z:-]*__[A-Z]\"\n",
    "```\n",
    "One capital letter, two underscores, from 1 to 4 numbers, two underscores, from 0 to infinite small letters, : or -, two underscores and one capital letter. Examples:\n",
    "\n",
    "```\n",
    "B__2786__prxywflh-bityeqwdmuzbygwpjtadzbvjyzmq:prucyz-rkrgjmytczdjejsgvpn-__X\n",
    "Q__5879__rarwqytmpqa-l__M\n",
    "A__45__lwonmfel-qdbcd__X\n",
    "T__5776__lmeqq__T\n",
    "A__126____T\n",
    "\n",
    "```\n",
    "\n",
    "What does the following regex match (assuming only IUPAC ambiguous nucleotide alphabet)?\n",
    "\n",
    "```\n",
    "regex = \"[ATCG]*([^ATCG]+[ATCG]*)*\"\n",
    "```\n",
    "macthes all DNA strings with zero or more A,T,C,G bases, followed by a non A,T,C,G base (e.g. N), followed by zero or more A,T,C,G bases:\n",
    "```\n",
    "ATACAYYATATACA\n",
    "CGCTTANTAT\n",
    "AAATTCGW\n",
    "TTACWWWWWWWWA\n",
    "```\n",
    "\n",
    "What does the following regex encode for?\n",
    "\n",
    "```\n",
    "regex=\"^\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}$```\n",
    "```\n",
    "it matches:\n",
    "\n",
    "```\n",
    "61.92.71.6\n",
    "497.7.0.88\n",
    "168.9.620.755\n",
    "21.5.5.1\n",
    "6.994.06.45\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In python there is a specific regular expression module to deal with these. It is called ```re``` and must be imported with ```import re``` before using it.\n",
    "\n",
    "### Match, search, finditer and findall\n",
    "The first three methods we are going to see are *match*, *search* and *finditer*, which return a ```MatchObject```:\n",
    "\n",
    "* ```re.match(regex, str)``` where ```regex``` is a string representing the regular expression, ```str``` is the input string. This method tries to match the regex on the string starting **from the beginning** (i.e. left-to-right) of the string. It will return an ```MatchObject``` or None if no match could be found;\n",
    "\n",
    "* ```re.search(regex,str)``` searches ```regex``` in the whole string and returns a ```MatchObject``` with the first occurrence of the pattern or None if the regex could not match anything;\n",
    "\n",
    "* ```re.finditer(regex,str)```: returns an iterator to ```MatchObject``` instances over all **non-overlapping** matches for the regular expression pattern in string. The string is scanned left-to-right, and matches are returned in the order found;\n",
    "\n",
    "\n",
    "Given a ```MatchObject```, if not None, provides the following information:\n",
    "\n",
    "* ```MatchObject.group()``` : the matched string;\n",
    "\n",
    "* ```MatchObject.start()``` : the starting point of the matched string in the tested string (str);\n",
    "\n",
    "* ```MatchObject.end()``` : the ending point of the matched string in the tested string (str); \n",
    "\n",
    "* ```MatchObject.groups()``` : when defining a regular expression, we can define subgroups by using \"()\". This method returns a tuple containing all the subgroups.\n",
    "\n",
    "Another method available is:\n",
    "\n",
    "```re.findall(regex,str)``` returns a **list** with all the occurrences of the regex in the string. Note that **if a group is specified** (with \"()\"), this method reports only the group.\n",
    "\n",
    "Let's see some examples of match:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hi there, i am using Python and learning hOw to UsE Regular Expressions AKA REGEX\n",
      "123 is NOT in myStr\n",
      "m is None\n",
      "\n",
      "Capitalized words:\n",
      "Python\n",
      "UsE\n",
      "Regular\n",
      "Expressions\n",
      "AKA\n",
      "REGEX\n",
      "\n",
      "On the whole string with SEARCH:\n",
      " Python: starts:20 ends: 27 that is: \" Python\"\n",
      "\n",
      "Iteratively with finditer\n",
      "<class '_sre.SRE_Match'>\n",
      " Python: starts:20 ends: 27 that is: \" Python\"\n",
      "<class '_sre.SRE_Match'>\n",
      " UsE: starts:47 ends: 51 that is: \" UsE\"\n",
      "<class '_sre.SRE_Match'>\n",
      " Regular: starts:51 ends: 59 that is: \" Regular\"\n",
      "<class '_sre.SRE_Match'>\n",
      " Expressions: starts:59 ends: 71 that is: \" Expressions\"\n",
      "<class '_sre.SRE_Match'>\n",
      " AKA: starts:71 ends: 75 that is: \" AKA\"\n",
      "<class '_sre.SRE_Match'>\n",
      " REGEX: starts:75 ends: 81 that is: \" REGEX\"\n"
     ]
    }
   ],
   "source": [
    "%reset -f\n",
    "\n",
    "import re \n",
    "\n",
    "\n",
    "myStr = \"hi there, i am using Python and learning hOw to UsE Regular Expressions AKA REGEX\"\n",
    "\n",
    "m = re.search(\"123\", myStr)\n",
    "print(myStr)\n",
    "if(m):\n",
    "    print(\"123 is in myStr\")\n",
    "else:\n",
    "    print(\"123 is NOT in myStr\")\n",
    "    print(\"m is {}\".format(m))\n",
    "    \n",
    "a = myStr.split()\n",
    "print(\"\\nCapitalized words:\")\n",
    "for word in a:\n",
    "    match = re.match(\"[A-Z]+[a-zA-Z]*\",word)\n",
    "    if(match):\n",
    "        print(word)\n",
    "        \n",
    "print(\"\\nOn the whole string with SEARCH:\")\n",
    "result = re.search(\" [A-Z]+[a-zA-Z]*|^[A-Z]+[a-zA-Z]*| [A-Z]+[a-zA-Z]*$\", myStr)\n",
    "\n",
    "print(\"{}: starts:{} ends: {} that is: \\\"{}\\\"\".format(result.group(), \n",
    "                                        result.start(), \n",
    "                                        result.end(), \n",
    "                                        myStr[result.start():result.end()]))\n",
    "\n",
    "print(\"\\nIteratively with finditer\")\n",
    "#get all the words starting with a capital letter\n",
    "for m in re.finditer(\" [A-Z]+[a-zA-Z]*|^[A-Z]+[a-zA-Z]*| [A-Z]+[a-zA-Z]*$\", myStr):\n",
    "    print(type(m))\n",
    "    print(\"{}: starts:{} ends: {} that is: \\\"{}\\\"\".format(m.group(), \n",
    "                                        m.start(), \n",
    "                                        m.end(), \n",
    "                                        myStr[m.start():m.end()]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Example:**\n",
    "There are three types of valid Uniprot accessions:\n",
    "\n",
    "1. [O,P,Q][0-9][A-Z,0-9][A-Z,0-9][A-Z,0-9][0-9]\n",
    "2. [A-N,R-Z][0-9][A-Z][A-Z,0-9][A-Z,0-9][0-9]\n",
    "3. [A-N,R-Z][0-9][A-Z][A-Z,0-9][A-Z,0-9][0-9][A-Z][A-Z,0-9][A-Z,0-9][0-9]\n",
    "\n",
    "Devise a single regex capable of capturing all of them and use these to filter out the strings that do not encode for valid Uniprot accessions from the following list:\n",
    "\n",
    " [\"P68250\", \"SRX4799393\", \"XT44312B\",\"LFKT12DDF\", \"A0A022YWF9\", \n",
    " \"PRJNA230403\", \"P15163\", \"Q86US8\",\"W0075458831\", \n",
    " \"A0A2K5MMC5\", \"SAMN10177882\",\"Q8LKW0\", \"Q8IEH2_PLAF7\"]\n",
    "\n",
    "Test the strings that passed the filter by accessing: \n",
    "https://www.uniprot.org/uniprot/xxxxxxx where \"xxxxxxx\" is a valid Uniprot accession."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://www.uniprot.org/uniprot/P68250\n",
      "skipping:SRX4799393\n",
      "skipping:XT44312B\n",
      "skipping:LFKT12DDF\n",
      "https://www.uniprot.org/uniprot/A0A022YWF9\n",
      "skipping:PRJNA230403\n",
      "https://www.uniprot.org/uniprot/P15163\n",
      "https://www.uniprot.org/uniprot/Q86US8\n",
      "skipping:W0075458831\n",
      "https://www.uniprot.org/uniprot/A0A2K5MMC5\n",
      "skipping:SAMN10177882\n",
      "https://www.uniprot.org/uniprot/Q8LKW0\n",
      "https://www.uniprot.org/uniprot/Q8IEH2\n"
     ]
    }
   ],
   "source": [
    "%reset -f \n",
    "import re\n",
    "\n",
    "accessionList = [\"P68250\", \"SRX4799393\", \"XT44312B\",\"LFKT12DDF\",\n",
    "                 \"A0A022YWF9\", \"PRJNA230403\", \"P15163\", \n",
    "                 \"Q86US8\",\"W0075458831\", \"A0A2K5MMC5\", \"SAMN10177882\",\n",
    "                \"Q8LKW0\", \"Q8IEH2_PLAF7\"]\n",
    "for el in accessionList:\n",
    "    match = re.match(\"[OPQ][0-9][A-Z0-9]{3}[0-9]|[A-NR-Z][0-9]([A-Z][A-Z0-9]{2}[0-9]){1,2}\",el)\n",
    "    if(match):\n",
    "        print(\"https://www.uniprot.org/uniprot/\" + match.group())\n",
    "    else:\n",
    "        print(\"skipping:\" + el)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Example:**\n",
    "The BisI restriction enzyme cuts the DNA at a specific site (restriction site): \n",
    "```\n",
    "GCNGC where N is any base \n",
    "```\n",
    "given the following DNA string:\n",
    "```\n",
    "DNA = \"ACAAAGACCAAGAAAATGGAATACTCGGAGGGGTAGACCACTAGAAGTGAGCTGCATAAGCGCCGCAAGCGGATTATTTGCCACTCAGTCATTGGCAGGAGTTGCAGGCTGCCAGAGTTCGACAGTTGCAGTATGCCGCGGATCCACGTTTGGAGCAGCGCCAGTTTGGATAGACACCAGGACCCAGAAGCATTACAAAATGCGGCGTAACCTCGCTGTAGACCATGAAACTACGACCGGGTGAATCAGTCCACTCTGCTGCTGGAGCTACCATTGGCAGCACCGGTTGTGGGTAAGGAAAGCTCACCGTTAATCTACTGCAGATGGACAGAGTCCGGGTCCTACACACGCGGCGAGACACCCATTTGTGGCCAGTAGCCGCTTCAGGTCGAGTAATCTGCCAGCAAATCGCGGCGAGCGAGTGACTTGTCGACGAAGCAAAGCAATGCAAACCGTTGGCGGCAGCAATCTCCTGCATATGATGGTGGGACGCTACAATAATCATGTACCAGGCAGCCTATGCATAATATGGCTGCCCACCTACGGGCTATTATAGTGCGACAGCTACCAAAAAAGGTCGCTAGCATGCGCGATACAGCTAACTTTTACGCAGCATACCTTTCCAGCTAGCCGCGATTGT\" \n",
    "```\n",
    "write a regex and some python code that answers the following questions:\n",
    "\n",
    "1. print the start and end position of the first restriction site and the corresponding sequence;\n",
    "2. print all the positions of the restriction sites and the corresponding sequence;\n",
    "3. count how many restrictions sites are there and print a list with all the distances between consecutive restriction sites\n",
    "4. define two groups to get the whole restriction site and also the base in between GCs;\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DNA does NOT start with a restriction site\n",
      "\n",
      "1st Restriction site starts at 50 ends at 55: GCTGC\n",
      "\n",
      "1 res. site starts at 50 ends at 55: GCTGC\n",
      "2 res. site starts at 61 ends at 66: GCCGC\n",
      "3 res. site starts at 107 ends at 112: GCTGC\n",
      "4 res. site starts at 134 ends at 139: GCCGC\n",
      "5 res. site starts at 154 ends at 159: GCAGC\n",
      "6 res. site starts at 201 ends at 206: GCGGC\n",
      "7 res. site starts at 257 ends at 262: GCTGC\n",
      "8 res. site starts at 276 ends at 281: GCAGC\n",
      "9 res. site starts at 349 ends at 354: GCGGC\n",
      "10 res. site starts at 377 ends at 382: GCCGC\n",
      "11 res. site starts at 410 ends at 415: GCGGC\n",
      "12 res. site starts at 458 ends at 463: GCGGC\n",
      "13 res. site starts at 512 ends at 517: GCAGC\n",
      "14 res. site starts at 531 ends at 536: GCTGC\n",
      "15 res. site starts at 609 ends at 614: GCAGC\n",
      "16 res. site starts at 629 ends at 634: GCCGC\n",
      "\n",
      "We have 16 restriction sites\n",
      "Distances:\n",
      "[49, 5, 40, 21, 14, 41, 50, 13, 67, 22, 27, 42, 48, 13, 72, 14]\n",
      "res. site S:50 E:55: GCTGC whole group: GCTGC base: T\n",
      "res. site S:61 E:66: GCCGC whole group: GCCGC base: C\n",
      "res. site S:107 E:112: GCTGC whole group: GCTGC base: T\n",
      "res. site S:134 E:139: GCCGC whole group: GCCGC base: C\n",
      "res. site S:154 E:159: GCAGC whole group: GCAGC base: A\n",
      "res. site S:201 E:206: GCGGC whole group: GCGGC base: G\n",
      "res. site S:257 E:262: GCTGC whole group: GCTGC base: T\n",
      "res. site S:276 E:281: GCAGC whole group: GCAGC base: A\n",
      "res. site S:349 E:354: GCGGC whole group: GCGGC base: G\n",
      "res. site S:377 E:382: GCCGC whole group: GCCGC base: C\n",
      "res. site S:410 E:415: GCGGC whole group: GCGGC base: G\n",
      "res. site S:458 E:463: GCGGC whole group: GCGGC base: G\n",
      "res. site S:512 E:517: GCAGC whole group: GCAGC base: A\n",
      "res. site S:531 E:536: GCTGC whole group: GCTGC base: T\n",
      "res. site S:609 E:614: GCAGC whole group: GCAGC base: A\n",
      "res. site S:629 E:634: GCCGC whole group: GCCGC base: C\n",
      "\n",
      "Findall test (tuple with all groups!):\n",
      "[('GCTGC', 'T'), ('GCCGC', 'C'), ('GCTGC', 'T'), ('GCCGC', 'C'), ('GCAGC', 'A'), ('GCGGC', 'G'), ('GCTGC', 'T'), ('GCAGC', 'A'), ('GCGGC', 'G'), ('GCCGC', 'C'), ('GCGGC', 'G'), ('GCGGC', 'G'), ('GCAGC', 'A'), ('GCTGC', 'T'), ('GCAGC', 'A'), ('GCCGC', 'C')]\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "DNA = \"\"\"ACAAAGACCAAGAAAATGGAATACTCGGAGGGGTAGACCACTAGAAGTGAGCTGCATAAGCGCCGCAAGCGGAT\n",
    "TATTTGCCACTCAGTCATTGGCAGGAGTTGCAGGCTGCCAGAGTTCGACAGTTGCAGTATGCCGCGGATCCACGTTTGGAGCA\n",
    "GCGCCAGTTTGGATAGACACCAGGACCCAGAAGCATTACAAAATGCGGCGTAACCTCGCTGTAGACCATGAAACTACGACCGG\n",
    "GTGAATCAGTCCACTCTGCTGCTGGAGCTACCATTGGCAGCACCGGTTGTGGGTAAGGAAAGCTCACCGTTAATCTACTGCAG\n",
    "ATGGACAGAGTCCGGGTCCTACACACGCGGCGAGACACCCATTTGTGGCCAGTAGCCGCTTCAGGTCGAGTAATCTGCCAGCA\n",
    "AATCGCGGCGAGCGAGTGACTTGTCGACGAAGCAAAGCAATGCAAACCGTTGGCGGCAGCAATCTCCTGCATATGATGGTGGG\n",
    "ACGCTACAATAATCATGTACCAGGCAGCCTATGCATAATATGGCTGCCCACCTACGGGCTATTATAGTGCGACAGCTACCAAA\n",
    "AAAGGTCGCTAGCATGCGCGATACAGCTAACTTTTACGCAGCATACCTTTCCAGCTAGCCGCGATTGT\"\"\" \n",
    "\n",
    "DNA = DNA.replace(\"\\n\",\"\")\n",
    "regex = \"GC[ATCG]GC\"\n",
    "#let's try with match\n",
    "match = re.match(regex,DNA)\n",
    "if match:\n",
    "    print(\"DNA starts with a restriction site\\n\")\n",
    "else:\n",
    "    print(\"DNA does NOT start with a restriction site\\n\")\n",
    "#with search\n",
    "first = re.search(regex,DNA)\n",
    "if first:\n",
    "    print(\"1st Restriction site starts at {} ends at {}: {}\\n\".format(first.start(),\n",
    "                                                                first.end(),\n",
    "                                                                first.group()\n",
    "                                                               ))\n",
    "allrest = re.finditer(regex,DNA)\n",
    "if allrest: \n",
    "    cnt = 0\n",
    "    curPos = 0\n",
    "    distances = []\n",
    "    for site in allrest:\n",
    "        print(\"{} res. site starts at {} ends at {}: {}\".format(cnt + 1,site.start(),\n",
    "                                                                site.end(),\n",
    "                                                                site.group()\n",
    "                                                               ))\n",
    "        cnt +=1\n",
    "        distances.append(site.start() - curPos -1 )\n",
    "        curPos = site.end()\n",
    "\n",
    "print(\"\\nWe have {} restriction sites\".format(cnt))\n",
    "print(\"Distances:\\n{}\".format(distances))\n",
    "\n",
    "regex = \"(GC([ATCG])GC)\"\n",
    "allrest = re.finditer(regex,DNA)\n",
    "if allrest: \n",
    "    for site in allrest:\n",
    "        print(\"res. site S:{} E:{}: {} whole group: {} base: {}\".format(site.start(),\n",
    "                                                                site.end(),\n",
    "                                                                site.group(),\n",
    "                                                                site.groups()[0],\n",
    "                                                                site.groups()[1]\n",
    "                                                               ))\n",
    "\n",
    "print(\"\\nFindall test (tuple with all groups!):\")\n",
    "allres = re.findall(regex,DNA)\n",
    "print(allres)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A lot of information on regex can be found [here](https://docs.python.org/3.7/library/re.html) and [here](https://docs.python.org/3/howto/regex.html).\n",
    "\n",
    "If you want to practice with regular expressions, you can use this website: [https://regex101.com/r/gU7oT6/2](https://regex101.com/r/gU7oT6/2)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercises\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. The following function is supposed to get two lists of integers (let's call then X and Y) and return the list of elements that are contained in both (let's call it B). Is it correct? Devise a unit test to check if it is correct or not (things to check are the length of the result, transitivity, repeated elements in one of the lists, one list empty...). In the latter case propose a correct version of the function.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A, B: [1, 4, 7]\n",
      "A, C: []\n",
      "B, C: [120]\n",
      "A, D: []\n"
     ]
    }
   ],
   "source": [
    "def myListIntersection(X,Y):\n",
    "    tmp = X + Y\n",
    "    vals = [x for x in tmp if tmp.count(x) == 2]\n",
    "    return list(set(vals))\n",
    "\n",
    "A = [1, 2, 3, 4, 7, 12]\n",
    "B = [4, 1, 7, 120]\n",
    "C = [120, 6]\n",
    "D = []\n",
    "\n",
    "print(\"A, B: {}\".format(myListIntersection(A,B)))\n",
    "print(\"A, C: {}\".format(myListIntersection(A,C)))\n",
    "print(\"B, C: {}\".format(myListIntersection(B,C)))\n",
    "print(\"A, D: {}\".format(myListIntersection(A,D)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex1');\">Show/Hide Solution</div>\n",
    "<div id=\"ex1\" style=\"display:none;\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%reset -f\n",
    "\n",
    "import unittest\n",
    "import random\n",
    "\n",
    "#this has problems (test X=[1,1,2,2,3], Y=[4,3])\n",
    "def myListIntersection2(X,Y):\n",
    "    tmp = X + Y\n",
    "    vals = [x for x in tmp if tmp.count(x) == 2]\n",
    "    return list(set(vals))\n",
    "\n",
    "##Correct!\n",
    "def myListIntersection(X,Y):\n",
    "    inx = [ x for x in X if x in Y]\n",
    "    iny = [y for y in Y if y in X]\n",
    "    return list(set(inx + iny))\n",
    "    \n",
    "A = [1, 2, 3, 4, 7, 12]\n",
    "B = [4, 1, 7, 120]\n",
    "C = [120, 6]\n",
    "D = []\n",
    "\n",
    "\n",
    "class Test(unittest.TestCase):\n",
    "    def __init__(self, *args, **kwargs):\n",
    "        super(Test, self).__init__(*args, **kwargs)\n",
    "        #create two test lists:\n",
    "        self.x = []\n",
    "        self.y = []\n",
    "        for i in range(15):    \n",
    "            self.x.append(random.randint(0,10))\n",
    "            self.y.append(random.randint(0,10))\n",
    "        #print(\"{}\\n{}\".format(self.x,self.y))\n",
    "                          \n",
    "    def test_reslen(self):\n",
    "        r = myListIntersection(self.x, self.y)\n",
    "        self.assertFalse(len(r) > len(self.x))\n",
    "        self.assertFalse(len(r) > len(self.y))\n",
    "        s1 = [a for a in self.x if a in self.y]\n",
    "        s2 = [a for a in self.y if a in self.x]\n",
    "        S = set(s1+s2)\n",
    "        self.assertTrue(len(r) == len(S))\n",
    "    \n",
    "    def test_empty(self):\n",
    "            self.assertEqual(myListIntersection(self.x, []),[])\n",
    "            self.assertEqual(myListIntersection([], self.y),[]) \n",
    "    \n",
    "    def test_transitivity(self):\n",
    "            v = myListIntersection(self.x, self.y).sort()\n",
    "            v1 = myListIntersection(self.y, self.x).sort()              \n",
    "            self.assertEqual(v,v1)\n",
    "    \n",
    "    def test_doubleEls(self):\n",
    "        dX = self.x + self.x\n",
    "        dY = self.y + self.y\n",
    "        v1 = myListIntersection(dX,dY)\n",
    "        v1.sort()\n",
    "        v2 = myListIntersection(self.x,self.y)\n",
    "        v2.sort()\n",
    "        self.assertEqual(v1, v2)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    print(\"A, B: {}\".format(myListIntersection(A,B)))\n",
    "    print(\"A, C: {}\".format(myListIntersection(A,C)))\n",
    "    print(\"B, C: {}\".format(myListIntersection(B,C)))\n",
    "    print(\"A, D: {}\".format(myListIntersection(A,D)))   \n",
    "    unittest.main()\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-info\">\n",
    "\n",
    "**Note:** \n",
    "Note the line that I used to initialize the Test class. \n",
    "```\n",
    "def __init__(self, *args, **kwargs):\n",
    "        super(Test, self).__init__(*args, **kwargs)\n",
    "```\n",
    "this allows us to define the random test data within the Test class (these lines are basically because we need to call the super-class constructor with all the parameters it needs). \n",
    "\n",
    "</div>\n",
    "\n",
    "\n",
    "You can find a solution to run unittests here: [pract13_ex1.py](file_samples/pract13_ex1.py)\n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2. CRISPR-Cas9 is quite a neat system to perform genome editing. Guide RNAs (gRNAs) can transport Cas9 to anywhere in the genome for gene editing, but no editing can occur at any site other than one at which Cas9 recognizes the protospacer adjacent motif (PAM). The PAM site is a 2-6 base pair DNA sequence immediately following the DNA sequence targeted by the Cas9 nuclease in the CRISPR bacterial adaptive immune system. Some used PAMs are the following:\n",
    "```\n",
    "NGG (where N is any base)\n",
    "NGA\n",
    "YG (where Y is a Pyrimidine, i.e. C or T)\n",
    "TTTN\n",
    "YTN\n",
    "```\n",
    "write a function that loads the fasta sequences [contig82.fasta](file_samples/contig82.fasta) and for each sequence reports the number of sites of each of the above PAMs and its frequency (i.e. number over the length of the sequence). Hint: load the sequence with biopython and SeqIO.\n",
    "\n",
    "\n",
    "\n",
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex2');\">Show/Hide Solution</div>\n",
    "<div id=\"ex2\" style=\"display:none;\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MDC020656.85 (len:2802):\n",
      "\tNGG:128 (1 on 21.890625 bases)\n",
      "\tNGA:163 (1 on 17.190184049079754 bases)\n",
      "\tYG:226 (1 on 12.398230088495575 bases)\n",
      "\tTTTN:62 (1 on 45.193548387096776 bases)\n",
      "\tYTN:256 (1 on 10.9453125 bases)\n",
      "MDC001115.177 (len:3118):\n",
      "\tNGG:74 (1 on 42.13513513513514 bases)\n",
      "\tNGA:159 (1 on 19.61006289308176 bases)\n",
      "\tYG:215 (1 on 14.502325581395349 bases)\n",
      "\tTTTN:74 (1 on 42.13513513513514 bases)\n",
      "\tYTN:337 (1 on 9.252225519287833 bases)\n",
      "MDC013284.379 (len:5173):\n",
      "\tNGG:182 (1 on 28.423076923076923 bases)\n",
      "\tNGA:307 (1 on 16.850162866449512 bases)\n",
      "\tYG:456 (1 on 11.344298245614034 bases)\n",
      "\tTTTN:162 (1 on 31.932098765432098 bases)\n",
      "\tYTN:577 (1 on 8.965337954939342 bases)\n",
      "MDC018185.243 (len:22724):\n",
      "\tNGG:723 (1 on 31.43015214384509 bases)\n",
      "\tNGA:1230 (1 on 18.47479674796748 bases)\n",
      "\tYG:1932 (1 on 11.761904761904763 bases)\n",
      "\tTTTN:538 (1 on 42.237918215613384 bases)\n",
      "\tYTN:2424 (1 on 9.374587458745875 bases)\n",
      "MDC018185.241 (len:23761):\n",
      "\tNGG:713 (1 on 33.32538569424965 bases)\n",
      "\tNGA:1420 (1 on 16.733098591549297 bases)\n",
      "\tYG:1987 (1 on 11.958228485153498 bases)\n",
      "\tTTTN:615 (1 on 38.635772357723575 bases)\n",
      "\tYTN:2373 (1 on 10.01306363253266 bases)\n",
      "MDC004527.213 (len:3551):\n",
      "\tNGG:86 (1 on 41.2906976744186 bases)\n",
      "\tNGA:136 (1 on 26.110294117647058 bases)\n",
      "\tYG:301 (1 on 11.79734219269103 bases)\n",
      "\tTTTN:85 (1 on 41.77647058823529 bases)\n",
      "\tYTN:440 (1 on 8.070454545454545 bases)\n",
      "MDC003661.174 (len:3334):\n",
      "\tNGG:141 (1 on 23.645390070921987 bases)\n",
      "\tNGA:188 (1 on 17.73404255319149 bases)\n",
      "\tYG:344 (1 on 9.69186046511628 bases)\n",
      "\tTTTN:70 (1 on 47.628571428571426 bases)\n",
      "\tYTN:359 (1 on 9.286908077994429 bases)\n",
      "MDC012176.157 (len:2236):\n",
      "\tNGG:55 (1 on 40.654545454545456 bases)\n",
      "\tNGA:114 (1 on 19.614035087719298 bases)\n",
      "\tYG:164 (1 on 13.634146341463415 bases)\n",
      "\tTTTN:28 (1 on 79.85714285714286 bases)\n",
      "\tYTN:203 (1 on 11.014778325123153 bases)\n",
      "MDC001204.812 (len:5804):\n",
      "\tNGG:199 (1 on 29.165829145728644 bases)\n",
      "\tNGA:328 (1 on 17.695121951219512 bases)\n",
      "\tYG:544 (1 on 10.669117647058824 bases)\n",
      "\tTTTN:145 (1 on 40.02758620689655 bases)\n",
      "\tYTN:620 (1 on 9.361290322580645 bases)\n",
      "MDC001204.810 (len:9798):\n",
      "\tNGG:249 (1 on 39.34939759036145 bases)\n",
      "\tNGA:459 (1 on 21.34640522875817 bases)\n",
      "\tYG:882 (1 on 11.108843537414966 bases)\n",
      "\tTTTN:358 (1 on 27.368715083798882 bases)\n",
      "\tYTN:1150 (1 on 8.52 bases)\n",
      "MDC004389.256 (len:5218):\n",
      "\tNGG:160 (1 on 32.6125 bases)\n",
      "\tNGA:284 (1 on 18.37323943661972 bases)\n",
      "\tYG:433 (1 on 12.05080831408776 bases)\n",
      "\tTTTN:52 (1 on 100.34615384615384 bases)\n",
      "\tYTN:413 (1 on 12.634382566585957 bases)\n",
      "MDC024257.15 (len:4445):\n",
      "\tNGG:138 (1 on 32.210144927536234 bases)\n",
      "\tNGA:233 (1 on 19.07725321888412 bases)\n",
      "\tYG:397 (1 on 11.196473551637279 bases)\n",
      "\tTTTN:120 (1 on 37.041666666666664 bases)\n",
      "\tYTN:473 (1 on 9.397463002114165 bases)\n",
      "MDC018297.229 (len:14841):\n",
      "\tNGG:599 (1 on 24.776293823038397 bases)\n",
      "\tNGA:860 (1 on 17.256976744186048 bases)\n",
      "\tYG:1444 (1 on 10.27770083102493 bases)\n",
      "\tTTTN:408 (1 on 36.375 bases)\n",
      "\tYTN:1597 (1 on 9.293049467752034 bases)\n",
      "MDC001802.364 (len:2238):\n",
      "\tNGG:72 (1 on 31.083333333333332 bases)\n",
      "\tNGA:80 (1 on 27.975 bases)\n",
      "\tYG:222 (1 on 10.08108108108108 bases)\n",
      "\tTTTN:58 (1 on 38.58620689655172 bases)\n",
      "\tYTN:263 (1 on 8.509505703422054 bases)\n",
      "MDC016621.241 (len:3562):\n",
      "\tNGG:82 (1 on 43.4390243902439 bases)\n",
      "\tNGA:199 (1 on 17.899497487437184 bases)\n",
      "\tYG:261 (1 on 13.647509578544062 bases)\n",
      "\tTTTN:50 (1 on 71.24 bases)\n",
      "\tYTN:334 (1 on 10.664670658682635 bases)\n",
      "MDC014057.243 (len:3169):\n",
      "\tNGG:124 (1 on 25.556451612903224 bases)\n",
      "\tNGA:199 (1 on 15.924623115577889 bases)\n",
      "\tYG:288 (1 on 11.003472222222221 bases)\n",
      "\tTTTN:33 (1 on 96.03030303030303 bases)\n",
      "\tYTN:251 (1 on 12.625498007968128 bases)\n",
      "MDC021015.302 (len:3085):\n",
      "\tNGG:168 (1 on 18.363095238095237 bases)\n",
      "\tNGA:235 (1 on 13.127659574468085 bases)\n",
      "\tYG:276 (1 on 11.177536231884059 bases)\n",
      "\tTTTN:46 (1 on 67.06521739130434 bases)\n",
      "\tYTN:193 (1 on 15.984455958549223 bases)\n",
      "MDC018185.242 (len:20357):\n",
      "\tNGG:641 (1 on 31.758190327613104 bases)\n",
      "\tNGA:1144 (1 on 17.79458041958042 bases)\n",
      "\tYG:1776 (1 on 11.462274774774775 bases)\n",
      "\tTTTN:466 (1 on 43.68454935622318 bases)\n",
      "\tYTN:2007 (1 on 10.142999501743896 bases)\n",
      "MDC051782.000 (len:11571):\n",
      "\tNGG:356 (1 on 32.502808988764045 bases)\n",
      "\tNGA:541 (1 on 21.388170055452864 bases)\n",
      "\tYG:1029 (1 on 11.244897959183673 bases)\n",
      "\tTTTN:309 (1 on 37.44660194174757 bases)\n",
      "\tYTN:1236 (1 on 9.361650485436893 bases)\n",
      "MDC017187.314 (len:5870):\n",
      "\tNGG:237 (1 on 24.76793248945148 bases)\n",
      "\tNGA:410 (1 on 14.317073170731707 bases)\n",
      "\tYG:605 (1 on 9.702479338842975 bases)\n",
      "\tTTTN:122 (1 on 48.114754098360656 bases)\n",
      "\tYTN:572 (1 on 10.262237762237762 bases)\n",
      "MDC017187.311 (len:7495):\n",
      "\tNGG:242 (1 on 30.97107438016529 bases)\n",
      "\tNGA:388 (1 on 19.317010309278352 bases)\n",
      "\tYG:691 (1 on 10.846599131693198 bases)\n",
      "\tTTTN:221 (1 on 33.914027149321264 bases)\n",
      "\tYTN:832 (1 on 9.008413461538462 bases)\n",
      "MDC012865.410 (len:3034):\n",
      "\tNGG:121 (1 on 25.074380165289256 bases)\n",
      "\tNGA:192 (1 on 15.802083333333334 bases)\n",
      "\tYG:343 (1 on 8.845481049562682 bases)\n",
      "\tTTTN:87 (1 on 34.87356321839081 bases)\n",
      "\tYTN:315 (1 on 9.631746031746031 bases)\n",
      "MDC000427.83 (len:2226):\n",
      "\tNGG:57 (1 on 39.05263157894737 bases)\n",
      "\tNGA:140 (1 on 15.9 bases)\n",
      "\tYG:145 (1 on 15.351724137931035 bases)\n",
      "\tTTTN:44 (1 on 50.59090909090909 bases)\n",
      "\tYTN:197 (1 on 11.299492385786802 bases)\n",
      "MDC017187.319 (len:13203):\n",
      "\tNGG:484 (1 on 27.27892561983471 bases)\n",
      "\tNGA:731 (1 on 18.06155950752394 bases)\n",
      "\tYG:1261 (1 on 10.47026169706582 bases)\n",
      "\tTTTN:369 (1 on 35.78048780487805 bases)\n",
      "\tYTN:1488 (1 on 8.872983870967742 bases)\n",
      "MDC017187.318 (len:3913):\n",
      "\tNGG:171 (1 on 22.883040935672515 bases)\n",
      "\tNGA:283 (1 on 13.826855123674912 bases)\n",
      "\tYG:428 (1 on 9.142523364485982 bases)\n",
      "\tTTTN:97 (1 on 40.34020618556701 bases)\n",
      "\tYTN:409 (1 on 9.56723716381418 bases)\n",
      "MDC004081.319 (len:2354):\n",
      "\tNGG:35 (1 on 67.25714285714285 bases)\n",
      "\tNGA:65 (1 on 36.215384615384615 bases)\n",
      "\tYG:117 (1 on 20.11965811965812 bases)\n",
      "\tTTTN:44 (1 on 53.5 bases)\n",
      "\tYTN:160 (1 on 14.7125 bases)\n",
      "MDC021913.275 (len:3766):\n",
      "\tNGG:98 (1 on 38.42857142857143 bases)\n",
      "\tNGA:157 (1 on 23.987261146496817 bases)\n",
      "\tYG:250 (1 on 15.064 bases)\n",
      "\tTTTN:38 (1 on 99.10526315789474 bases)\n",
      "\tYTN:230 (1 on 16.37391304347826 bases)\n",
      "MDC015147.205 (len:4626):\n",
      "\tNGG:163 (1 on 28.380368098159508 bases)\n",
      "\tNGA:272 (1 on 17.00735294117647 bases)\n",
      "\tYG:469 (1 on 9.863539445628998 bases)\n",
      "\tTTTN:167 (1 on 27.70059880239521 bases)\n",
      "\tYTN:501 (1 on 9.233532934131736 bases)\n",
      "MDC000038.355 (len:8671):\n",
      "\tNGG:221 (1 on 39.23529411764706 bases)\n",
      "\tNGA:427 (1 on 20.306791569086652 bases)\n",
      "\tYG:642 (1 on 13.506230529595015 bases)\n",
      "\tTTTN:226 (1 on 38.36725663716814 bases)\n",
      "\tYTN:900 (1 on 9.634444444444444 bases)\n",
      "MDC016032.95 (len:3946):\n",
      "\tNGG:169 (1 on 23.349112426035504 bases)\n",
      "\tNGA:291 (1 on 13.560137457044673 bases)\n",
      "\tYG:352 (1 on 11.210227272727273 bases)\n",
      "\tTTTN:82 (1 on 48.1219512195122 bases)\n",
      "\tYTN:363 (1 on 10.870523415977962 bases)\n",
      "MDC052568.000 (len:19684):\n",
      "\tNGG:557 (1 on 35.33931777378815 bases)\n",
      "\tNGA:1025 (1 on 19.20390243902439 bases)\n",
      "\tYG:1729 (1 on 11.384615384615385 bases)\n",
      "\tTTTN:497 (1 on 39.605633802816904 bases)\n",
      "\tYTN:2099 (1 on 9.377798951881848 bases)\n",
      "MDC008119.414 (len:2507):\n",
      "\tNGG:133 (1 on 18.849624060150376 bases)\n",
      "\tNGA:123 (1 on 20.382113821138212 bases)\n",
      "\tYG:263 (1 on 9.532319391634982 bases)\n",
      "\tTTTN:60 (1 on 41.78333333333333 bases)\n",
      "\tYTN:298 (1 on 8.412751677852349 bases)\n",
      "MDC026201.7 (len:2514):\n",
      "\tNGG:87 (1 on 28.896551724137932 bases)\n",
      "\tNGA:157 (1 on 16.012738853503183 bases)\n",
      "\tYG:197 (1 on 12.761421319796954 bases)\n",
      "\tTTTN:34 (1 on 73.94117647058823 bases)\n",
      "\tYTN:236 (1 on 10.652542372881356 bases)\n",
      "MDC003995.601 (len:15483):\n",
      "\tNGG:444 (1 on 34.87162162162162 bases)\n",
      "\tNGA:685 (1 on 22.602919708029198 bases)\n",
      "\tYG:1190 (1 on 13.0109243697479 bases)\n",
      "\tTTTN:483 (1 on 32.05590062111801 bases)\n",
      "\tYTN:1678 (1 on 9.227056019070321 bases)\n",
      "MDC009211.561 (len:7009):\n",
      "\tNGG:185 (1 on 37.88648648648649 bases)\n",
      "\tNGA:359 (1 on 19.52367688022284 bases)\n",
      "\tYG:640 (1 on 10.9515625 bases)\n",
      "\tTTTN:117 (1 on 59.9059829059829 bases)\n",
      "\tYTN:722 (1 on 9.70775623268698 bases)\n",
      "MDC009211.567 (len:3428):\n",
      "\tNGG:115 (1 on 29.808695652173913 bases)\n",
      "\tNGA:157 (1 on 21.8343949044586 bases)\n",
      "\tYG:277 (1 on 12.375451263537906 bases)\n",
      "\tTTTN:122 (1 on 28.098360655737704 bases)\n",
      "\tYTN:383 (1 on 8.950391644908617 bases)\n",
      "MDC054294.001 (len:21361):\n",
      "\tNGG:692 (1 on 30.86849710982659 bases)\n",
      "\tNGA:1113 (1 on 19.19227313566936 bases)\n",
      "\tYG:2014 (1 on 10.606256206554121 bases)\n",
      "\tTTTN:584 (1 on 36.57705479452055 bases)\n",
      "\tYTN:2308 (1 on 9.2551993067591 bases)\n",
      "MDC004364.265 (len:7776):\n",
      "\tNGG:285 (1 on 27.28421052631579 bases)\n",
      "\tNGA:460 (1 on 16.904347826086955 bases)\n",
      "\tYG:652 (1 on 11.92638036809816 bases)\n",
      "\tTTTN:190 (1 on 40.92631578947368 bases)\n",
      "\tYTN:777 (1 on 10.007722007722007 bases)\n",
      "MDC002360.219 (len:2613):\n",
      "\tNGG:80 (1 on 32.6625 bases)\n",
      "\tNGA:142 (1 on 18.401408450704224 bases)\n",
      "\tYG:205 (1 on 12.746341463414634 bases)\n",
      "\tTTTN:61 (1 on 42.83606557377049 bases)\n",
      "\tYTN:291 (1 on 8.97938144329897 bases)\n",
      "MDC003408.117 (len:3669):\n",
      "\tNGG:128 (1 on 28.6640625 bases)\n",
      "\tNGA:209 (1 on 17.555023923444978 bases)\n",
      "\tYG:270 (1 on 13.588888888888889 bases)\n",
      "\tTTTN:60 (1 on 61.15 bases)\n",
      "\tYTN:315 (1 on 11.647619047619047 bases)\n",
      "MDC015155.172 (len:2253):\n",
      "\tNGG:118 (1 on 19.093220338983052 bases)\n",
      "\tNGA:164 (1 on 13.737804878048781 bases)\n",
      "\tYG:235 (1 on 9.587234042553192 bases)\n",
      "\tTTTN:46 (1 on 48.97826086956522 bases)\n",
      "\tYTN:219 (1 on 10.287671232876713 bases)\n",
      "MDC053310.000 (len:12789):\n",
      "\tNGG:420 (1 on 30.45 bases)\n",
      "\tNGA:642 (1 on 19.92056074766355 bases)\n",
      "\tYG:1026 (1 on 12.464912280701755 bases)\n",
      "\tTTTN:328 (1 on 38.99085365853659 bases)\n",
      "\tYTN:1273 (1 on 10.046347211311861 bases)\n",
      "MDC019140.398 (len:3256):\n",
      "\tNGG:103 (1 on 31.611650485436893 bases)\n",
      "\tNGA:163 (1 on 19.975460122699385 bases)\n",
      "\tYG:245 (1 on 13.289795918367346 bases)\n",
      "\tTTTN:68 (1 on 47.88235294117647 bases)\n",
      "\tYTN:324 (1 on 10.049382716049383 bases)\n",
      "MDC019140.399 (len:4394):\n",
      "\tNGG:133 (1 on 33.037593984962406 bases)\n",
      "\tNGA:249 (1 on 17.646586345381525 bases)\n",
      "\tYG:433 (1 on 10.147806004618937 bases)\n",
      "\tTTTN:169 (1 on 26.0 bases)\n",
      "\tYTN:518 (1 on 8.482625482625483 bases)\n",
      "MDC011390.337 (len:7208):\n",
      "\tNGG:232 (1 on 31.06896551724138 bases)\n",
      "\tNGA:307 (1 on 23.478827361563518 bases)\n",
      "\tYG:661 (1 on 10.904689863842663 bases)\n",
      "\tTTTN:144 (1 on 50.05555555555556 bases)\n",
      "\tYTN:827 (1 on 8.71584038694075 bases)\n",
      "MDC007154.375 (len:12886):\n",
      "\tNGG:412 (1 on 31.276699029126213 bases)\n",
      "\tNGA:790 (1 on 16.31139240506329 bases)\n",
      "\tYG:1204 (1 on 10.70265780730897 bases)\n",
      "\tTTTN:286 (1 on 45.05594405594405 bases)\n",
      "\tYTN:1298 (1 on 9.927580893682588 bases)\n",
      "MDC006346.716 (len:7321):\n",
      "\tNGG:237 (1 on 30.890295358649787 bases)\n",
      "\tNGA:409 (1 on 17.899755501222494 bases)\n",
      "\tYG:607 (1 on 12.060955518945635 bases)\n",
      "\tTTTN:170 (1 on 43.06470588235294 bases)\n",
      "\tYTN:739 (1 on 9.906630581867388 bases)\n",
      "MDC010588.505 (len:3598):\n",
      "\tNGG:107 (1 on 33.626168224299064 bases)\n",
      "\tNGA:233 (1 on 15.44206008583691 bases)\n",
      "\tYG:325 (1 on 11.070769230769232 bases)\n",
      "\tTTTN:42 (1 on 85.66666666666667 bases)\n",
      "\tYTN:309 (1 on 11.64401294498382 bases)\n",
      "MDC002519.240 (len:26566):\n",
      "\tNGG:813 (1 on 32.67650676506765 bases)\n",
      "\tNGA:1453 (1 on 18.283551273227804 bases)\n",
      "\tYG:2343 (1 on 11.33845497225779 bases)\n",
      "\tTTTN:808 (1 on 32.87871287128713 bases)\n",
      "\tYTN:2880 (1 on 9.224305555555556 bases)\n",
      "MDC031322.5 (len:5249):\n",
      "\tNGG:159 (1 on 33.0125786163522 bases)\n",
      "\tNGA:377 (1 on 13.923076923076923 bases)\n",
      "\tYG:485 (1 on 10.822680412371135 bases)\n",
      "\tTTTN:132 (1 on 39.765151515151516 bases)\n",
      "\tYTN:489 (1 on 10.734151329243353 bases)\n",
      "MDC010588.502 (len:3858):\n",
      "\tNGG:129 (1 on 29.906976744186046 bases)\n",
      "\tNGA:253 (1 on 15.24901185770751 bases)\n",
      "\tYG:451 (1 on 8.554323725055433 bases)\n",
      "\tTTTN:100 (1 on 38.58 bases)\n",
      "\tYTN:404 (1 on 9.549504950495049 bases)\n",
      "MDC006346.711 (len:20736):\n",
      "\tNGG:653 (1 on 31.754977029096477 bases)\n",
      "\tNGA:1151 (1 on 18.015638575152042 bases)\n",
      "\tYG:1797 (1 on 11.53923205342237 bases)\n",
      "\tTTTN:497 (1 on 41.72233400402415 bases)\n",
      "\tYTN:2040 (1 on 10.16470588235294 bases)\n",
      "MDC011551.182 (len:37892):\n",
      "\tNGG:1158 (1 on 32.721934369602764 bases)\n",
      "\tNGA:2252 (1 on 16.825932504440498 bases)\n",
      "\tYG:3279 (1 on 11.555962183592559 bases)\n",
      "\tTTTN:914 (1 on 41.457330415754925 bases)\n",
      "\tYTN:3924 (1 on 9.656472986748216 bases)\n",
      "MDC002717.156 (len:3104):\n",
      "\tNGG:120 (1 on 25.866666666666667 bases)\n",
      "\tNGA:167 (1 on 18.58682634730539 bases)\n",
      "\tYG:323 (1 on 9.609907120743035 bases)\n",
      "\tTTTN:56 (1 on 55.42857142857143 bases)\n",
      "\tYTN:285 (1 on 10.891228070175439 bases)\n",
      "MDC006346.719 (len:9300):\n",
      "\tNGG:353 (1 on 26.345609065155806 bases)\n",
      "\tNGA:643 (1 on 14.463452566096423 bases)\n",
      "\tYG:965 (1 on 9.637305699481866 bases)\n",
      "\tTTTN:218 (1 on 42.6605504587156 bases)\n",
      "\tYTN:996 (1 on 9.337349397590362 bases)\n",
      "MDC007838.447 (len:6867):\n",
      "\tNGG:186 (1 on 36.91935483870968 bases)\n",
      "\tNGA:303 (1 on 22.663366336633665 bases)\n",
      "\tYG:548 (1 on 12.531021897810218 bases)\n",
      "\tTTTN:165 (1 on 41.61818181818182 bases)\n",
      "\tYTN:719 (1 on 9.550764951321279 bases)\n",
      "MDC007018.186 (len:3002):\n",
      "\tNGG:118 (1 on 25.440677966101696 bases)\n",
      "\tNGA:166 (1 on 18.08433734939759 bases)\n",
      "\tYG:287 (1 on 10.45993031358885 bases)\n",
      "\tTTTN:57 (1 on 52.666666666666664 bases)\n",
      "\tYTN:264 (1 on 11.371212121212121 bases)\n",
      "MDC017873.233 (len:2238):\n",
      "\tNGG:46 (1 on 48.65217391304348 bases)\n",
      "\tNGA:103 (1 on 21.728155339805824 bases)\n",
      "\tYG:139 (1 on 16.100719424460433 bases)\n",
      "\tTTTN:44 (1 on 50.86363636363637 bases)\n",
      "\tYTN:177 (1 on 12.64406779661017 bases)\n",
      "MDC016296.138 (len:8199):\n",
      "\tNGG:251 (1 on 32.66533864541832 bases)\n",
      "\tNGA:435 (1 on 18.848275862068967 bases)\n",
      "\tYG:758 (1 on 10.816622691292876 bases)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tTTTN:249 (1 on 32.9277108433735 bases)\n",
      "\tYTN:947 (1 on 8.657866948257656 bases)\n",
      "MDC019067.226 (len:2238):\n",
      "\tNGG:58 (1 on 38.58620689655172 bases)\n",
      "\tNGA:103 (1 on 21.728155339805824 bases)\n",
      "\tYG:206 (1 on 10.864077669902912 bases)\n",
      "\tTTTN:97 (1 on 23.072164948453608 bases)\n",
      "\tYTN:278 (1 on 8.050359712230216 bases)\n",
      "MDC036568.1 (len:2133):\n",
      "\tNGG:49 (1 on 43.53061224489796 bases)\n",
      "\tNGA:100 (1 on 21.33 bases)\n",
      "\tYG:182 (1 on 11.719780219780219 bases)\n",
      "\tTTTN:53 (1 on 40.24528301886792 bases)\n",
      "\tYTN:241 (1 on 8.850622406639005 bases)\n",
      "MDC014019.318 (len:2896):\n",
      "\tNGG:131 (1 on 22.106870229007633 bases)\n",
      "\tNGA:143 (1 on 20.251748251748253 bases)\n",
      "\tYG:270 (1 on 10.725925925925926 bases)\n",
      "\tTTTN:85 (1 on 34.07058823529412 bases)\n",
      "\tYTN:354 (1 on 8.180790960451978 bases)\n",
      "MDC007995.528 (len:3172):\n",
      "\tNGG:102 (1 on 31.098039215686274 bases)\n",
      "\tNGA:169 (1 on 18.76923076923077 bases)\n",
      "\tYG:283 (1 on 11.208480565371024 bases)\n",
      "\tTTTN:109 (1 on 29.10091743119266 bases)\n",
      "\tYTN:364 (1 on 8.714285714285714 bases)\n",
      "MDC026961.60 (len:4594):\n",
      "\tNGG:179 (1 on 25.664804469273744 bases)\n",
      "\tNGA:262 (1 on 17.53435114503817 bases)\n",
      "\tYG:423 (1 on 10.860520094562649 bases)\n",
      "\tTTTN:101 (1 on 45.48514851485149 bases)\n",
      "\tYTN:455 (1 on 10.096703296703296 bases)\n",
      "MDC013443.168 (len:4331):\n",
      "\tNGG:117 (1 on 37.01709401709402 bases)\n",
      "\tNGA:216 (1 on 20.050925925925927 bases)\n",
      "\tYG:375 (1 on 11.549333333333333 bases)\n",
      "\tTTTN:133 (1 on 32.56390977443609 bases)\n",
      "\tYTN:510 (1 on 8.492156862745098 bases)\n",
      "MDC022800.298 (len:2497):\n",
      "\tNGG:87 (1 on 28.701149425287355 bases)\n",
      "\tNGA:188 (1 on 13.28191489361702 bases)\n",
      "\tYG:264 (1 on 9.458333333333334 bases)\n",
      "\tTTTN:72 (1 on 34.68055555555556 bases)\n",
      "\tYTN:268 (1 on 9.317164179104477 bases)\n",
      "MDC021558.159 (len:3243):\n",
      "\tNGG:98 (1 on 33.09183673469388 bases)\n",
      "\tNGA:162 (1 on 20.01851851851852 bases)\n",
      "\tYG:321 (1 on 10.102803738317757 bases)\n",
      "\tTTTN:86 (1 on 37.7093023255814 bases)\n",
      "\tYTN:335 (1 on 9.680597014925374 bases)\n",
      "MDC002479.192 (len:5825):\n",
      "\tNGG:235 (1 on 24.78723404255319 bases)\n",
      "\tNGA:426 (1 on 13.673708920187794 bases)\n",
      "\tYG:611 (1 on 9.53355155482815 bases)\n",
      "\tTTTN:134 (1 on 43.47014925373134 bases)\n",
      "\tYTN:607 (1 on 9.596375617792422 bases)\n",
      "MDC010751.301 (len:5915):\n",
      "\tNGG:222 (1 on 26.644144144144143 bases)\n",
      "\tNGA:382 (1 on 15.484293193717278 bases)\n",
      "\tYG:580 (1 on 10.198275862068966 bases)\n",
      "\tTTTN:133 (1 on 44.473684210526315 bases)\n",
      "\tYTN:572 (1 on 10.340909090909092 bases)\n",
      "MDC000219.190 (len:9468):\n",
      "\tNGG:249 (1 on 38.024096385542165 bases)\n",
      "\tNGA:481 (1 on 19.683991683991685 bases)\n",
      "\tYG:745 (1 on 12.708724832214765 bases)\n",
      "\tTTTN:329 (1 on 28.778115501519757 bases)\n",
      "\tYTN:1039 (1 on 9.112608277189606 bases)\n",
      "MDC020963.161 (len:2535):\n",
      "\tNGG:51 (1 on 49.705882352941174 bases)\n",
      "\tNGA:105 (1 on 24.142857142857142 bases)\n",
      "\tYG:199 (1 on 12.738693467336683 bases)\n",
      "\tTTTN:55 (1 on 46.09090909090909 bases)\n",
      "\tYTN:319 (1 on 7.946708463949843 bases)\n",
      "MDC020963.162 (len:12994):\n",
      "\tNGG:443 (1 on 29.331828442437924 bases)\n",
      "\tNGA:826 (1 on 15.731234866828087 bases)\n",
      "\tYG:1232 (1 on 10.547077922077921 bases)\n",
      "\tTTTN:231 (1 on 56.25108225108225 bases)\n",
      "\tYTN:1220 (1 on 10.650819672131147 bases)\n",
      "MDC051637.000 (len:14987):\n",
      "\tNGG:423 (1 on 35.43026004728132 bases)\n",
      "\tNGA:706 (1 on 21.228045325779036 bases)\n",
      "\tYG:1292 (1 on 11.59984520123839 bases)\n",
      "\tTTTN:448 (1 on 33.453125 bases)\n",
      "\tYTN:1662 (1 on 9.017448856799037 bases)\n",
      "MDC005174.220 (len:7760):\n",
      "\tNGG:234 (1 on 33.162393162393165 bases)\n",
      "\tNGA:464 (1 on 16.724137931034484 bases)\n",
      "\tYG:667 (1 on 11.634182908545727 bases)\n",
      "\tTTTN:168 (1 on 46.19047619047619 bases)\n",
      "\tYTN:734 (1 on 10.572207084468666 bases)\n",
      "MDC040033.7 (len:7296):\n",
      "\tNGG:262 (1 on 27.84732824427481 bases)\n",
      "\tNGA:463 (1 on 15.758099352051836 bases)\n",
      "\tYG:751 (1 on 9.715046604527297 bases)\n",
      "\tTTTN:133 (1 on 54.857142857142854 bases)\n",
      "\tYTN:760 (1 on 9.6 bases)\n",
      "MDC019674.147 (len:3615):\n",
      "\tNGG:100 (1 on 36.15 bases)\n",
      "\tNGA:203 (1 on 17.807881773399014 bases)\n",
      "\tYG:271 (1 on 13.339483394833948 bases)\n",
      "\tTTTN:148 (1 on 24.425675675675677 bases)\n",
      "\tYTN:444 (1 on 8.141891891891891 bases)\n",
      "MDC010450.877 (len:2432):\n",
      "\tNGG:71 (1 on 34.25352112676056 bases)\n",
      "\tNGA:119 (1 on 20.436974789915965 bases)\n",
      "\tYG:195 (1 on 12.471794871794872 bases)\n",
      "\tTTTN:52 (1 on 46.76923076923077 bases)\n",
      "\tYTN:231 (1 on 10.528138528138529 bases)\n",
      "MDC007097.457 (len:5313):\n",
      "\tNGG:179 (1 on 29.681564245810055 bases)\n",
      "\tNGA:337 (1 on 15.765578635014837 bases)\n",
      "\tYG:512 (1 on 10.376953125 bases)\n",
      "\tTTTN:114 (1 on 46.60526315789474 bases)\n",
      "\tYTN:480 (1 on 11.06875 bases)\n",
      "MDC016278.70 (len:3738):\n",
      "\tNGG:92 (1 on 40.630434782608695 bases)\n",
      "\tNGA:158 (1 on 23.658227848101266 bases)\n",
      "\tYG:265 (1 on 14.105660377358491 bases)\n",
      "\tTTTN:104 (1 on 35.94230769230769 bases)\n",
      "\tYTN:391 (1 on 9.560102301790282 bases)\n",
      "MDC013723.254 (len:2447):\n",
      "\tNGG:75 (1 on 32.626666666666665 bases)\n",
      "\tNGA:122 (1 on 20.057377049180328 bases)\n",
      "\tYG:232 (1 on 10.547413793103448 bases)\n",
      "\tTTTN:64 (1 on 38.234375 bases)\n",
      "\tYTN:295 (1 on 8.294915254237289 bases)\n",
      "MDC002838.179 (len:2601):\n",
      "\tNGG:72 (1 on 36.125 bases)\n",
      "\tNGA:158 (1 on 16.462025316455698 bases)\n",
      "\tYG:178 (1 on 14.612359550561798 bases)\n",
      "\tTTTN:68 (1 on 38.25 bases)\n",
      "\tYTN:252 (1 on 10.321428571428571 bases)\n",
      "MDC009771.217 (len:12554):\n",
      "\tNGG:318 (1 on 39.477987421383645 bases)\n",
      "\tNGA:633 (1 on 19.832543443917853 bases)\n",
      "\tYG:1029 (1 on 12.20019436345967 bases)\n",
      "\tTTTN:366 (1 on 34.30054644808743 bases)\n",
      "\tYTN:1436 (1 on 8.742339832869082 bases)\n"
     ]
    }
   ],
   "source": [
    "%reset -f\n",
    "\n",
    "from Bio import SeqIO\n",
    "import re\n",
    "\n",
    "def countPAMs(filename):\n",
    "    for seq_record in SeqIO.parse(filename, \"fasta\"):\n",
    "        s = seq_record.seq\n",
    "        l = len(s)\n",
    "        ident = seq_record.id\n",
    "        m = re.findall(\"[ATCG]GG\", str(s))\n",
    "        NGG = len(m)\n",
    "        m = re.findall(\"[ATCG]GA\", str(s))\n",
    "        NGA = len(m)\n",
    "        m = re.findall(\"[C|T]G\", str(s))\n",
    "        YG = len(m)\n",
    "        m = re.findall(\"TTT[ATCG]\", str(s))\n",
    "        TTTN = len(m)\n",
    "        m = re.findall(\"[C|T]T[ATCG]\", str(s))\n",
    "        YTN = len(m)\n",
    "        print(\"{} (len:{}):\\n\\tNGG:{} (1 on {} bases)\".format(ident, l,NGG, l/NGG))\n",
    "        print(\"\\tNGA:{} (1 on {} bases)\".format(NGA, l/NGA))\n",
    "        print(\"\\tYG:{} (1 on {} bases)\".format(YG, l/YG))\n",
    "        print(\"\\tTTTN:{} (1 on {} bases)\".format( TTTN, l/TTTN))\n",
    "        print(\"\\tYTN:{} (1 on {} bases)\".format( YTN, l/YTN))\n",
    "        \n",
    "        \n",
    "fn = \"file_samples/contigs82.fasta\"\n",
    "countPAMs(fn)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3. Write a python function ```sortCSV(mystr)``` that gets a comma separated string and returns a comma separated string with all elements sorted in alphabetically decreasing order. Define some unittests to check if the function has been implemented correctly (some things to check: length of initial string is the same as that of the final string, number of elements is the same, each element in the output string must come after the next one in lexicographical order,...). \n",
    "\n",
    "\n",
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex3');\">Show/Hide Solution</div>\n",
    "<div id=\"ex3\" style=\"display:none;\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%reset -f\n",
    "\n",
    "import random\n",
    "import unittest\n",
    "\n",
    "def sortCSV(mystr):\n",
    "    tmp = mystr.split(\",\")\n",
    "    tmp.sort(reverse=True)\n",
    "    return \",\".join(tmp)\n",
    "\n",
    "\n",
    "\n",
    "class Testing(unittest.TestCase):\n",
    "    def __init__(self, *args, **kwargs):\n",
    "        super(Testing, self).__init__(*args, **kwargs)\n",
    "        #create a random string\n",
    "        self.alphabet = \"abcdefghkjilmnopqrstuvwyz\"\n",
    "        self.data = \"\"\n",
    "        #create 15 random strings\n",
    "        for i in range(15):\n",
    "            word = \"\"\n",
    "            #each of them has a random length up to 20\n",
    "            j = random.randint(1,20)\n",
    "            for ind in range(j):\n",
    "                #pick up to 20 random letters\n",
    "                t = random.randint(1,len(self.alphabet)-1)\n",
    "                word += self.alphabet[t]\n",
    "            if(len(self.data) == 0):\n",
    "                self.data = word\n",
    "            else:\n",
    "                self.data += \",\" + word\n",
    "\n",
    "\n",
    "    def test_reslen(self):\n",
    "        self.assertTrue(len(self.data) == len(sortCSV(self.data)))\n",
    "\n",
    "    def test_elcount(self):\n",
    "        res = sortCSV(self.data).split(\",\")\n",
    "        self.assertTrue(len(self.data.split(\",\")) == len(res))\n",
    "\n",
    "    def test_elsorting(self):\n",
    "        res = sortCSV(self.data).split(\",\")\n",
    "        for ind in range(len(res)-1):\n",
    "            self.assertTrue(res[ind]> res[ind+1])\n",
    "\n",
    "    def test_empty(self):\n",
    "        self.assertEqual(sortCSV(\"\"),\"\")\n",
    "    \n",
    "    def test_onlyOne(self):\n",
    "        j = random.randint(1,20)\n",
    "        word = \"\"\n",
    "        for ind in range(j):\n",
    "            #pick up to 20 random letters\n",
    "            t = random.randint(0,len(self.alphabet)-1)\n",
    "            word += self.alphabet[t]\n",
    "        self.assertEqual(sortCSV(word), word)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    mystr = \"book,tree,final,example,testing,zed,all,hair,lady,figure,tap,spring,test,fin,tail\"\n",
    "    print(\"Original:\\n{}\".format(mystr))\n",
    "    print(\"Sorted:\\n{}\".format(sortCSV(mystr)))\n",
    "    unittest.main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "You can find a solution to run unittests here: [pract13_ex3.py](file_samples/pract13_ex3.py)\n",
    "\n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4. Solve a modified version of the following exercise of Practical 11 implementing a function that parses the \"ExpXml\" text through several regular expressions.\n",
    "\n",
    "The exercise is reported below:\n",
    "\n",
    "Write a python script that retrieves all the information present in SRA regarding PacBio sequencing performed on E.coli strain K12 (query term is E.coli K12 wgs PacBio). \n",
    "\n",
    "Print the number of results and for each id report the title, the experiment accession, the instrument, the library strategy, the library source, the total number of spots and total number of bases sequenced.\n",
    "Sample output:\n",
    "\n",
    "```\n",
    "Entries found: 9\n",
    "\n",
    "[1] Results for id 357838:\n",
    "Title: E. coli K12 PacBio RS C2 CCS sequencing\n",
    "Experiment accession: SRX255779\n",
    "Instrument: PACBIO_SMRT\n",
    "Library strategy: WGS\n",
    "Library source: GENOMIC\n",
    "Total spots:1798302\n",
    "Total bases:4228754616\n",
    "\n",
    " ...\n",
    " ...\n",
    "```\n",
    "\n",
    "A sample \"ExpXml\" string:\n",
    "\n",
    "```\n",
    "<Summary><Title>E. coli K12 PacBio RS C2 CCS sequencing</Title><Platform instrument_model=\"PacBio RS\">PACBIO_SMRT</Platform><Statistics total_runs=\"22\" total_spots=\"1798302\" total_bases=\"4228754616\" total_size=\"16799546700\" load_done=\"true\" cluster_name=\"public\"/></Summary><Submitter acc=\"SRA071585\" center_name=\"NBACC\" contact_name=\"Sergey Koren\" lab_name=\"\"/><Experiment acc=\"SRX255779\" ver=\"1\" status=\"public\" name=\"E. coli K12 PacBio RS C2 CCS sequencing\"/><Study acc=\"SRP020003\" name=\"Escherichia coli K12 Re-sequencing\"/><Organism taxid=\"511145\" ScientificName=\"Escherichia coli str. K-12 substr. MG1655\"/><Sample acc=\"SRS000462\" name=\"\"/><Instrument PACBIO_SMRT=\"PacBio RS\"/><Library_descriptor><LIBRARY_NAME>PacBio RS CCS</LIBRARY_NAME><LIBRARY_STRATEGY>WGS</LIBRARY_STRATEGY><LIBRARY_SOURCE>GENOMIC</LIBRARY_SOURCE><LIBRARY_SELECTION>unspecified</LIBRARY_SELECTION><LIBRARY_LAYOUT> <SINGLE/> </LIBRARY_LAYOUT></Library_descriptor><Bioproject>PRJNA194437</Bioproject><Biosample>SAMN00000224</Biosample>\n",
    "```\n",
    "\n",
    "\n",
    "\n",
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex4');\">Show/Hide Solution</div>\n",
    "<div id=\"ex4\" style=\"display:none;\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Entries found: 9\n",
      "\n",
      "[1] Results for id 357838:\n",
      "Title: E. coli K12 PacBio RS C2 CCS sequencing\n",
      "Experiment accession: SRX255779\n",
      "Instrument: PACBIO_SMRT\n",
      "Library strategy: WGS\n",
      "Library source: GENOMIC\n",
      "Total spots:1798302\n",
      "Total bases:4228754616\n",
      "\n",
      "[2] Results for id 357018:\n",
      "Title: E. coli K12 MG1655 PacBio RS C2 Sequencing\n",
      "Experiment accession: SRX255228\n",
      "Instrument: PACBIO_SMRT\n",
      "Library strategy: WGS\n",
      "Library source: GENOMIC\n",
      "Total spots:1389597\n",
      "Total bases:3032527263\n",
      "\n",
      "[3] Results for id 357016:\n",
      "Title: E. Coli K12 454 Sequencing To Use For PacBio RS Correction\n",
      "Experiment accession: SRX255226\n",
      "Instrument: LS454\n",
      "Library strategy: WGS\n",
      "Library source: GENOMIC\n",
      "Total spots:1385764\n",
      "Total bases:720878709\n",
      "\n",
      "[4] Results for id 285700:\n",
      "Title: WGA E. coli genome DNA sequenced by C2 chemistry (Control)\n",
      "Experiment accession: SRX209661\n",
      "Instrument: PACBIO_SMRT\n",
      "Library strategy: WGS\n",
      "Library source: GENOMIC\n",
      "Total spots:245223\n",
      "Total bases:373483301\n",
      "\n",
      "[5] Results for id 285699:\n",
      "Title: WGA E. coli genome DNA sequenced by C2 chemistry (Null)\n",
      "Experiment accession: SRX209660\n",
      "Instrument: PACBIO_SMRT\n",
      "Library strategy: WGS\n",
      "Library source: GENOMIC\n",
      "Total spots:245223\n",
      "Total bases:432837815\n",
      "\n",
      "[6] Results for id 285698:\n",
      "Title: Native E. coli genome DNA sequenced by C2 chemistry\n",
      "Experiment accession: SRX209659\n",
      "Instrument: PACBIO_SMRT\n",
      "Library strategy: WGS\n",
      "Library source: GENOMIC\n",
      "Total spots:245223\n",
      "Total bases:340971357\n",
      "\n",
      "[7] Results for id 285697:\n",
      "Title: WGA E. coli genome DNA sequenced by FCR chemistry\n",
      "Experiment accession: SRX209658\n",
      "Instrument: PACBIO_SMRT\n",
      "Library strategy: WGS\n",
      "Library source: GENOMIC\n",
      "Total spots:2917014\n",
      "Total bases:2838893381\n",
      "\n",
      "[8] Results for id 285673:\n",
      "Title: WGA E. coli Plasmid DNA with m4C\n",
      "Experiment accession: SRX209634\n",
      "Instrument: PACBIO_SMRT\n",
      "Library strategy: WGS\n",
      "Library source: GENOMIC\n",
      "Total spots:121922\n",
      "Total bases:183635674\n",
      "\n",
      "[9] Results for id 285672:\n",
      "Title: Native E. coli plasmid DNA with m4C\n",
      "Experiment accession: SRX209633\n",
      "Instrument: PACBIO_SMRT\n",
      "Library strategy: WGS\n",
      "Library source: GENOMIC\n",
      "Total spots:60961\n",
      "Total bases:168647652\n"
     ]
    }
   ],
   "source": [
    "%reset -f \n",
    "\n",
    "import re\n",
    "from Bio import Entrez\n",
    "\n",
    "\n",
    "\n",
    "def parseExp(expStr):\n",
    "    m = re.search(\"<Title>([A-Za-z0-9_\\. \\(\\)]*)</Title>\", expStr)\n",
    "    if(m):\n",
    "        title = \"Title: \" + m.group()[7:-8]\n",
    "        print(title)\n",
    "        \n",
    "    m = re.search(\"<Experiment acc=\\\"([A-Z0-9]*)\\\"\", expStr)\n",
    "    if(m):\n",
    "        acc = \"Experiment accession: \" + m.groups()[0]\n",
    "        print(acc)\n",
    "    m = re.search(\"<Platform ([A-Za-z0-9_=\\\" \\(\\)]*)>([A-Za-z0-9_\\(\\)]*)</Platform>\", expStr)\n",
    "    if(m):\n",
    "        platform = \"Instrument: \" + m.groups()[1]\n",
    "        print(platform)\n",
    "    m = re.search(\"<LIBRARY_STRATEGY>([A-Za-z0-9_=\\\" \\(\\)]*)</LIBRARY_STRATEGY>\", expStr)\n",
    "    if(m):\n",
    "        src = \"Library strategy: \"  + m.groups()[0]\n",
    "        print(src)\n",
    "    m = re.search(\"<LIBRARY_SOURCE>([A-Za-z0-9_=\\\" \\(\\)]*)</LIBRARY_SOURCE>\", expStr)\n",
    "    if(m):\n",
    "        src = \"Library source: \"  + m.groups()[0]    \n",
    "        print(src)\n",
    "    m = re.search(\"total_spots=\\\"([0-9]*)\\\" total_bases=\\\"([0-9]*)\\\"\",expStr)\n",
    "    if(m):\n",
    "        spots = \"Total spots:\" + m.groups()[0] +\"\\nTotal bases:\" + m.groups()[1]\n",
    "        print(spots)\n",
    "\n",
    "    \n",
    "Entrez.email = \"my_email\"\n",
    "handle = Entrez.esearch(db=\"sra\", term=\"E.coli K12 wgs PacBio\", retmax = 10)\n",
    "res = Entrez.read(handle)\n",
    "#uncomment to see all fields:\n",
    "#for el in res.keys():\n",
    "#    print(el , \" : \", res[el])\n",
    "\n",
    "print(\"Entries found: {}\".format(res[\"Count\"]))\n",
    "\n",
    "cnt = 1\n",
    "for ids in res[\"IdList\"]:\n",
    "    print(\"\\n[{}] Results for id {}:\".format(cnt, ids))\n",
    "    handle = Entrez.esummary(db=\"sra\",  id = ids)\n",
    "    res = Entrez.read(handle)\n",
    "    cnt += 1\n",
    "          \n",
    "    for r in res:\n",
    "        info = r['ExpXml']\n",
    "        #print(info)\n",
    "        parseExp(info)\n",
    "        \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "5. The file (DNA_seq.fasta)[file_samples/DNA_seq.fasta] contains a synthetic DNA sequence. Let's assume to have two restriction enzymes LagI and JagII that respectively cut at the site CNC/ATT, and GAGRK/TNG (where N is any site, R is A or G and K is A or C or T. Note that \"/\" is just a representation of where the enzyme cuts, therefore we do not need to specify this in the regular expression, but we need to take it into account when we cut the DNA.\n",
    "\n",
    "Ex. if the sequence is:\n",
    "ATACATTCCCCCGGAATCGCCCCCCCTCCATTCC\n",
    "digesting the sequence with LagI would give:\n",
    "[\"ATAC\",\"ATTCCCCCGGAATCGCCCCCCCTCC\", \"ATTCC\"]\n",
    "digesting this further with JagII would give:\n",
    "[\"ATAC\",\"ATTCCCCCGGAA\", \"TCGCCCCCCCTCC\", \"ATTCC\"]\n",
    "\n",
    "Write a python script that simulates a digestion with LagI, size selection to keep only the fragments higher than 50 base pairs, and a digestion with JagII, printing the lengths of the obtained fragments. What happens to the fragments if we digest first with JagII and then with LagI?\n",
    "\n",
    "<div class=\"tggle\" onclick=\"toggleVisibility('ex5');\">Show/Hide Solution</div>\n",
    "<div id=\"ex5\" style=\"display:none;\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initial sequence:\n",
      "CTTACATGGCAATAACCCCCCGTTTCTACTTCTAGAGGAGAAAAGTATTGACATGAGCGCTCCCGGCACAAGGGCCAAAGAAGTCTCCAATTTCTTATTTCCGAATGACATGCGTCTCCTTGCGGGTAAATCACCGACCGCAATTCATAGAAGCCTGGGGGAACAGATAGGTCTAATTAGCTTAAGAGAGTAAATCCTGGGATCATTCAGTAGTAACCATAAACTTACGCTGGGGCTTCTTCGGCGGATTTTTACAGTTACCAACCAGGAGATTTGAAGTAAATCAGTTGAGGATTTAGCCGCGCTATCCGGTAATCTCCAAATTAAAACATACCGTTCCATGAAGGCTAGAATTACTTACCGGCCTTTTCCATGCCTGCGCTATACCCCCCCACTCTCCCGCTTATCCGTCCGAGCGGAGGCAGTGCGATCCTCCGTTAAGATATTCTTACGTGTGACGTAGCTATGTATTTTGCAGAGCTGGCGAACGCGTTGAACACTTCACAGATGGTAGGGATTCGGGTAAAGGGCGTATAATTGGGGACTAACATAGGCGTAGACTACGATGGCGCCAACTCAATCGCAGCTCGAGCGCCCTGAATAACGTACTCATCTCAACTCATTCTCGGCAATCTACCGAGCGACTCGATTATCAACGGCTGTCTAGCAGTTCTAATCTTTTGCCAGCATCGTAATAGCCTCCAAGAGATTGATGATAGCTATCGGCACAGAACTGAGACGGCGCCGATGGATAGCGGACTTTCGGTCAACCACAATTCCCCACGGGACAGGTCCTGCGGTGCGCATCACTCTGAATGTACAAGCAACCCAAGTGGGCCGAGCCTGGACTCAGCTGGTTCCTGCGTGAGCTCGAGACTCGGGATGACAGCTCTTTAAACATAGAGCGGGGGCGTCGAACGGTCGAGAAAGTCATAGTACCTCGGGTACCAACTTACTCAGGTTATTGCTTGAAGCTGTACTATTTTAGGGGGGGAGCGCTGAAGGTCTCTTCTTCTCATGACTGAACTCGCGAGGGTCGTGAAGTCGGTTCCTTCAATGGTTAAAAAACAAAGGCTTACTGTGCGCAGAGGAACGCCCATCTAGCGGCTGGCGTCTTGAATGCTCGGTCCCCTTTGTCATTCCGGATTAATCCATTTCCCTCATTCACGAGCTTGCGAAGTCTACATTGGTATATGAATGCGACCTAGAAGAGGGCGCTTAAAATTGGCAGTGGTTGATGCTCTAAACTCCATTTGGTTTACTCGTGCATCACCGCGATAGGCTGACAAAGGTTTAACATTGAATAGCAAGGCACTTCCGGTCTCAATGAACGGCCGGGAAAGGTACGCGCGCGGTATGGGAGGATCAAGGGGCCAATAGAGAGGCTCCTCTCTCACTCGCTAGGAGGCAAATGTAAAACAATGGTTACTGCATCGATACATAAAACATGTCCATCGGTTGCCCAAAGTGTTAAGTGTCTATCACCCCTAGGGCCGTTTCCCGCATATAAACGCCAGGTTGTATCCGCATTTGATGCTACCGTGGATGAGTCTGCGTCGAGCGCGCCGCACGAATGTTGCAATGTATTGCATGAGTAGGGTTGACTAAGAGCCGTTAGATGCGTCGCTGTACTAATAGTTGTCGACAGACCGTCGAGATTAGAAAATGGTACCAGCATTTTCGGAGGTTCTCTAACTAGTATGGATTGCGGTGTCTTCACTGTGCTGCGGCTACCCATCGCCTGAAATCCAGCTGGTGTCAAGCCATCCCCTCTCCGGGACGCCGCATGTAGTGAAACATATACGTTGCACGGGTTCACCGCGGTCCGTTCTGAGTCGACCAAGGACACAATCGAGCTCCGATCCGTACCCTCGACAAACTTGTACCCGACCCCCGGAGCTTGCCAGCTCCTCGGGTATCATGGAGCCTGTGGTTCATCGCGTCCGATATCAAACTTCGTCATGATAAAGTCCCCCCCTCGGGAGTACCAGAGAAGATGACTACTGAGTTGTGCGATAAACGGATAAA\n",
      "\n",
      "\n",
      "LagI restriction:\n",
      "\tRestriction sites:\n",
      "\t618 624 CTCATT\n",
      "\tRestriction sites:\n",
      "\t1159 1165 CTCATT\n",
      "\tRestriction sites:\n",
      "\t1525 1531 CGCATT\n",
      "Lengths:\n",
      "[621, 541, 366, 500]\n",
      "\n",
      "JagII restriction:\n",
      "\tRestriction sites:\n",
      "\t268 276 GAGATTTG\n",
      "\tRestriction sites:\n",
      "\t251 259 GAGACTCG\n",
      "\tRestriction sites:\n",
      "\t126 134 GAGATTAG\n",
      "Final lengths: [273, 348, 256, 285, 366, 131, 369]\n",
      "\n",
      "###########################\n",
      "##### JagII first #########\n",
      "###########################\n",
      "LagI restriction:\n",
      "\tRestriction sites:\n",
      "\t268 276 GAGATTTG\n",
      "\tRestriction sites:\n",
      "\t872 880 GAGACTCG\n",
      "\tRestriction sites:\n",
      "\t1654 1662 GAGATTAG\n",
      "Lengths:\n",
      "[273, 604, 782, 369]\n",
      "\n",
      "JagII restriction:\n",
      "\tRestriction sites:\n",
      "\t345 351 CTCATT\n",
      "\tRestriction sites:\n",
      "\t282 288 CTCATT\n",
      "\tRestriction sites:\n",
      "\t648 654 CGCATT\n",
      "Final lengths: [273, 348, 256, 285, 366, 131, 369]\n"
     ]
    }
   ],
   "source": [
    "%reset -f \n",
    "\n",
    "from Bio import SeqIO\n",
    "import re\n",
    "\n",
    "#in our case overhang can be 3 or 4 (3 for LagI and 4 for JagII)\n",
    "def digestSequence(seq, regex, overhang):\n",
    "    digests = []\n",
    "    sP = 0\n",
    "    matches = re.finditer(regex,seq)\n",
    "    if matches:\n",
    "        for site in matches:\n",
    "            print(\"\\tRestriction sites:\")\n",
    "            print(\"\\t{} {} {}\".format(site.start(),\n",
    "                                   site.end(),\n",
    "                                   site.group()))\n",
    "            digests.append(seq[sP:site.start()+overhang])\n",
    "            sP = site.start() + overhang\n",
    "        #last element:\n",
    "        digests.append(seq[sP:])\n",
    "    return digests\n",
    "\n",
    "fn = \"file_samples/DNA_seq.fasta\"\n",
    "\n",
    "myseq = SeqIO.read(fn, \"fasta\")\n",
    "\n",
    "regexLagI = \"C[ATCG]CATT\"\n",
    "regexJagII = \"GAG[AG][ACT]T[ATCG]G\"\n",
    "\n",
    "\n",
    "s = str(myseq.seq)\n",
    "\n",
    "print(\"Initial sequence:\")\n",
    "print(s)\n",
    "print(\"\\n\")\n",
    "\n",
    "        \n",
    "print(\"LagI restriction:\")\n",
    "digests = digestSequence(s, regexLagI, 3)\n",
    "#filter digests\n",
    "dig = [x for x in digests if len(x) > 50]\n",
    "print(\"Lengths:\")\n",
    "print([len(x) for x in dig])\n",
    "finalDigests = []\n",
    "\n",
    "print(\"\\nJagII restriction:\")\n",
    "for d in dig:\n",
    "    tmp = digestSequence(d, regexJagII, 5)\n",
    "    for t in tmp:\n",
    "        finalDigests.append(t)\n",
    "\n",
    "#print(\"Final digests:\" + str(finalDigests))\n",
    "print(\"Final lengths: \" + str([len(x) for x in finalDigests])) \n",
    "\n",
    "\n",
    "#Other JagII first and LagI second:\n",
    "print(\"\\n###########################\")\n",
    "print(\"##### JagII first #########\")\n",
    "print(\"###########################\")\n",
    "print(\"LagI restriction:\")\n",
    "digests = digestSequence(s, regexJagII, 5)\n",
    "#filter digests\n",
    "dig = [x for x in digests if len(x) > 50]\n",
    "print(\"Lengths:\")\n",
    "print([len(x) for x in dig])\n",
    "finalDigests = []\n",
    "\n",
    "print(\"\\nJagII restriction:\")\n",
    "for d in dig:\n",
    "    tmp = digestSequence(d, regexLagI, 3)\n",
    "    for t in tmp:\n",
    "        finalDigests.append(t)\n",
    "\n",
    "#print(\"Final digests:\" + str(finalDigests))\n",
    "print(\"Final lengths: \" + str([len(x) for x in finalDigests])) \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "</div>"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Edit Metadata",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
